
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="An introduction to GWAS">
      
      
        <meta name="author" content="Matti Pirinen, University of Helsinki">
      
      
        <link rel="canonical" href="https://example.com/WhatIsAGWAS/">
      
      <link rel="icon" href="../images/favicon.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.4.2">
    
    
      
        <title>What is a GWAS? - Genome-Wide Association Studies</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.69437709.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="brightness" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#why-do-we-do-gwas-slides-2-6" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Genome-Wide Association Studies" class="md-header__button md-logo" aria-label="Genome-Wide Association Studies" data-md-component="logo">
      
  
  <svg id="svg" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="400" height="400" viewBox="0, 0, 400,400"><g id="svgg"><path id="path0" d="M252.159 229.900 C 252.285 230.725,252.660 234.190,252.993 237.600 C 253.325 241.010,253.986 247.267,254.461 251.505 C 254.935 255.743,255.242 259.291,255.143 259.390 C 255.044 259.489,253.801 258.480,252.382 257.148 C 245.622 250.806,237.506 246.093,232.113 245.377 C 229.077 244.975,216.726 246.493,204.400 248.784 C 168.099 255.530,160.505 257.424,151.304 262.026 C 144.390 265.484,144.854 265.417,145.220 262.904 C 145.373 261.857,145.965 257.670,146.535 253.600 C 147.106 249.530,147.761 244.940,147.991 243.400 C 149.347 234.312,149.976 229.442,149.814 229.281 C 148.602 228.068,126.791 239.440,112.000 248.997 C 111.230 249.494,106.910 252.199,102.400 255.008 C 93.435 260.590,73.920 273.186,73.600 273.596 C 73.490 273.737,72.545 274.435,71.499 275.146 C 63.066 280.885,52.296 290.509,48.688 295.530 L 47.176 297.635 49.088 298.460 C 50.140 298.914,52.350 299.878,54.000 300.603 C 55.650 301.327,59.790 303.126,63.200 304.600 C 66.610 306.074,70.750 307.874,72.400 308.600 C 74.050 309.326,78.190 311.126,81.600 312.600 C 85.010 314.074,89.150 315.872,90.800 316.596 C 100.451 320.827,107.469 323.890,114.200 326.806 C 118.380 328.618,124.590 331.317,128.000 332.806 C 141.853 338.852,153.544 343.938,163.400 348.206 C 178.480 354.735,191.643 360.462,195.991 362.387 C 200.692 364.467,199.085 364.830,215.400 358.000 C 222.880 354.868,230.485 351.697,232.300 350.953 C 234.115 350.209,237.085 348.992,238.900 348.248 C 240.715 347.504,248.050 344.464,255.200 341.491 C 280.764 330.863,296.635 324.275,298.800 323.392 C 300.010 322.898,307.390 319.841,315.200 316.597 C 323.010 313.354,330.390 310.294,331.600 309.798 C 344.055 304.688,356.228 299.532,356.517 299.243 C 356.982 298.778,356.300 297.507,354.242 295.002 C 347.499 286.797,336.884 278.831,311.284 262.766 C 302.348 257.158,296.621 253.481,288.600 248.199 C 273.968 238.565,266.911 234.595,256.539 230.165 C 251.568 228.041,251.878 228.060,252.159 229.900 " stroke="none" fill="#fbc34b" fill-rule="evenodd"></path><path id="path1" d="M64.400 277.358 C 59.670 279.400,53.370 282.112,50.400 283.385 C 47.430 284.658,43.920 286.181,42.600 286.771 C 41.280 287.361,36.805 289.274,32.655 291.022 C 23.862 294.725,23.422 295.100,24.975 297.560 C 25.539 298.453,26.000 299.343,26.000 299.538 C 26.000 299.733,26.330 300.321,26.733 300.846 C 27.136 301.371,27.901 302.602,28.433 303.583 C 28.965 304.563,29.715 305.688,30.100 306.083 C 30.485 306.477,30.800 307.044,30.800 307.342 C 30.800 307.639,31.250 308.369,31.800 308.962 C 32.350 309.556,32.800 310.126,32.800 310.229 C 32.800 310.529,34.932 313.725,35.700 314.575 C 36.085 315.002,36.400 315.461,36.400 315.596 C 36.400 316.080,42.942 324.669,45.225 327.183 C 45.789 327.804,46.533 328.743,46.878 329.271 C 48.447 331.664,62.148 345.666,67.358 350.200 C 68.875 351.520,70.526 352.983,71.027 353.450 C 71.527 353.918,72.505 354.705,73.200 355.200 C 73.895 355.695,74.944 356.526,75.531 357.047 C 78.066 359.293,78.439 359.600,78.641 359.600 C 78.914 359.600,81.504 361.532,82.796 362.700 C 83.344 363.195,84.000 363.600,84.254 363.600 C 84.508 363.600,85.140 364.047,85.658 364.593 C 86.176 365.139,86.748 365.589,86.928 365.593 C 87.109 365.597,88.392 366.410,89.779 367.400 C 91.166 368.390,92.482 369.200,92.702 369.200 C 92.922 369.200,93.333 369.478,93.616 369.819 C 93.898 370.159,95.090 370.892,96.265 371.448 C 97.439 372.004,98.400 372.592,98.400 372.756 C 98.400 372.919,99.525 373.581,100.900 374.226 C 102.275 374.872,103.595 375.580,103.833 375.800 C 104.255 376.190,114.591 381.378,119.400 383.615 C 120.720 384.229,122.790 385.043,124.000 385.424 C 125.210 385.806,126.298 386.268,126.419 386.451 C 126.653 386.808,132.654 388.800,133.495 388.800 C 133.773 388.800,134.000 388.949,134.000 389.132 C 134.000 389.314,135.845 389.974,138.100 390.599 C 140.355 391.224,142.290 391.850,142.400 391.990 C 142.624 392.275,149.658 394.000,150.596 394.000 C 150.928 394.000,151.200 394.144,151.200 394.320 C 151.200 394.638,158.369 396.165,162.200 396.663 C 163.300 396.806,164.290 397.055,164.400 397.216 C 164.609 397.524,171.117 398.407,179.600 399.279 C 186.331 399.971,213.692 399.976,220.400 399.286 C 228.777 398.425,235.285 397.550,235.960 397.194 C 236.312 397.008,239.390 396.373,242.800 395.782 C 246.210 395.192,249.090 394.583,249.200 394.429 C 249.310 394.275,251.200 393.730,253.400 393.216 C 255.600 392.702,257.490 392.165,257.600 392.021 C 257.710 391.878,259.689 391.259,261.997 390.647 C 264.305 390.035,266.285 389.392,266.397 389.218 C 266.509 389.044,268.220 388.410,270.200 387.809 C 272.180 387.208,273.890 386.588,274.000 386.433 C 274.110 386.277,275.370 385.730,276.800 385.218 C 278.230 384.706,279.766 384.063,280.214 383.791 C 280.662 383.518,282.282 382.796,283.814 382.187 C 285.346 381.578,286.690 380.949,286.800 380.790 C 286.910 380.631,287.990 380.082,289.200 379.571 C 290.410 379.060,291.490 378.519,291.600 378.368 C 291.710 378.217,292.970 377.527,294.400 376.835 C 297.895 375.143,301.296 373.278,301.574 372.901 C 301.698 372.732,302.790 372.065,304.000 371.418 C 305.210 370.772,306.380 370.096,306.600 369.917 C 306.820 369.737,308.530 368.607,310.400 367.404 C 312.270 366.201,314.166 364.853,314.614 364.409 C 315.062 363.964,315.621 363.600,315.856 363.600 C 316.091 363.600,316.731 363.195,317.278 362.700 C 318.586 361.516,320.697 360.000,321.038 360.000 C 321.187 360.000,321.703 359.595,322.185 359.100 C 322.667 358.605,324.018 357.480,325.187 356.600 C 327.283 355.022,328.333 354.150,331.926 351.000 C 337.457 346.151,344.569 339.153,348.558 334.635 C 349.901 333.113,351.360 331.477,351.800 330.998 C 353.342 329.318,355.014 327.294,356.600 325.187 C 357.480 324.018,358.605 322.667,359.100 322.185 C 359.595 321.703,360.000 321.187,360.000 321.038 C 360.000 320.697,361.516 318.586,362.700 317.278 C 363.195 316.731,363.600 316.091,363.600 315.856 C 363.600 315.621,363.964 315.062,364.409 314.614 C 364.853 314.166,366.201 312.270,367.404 310.400 C 368.607 308.530,369.737 306.820,369.917 306.600 C 370.096 306.380,370.773 305.210,371.421 304.000 C 372.070 302.790,372.915 301.451,373.300 301.025 C 374.331 299.883,374.160 299.331,372.700 299.093 C 371.985 298.977,370.410 298.430,369.200 297.877 C 367.990 297.325,363.516 295.321,359.259 293.424 C 355.001 291.526,350.823 289.620,349.975 289.187 C 347.743 288.048,347.766 288.156,350.306 290.700 C 354.686 295.089,357.206 298.554,356.517 299.243 C 356.228 299.532,344.055 304.688,331.600 309.798 C 330.390 310.294,323.010 313.354,315.200 316.597 C 307.390 319.841,300.010 322.898,298.800 323.392 C 296.635 324.275,280.764 330.863,255.200 341.491 C 248.050 344.464,240.715 347.504,238.900 348.248 C 237.085 348.992,234.115 350.209,232.300 350.953 C 230.485 351.697,222.880 354.868,215.400 358.000 C 199.085 364.830,200.692 364.467,195.991 362.387 C 191.643 360.462,178.480 354.735,163.400 348.206 C 153.544 343.938,141.853 338.852,128.000 332.806 C 124.590 331.317,118.380 328.618,114.200 326.806 C 107.469 323.890,100.451 320.827,90.800 316.596 C 89.150 315.872,85.010 314.074,81.600 312.600 C 78.190 311.126,74.050 309.326,72.400 308.600 C 70.750 307.874,66.610 306.074,63.200 304.600 C 59.790 303.126,55.650 301.327,54.000 300.603 C 52.350 299.878,50.140 298.914,49.088 298.460 L 47.176 297.635 48.688 295.530 C 52.276 290.537,60.995 282.685,70.542 275.850 C 72.270 274.612,73.530 273.610,73.342 273.622 C 73.154 273.635,69.130 275.316,64.400 277.358 " stroke="none" fill="#a41c24" fill-rule="evenodd"></path><path id="path2" d="M186.180 0.418 C 179.665 0.790,167.868 2.020,167.600 2.356 C 167.490 2.494,164.340 3.121,160.600 3.749 C 156.860 4.377,153.710 5.014,153.600 5.165 C 153.490 5.316,151.060 5.959,148.200 6.594 C 145.340 7.229,142.910 7.875,142.800 8.029 C 142.690 8.183,140.807 8.785,138.616 9.368 C 136.424 9.950,134.534 10.556,134.416 10.716 C 134.297 10.875,132.580 11.524,130.600 12.156 C 128.620 12.789,126.645 13.552,126.211 13.853 C 125.777 14.154,125.147 14.402,124.811 14.405 C 124.475 14.408,123.175 14.930,121.922 15.566 C 120.668 16.201,118.731 17.024,117.616 17.395 C 116.501 17.765,115.367 18.286,115.095 18.552 C 114.823 18.819,113.385 19.551,111.900 20.178 C 110.415 20.806,109.200 21.473,109.200 21.660 C 109.200 21.847,109.019 22.000,108.798 22.000 C 107.992 22.000,99.396 26.558,98.471 27.476 C 98.181 27.764,97.753 28.000,97.520 28.000 C 97.287 28.000,96.085 28.677,94.849 29.505 C 93.612 30.333,92.105 31.263,91.500 31.573 C 90.895 31.882,90.400 32.285,90.400 32.468 C 90.400 32.650,90.192 32.800,89.938 32.800 C 89.684 32.800,88.886 33.250,88.164 33.800 C 87.443 34.350,86.761 34.800,86.647 34.800 C 86.534 34.800,85.956 35.250,85.362 35.800 C 84.769 36.350,84.050 36.800,83.764 36.800 C 83.479 36.800,83.010 37.131,82.723 37.537 C 82.435 37.942,81.833 38.468,81.384 38.706 C 80.516 39.166,74.292 43.950,72.078 45.860 C 71.350 46.487,70.090 47.570,69.278 48.267 C 68.465 48.963,66.559 50.718,65.042 52.167 C 63.524 53.615,62.174 54.802,62.042 54.805 C 61.742 54.812,54.800 61.776,54.800 62.070 C 54.800 62.187,53.615 63.524,52.167 65.042 C 50.718 66.559,48.963 68.465,48.267 69.278 C 47.570 70.090,46.487 71.350,45.860 72.078 C 43.950 74.292,39.166 80.516,38.706 81.384 C 38.468 81.833,37.942 82.435,37.537 82.723 C 37.131 83.010,36.800 83.479,36.800 83.764 C 36.800 84.050,36.350 84.769,35.800 85.362 C 35.250 85.956,34.800 86.534,34.800 86.647 C 34.800 86.761,34.350 87.443,33.800 88.164 C 33.250 88.886,32.800 89.684,32.800 89.938 C 32.800 90.192,32.650 90.400,32.468 90.400 C 32.285 90.400,31.882 90.895,31.573 91.500 C 31.263 92.105,30.333 93.612,29.505 94.849 C 28.677 96.085,28.000 97.287,28.000 97.520 C 28.000 97.753,27.764 98.181,27.476 98.471 C 26.558 99.396,22.000 107.992,22.000 108.798 C 22.000 109.019,21.847 109.200,21.660 109.200 C 21.473 109.200,20.806 110.415,20.178 111.900 C 19.551 113.385,18.819 114.823,18.552 115.095 C 18.286 115.367,17.765 116.501,17.395 117.616 C 17.024 118.731,16.201 120.668,15.566 121.922 C 14.930 123.175,14.408 124.475,14.405 124.811 C 14.402 125.147,14.144 125.788,13.830 126.235 C 13.517 126.682,12.780 128.567,12.192 130.424 C 11.603 132.281,10.790 134.520,10.384 135.400 C 9.978 136.280,9.359 138.253,9.009 139.785 C 8.659 141.317,8.232 142.667,8.060 142.785 C 7.889 142.903,7.229 145.340,6.594 148.200 C 5.959 151.060,5.316 153.490,5.165 153.600 C 5.014 153.710,4.377 156.860,3.749 160.600 C 3.121 164.340,2.495 167.490,2.358 167.600 C 1.571 168.231,0.007 189.790,0.007 200.000 C 0.007 210.683,1.565 231.850,2.391 232.400 C 2.556 232.510,2.821 233.680,2.978 235.000 C 3.442 238.887,4.828 245.769,5.229 246.180 C 5.433 246.389,5.600 246.904,5.600 247.324 C 5.600 248.431,7.584 256.496,7.976 256.985 C 8.159 257.213,8.702 258.930,9.183 260.800 C 9.664 262.670,10.300 264.740,10.597 265.400 C 10.894 266.060,11.616 268.025,12.203 269.766 C 12.789 271.508,13.497 273.308,13.777 273.766 C 14.057 274.225,14.598 275.635,14.980 276.900 C 15.362 278.165,15.826 279.200,16.010 279.200 C 16.195 279.200,16.461 279.605,16.600 280.100 C 17.187 282.177,18.519 285.200,18.846 285.200 C 19.041 285.200,19.200 285.479,19.200 285.821 C 19.200 286.162,19.950 287.917,20.866 289.721 C 21.782 291.524,22.809 293.585,23.149 294.300 C 23.802 295.674,24.400 296.007,24.400 294.995 C 24.400 294.409,25.852 293.648,31.600 291.223 C 39.017 288.093,51.549 282.711,52.600 282.203 C 53.260 281.884,56.320 280.554,59.400 279.247 C 74.416 272.878,75.117 272.519,84.766 266.245 C 89.955 262.872,97.890 257.814,102.400 255.006 C 106.910 252.199,111.230 249.494,112.000 248.997 C 125.615 240.200,136.697 234.209,146.400 230.402 C 148.270 229.668,149.839 229.043,149.888 229.013 C 150.027 228.926,151.108 221.731,152.384 212.400 C 153.031 207.670,153.829 202.090,154.157 200.000 C 154.485 197.910,154.964 194.459,155.222 192.331 C 155.743 188.032,156.477 186.431,158.496 185.193 C 162.136 182.961,171.320 180.695,180.400 179.789 C 183.810 179.448,187.545 179.073,188.700 178.956 L 190.800 178.742 190.800 132.771 L 190.800 86.800 179.800 86.800 L 168.800 86.800 168.800 75.598 L 168.800 64.395 200.300 64.498 L 231.800 64.600 231.800 75.600 L 231.800 86.600 220.700 86.706 L 209.600 86.813 209.600 132.745 L 209.600 178.678 214.300 178.945 C 227.127 179.673,238.312 181.719,243.800 184.341 C 247.685 186.197,247.791 186.390,248.337 192.600 C 249.205 202.470,251.933 228.059,252.139 228.260 C 252.172 228.293,254.106 229.132,256.435 230.125 C 266.875 234.573,273.916 238.530,288.600 248.199 C 296.621 253.481,302.348 257.158,311.284 262.766 C 327.352 272.850,337.672 279.886,344.112 285.151 C 346.019 286.710,348.619 288.421,349.890 288.955 C 351.160 289.488,355.260 291.310,359.000 293.003 C 369.128 297.589,372.933 299.200,373.600 299.184 C 374.293 299.169,375.984 296.636,377.206 293.785 C 377.671 292.702,378.184 291.722,378.346 291.608 C 378.509 291.494,379.060 290.410,379.571 289.200 C 380.082 287.990,380.631 286.910,380.790 286.800 C 380.949 286.690,381.573 285.340,382.177 283.800 C 382.782 282.260,383.494 280.640,383.761 280.200 C 384.029 279.760,384.575 278.500,384.977 277.400 C 385.378 276.300,386.108 274.500,386.599 273.400 C 387.089 272.300,387.803 270.320,388.185 269.000 C 388.567 267.680,389.013 266.510,389.175 266.400 C 389.489 266.187,391.607 258.946,391.603 258.100 C 391.601 257.825,391.746 257.600,391.925 257.600 C 392.104 257.600,392.672 255.755,393.187 253.500 C 393.701 251.245,394.256 249.220,394.420 249.000 C 394.708 248.613,396.116 241.659,396.648 238.000 C 396.792 237.010,397.048 236.020,397.217 235.800 C 397.542 235.379,398.424 228.802,399.284 220.400 C 399.972 213.671,399.970 186.302,399.281 179.600 C 398.377 170.806,397.562 165.011,397.176 164.620 C 396.969 164.411,396.800 163.827,396.800 163.323 C 396.800 161.939,394.779 151.859,394.410 151.400 C 394.233 151.180,393.778 149.560,393.401 147.800 C 393.023 146.040,392.616 144.150,392.497 143.600 C 392.378 143.050,392.160 142.510,392.013 142.400 C 391.866 142.290,391.241 140.400,390.624 138.200 C 390.007 136.000,389.388 134.110,389.248 134.000 C 389.108 133.890,388.477 132.180,387.844 130.200 C 387.212 128.220,386.582 126.510,386.444 126.400 C 386.306 126.290,385.852 125.210,385.435 124.000 C 385.018 122.790,384.053 120.450,383.290 118.800 C 380.697 113.191,378.645 109.001,378.393 108.800 C 378.254 108.690,377.794 107.735,377.371 106.679 C 376.947 105.622,376.330 104.452,376.000 104.079 C 375.670 103.705,374.964 102.455,374.432 101.300 C 373.899 100.145,373.134 98.902,372.732 98.538 C 372.329 98.174,372.000 97.691,372.000 97.466 C 372.000 97.037,369.783 93.407,367.147 89.520 C 366.296 88.265,365.600 87.059,365.600 86.839 C 365.600 86.620,365.349 86.296,365.043 86.120 C 364.737 85.944,364.152 85.234,363.743 84.542 C 362.884 83.089,357.291 75.595,356.621 75.000 C 356.373 74.780,355.750 74.001,355.236 73.269 C 354.721 72.536,353.918 71.536,353.450 71.047 C 352.983 70.557,351.520 68.906,350.200 67.378 C 346.096 62.628,332.942 49.711,329.831 47.377 C 328.968 46.730,327.765 45.750,327.158 45.200 C 324.645 42.923,316.075 36.400,315.596 36.400 C 315.461 36.400,315.002 36.085,314.575 35.700 C 313.447 34.681,310.485 32.800,310.009 32.800 C 309.784 32.800,309.600 32.623,309.600 32.407 C 309.600 32.190,308.700 31.557,307.600 31.000 C 306.500 30.443,305.600 29.810,305.600 29.593 C 305.600 29.377,305.428 29.200,305.218 29.200 C 305.007 29.200,303.882 28.635,302.718 27.944 C 295.509 23.669,286.873 19.200,285.821 19.200 C 285.479 19.200,285.200 19.041,285.200 18.846 C 285.200 18.519,282.177 17.187,280.100 16.600 C 279.605 16.461,279.200 16.195,279.200 16.010 C 279.200 15.826,278.165 15.362,276.900 14.980 C 275.635 14.598,274.225 14.057,273.766 13.777 C 273.308 13.497,271.508 12.789,269.766 12.203 C 268.025 11.616,266.060 10.894,265.400 10.597 C 264.740 10.300,262.670 9.664,260.800 9.183 C 258.930 8.702,257.213 8.159,256.985 7.976 C 256.496 7.584,248.431 5.600,247.324 5.600 C 246.904 5.600,246.389 5.433,246.180 5.229 C 245.769 4.828,238.887 3.442,235.000 2.978 C 233.680 2.821,232.510 2.556,232.400 2.391 C 231.399 0.886,200.840 -0.418,186.180 0.418 M184.000 75.800 L 184.000 78.809 200.300 78.704 L 216.600 78.600 216.600 75.800 L 216.600 73.000 200.300 72.896 L 184.000 72.791 184.000 75.800 " stroke="none" fill="#b3dbf3" fill-rule="evenodd"></path><path id="path3" d="M168.800 75.800 L 168.800 86.800 180.000 86.800 L 191.200 86.800 191.200 132.813 C 191.200 169.280,191.096 178.855,190.700 178.968 C 190.425 179.047,187.616 179.323,184.457 179.582 C 172.546 180.558,160.814 183.407,157.798 186.055 C 156.211 187.448,155.898 188.596,154.807 197.004 C 154.351 200.522,153.542 206.280,153.011 209.800 C 152.480 213.320,151.671 219.170,151.215 222.800 C 150.759 226.430,150.207 230.390,149.988 231.600 C 149.769 232.810,149.228 236.770,148.786 240.400 C 148.344 244.030,147.728 248.440,147.417 250.200 C 146.604 254.793,145.494 264.027,145.725 264.258 C 145.833 264.367,147.767 263.452,150.021 262.227 C 152.276 261.001,155.219 259.620,156.561 259.158 C 157.902 258.696,159.360 258.155,159.800 257.957 C 162.057 256.941,175.745 253.831,188.600 251.413 C 191.460 250.875,195.060 250.158,196.600 249.819 C 198.140 249.480,202.190 248.750,205.600 248.195 C 209.010 247.641,213.780 246.829,216.200 246.391 C 221.890 245.360,232.822 244.890,235.017 245.582 C 240.193 247.213,247.240 251.771,252.931 257.168 L 254.863 259.000 254.611 257.200 C 254.473 256.210,253.913 250.810,253.366 245.200 C 252.820 239.590,252.190 233.290,251.966 231.200 C 251.742 229.110,251.026 222.090,250.376 215.600 C 249.725 209.110,248.924 201.370,248.595 198.400 C 248.267 195.430,247.982 191.749,247.962 190.220 C 247.928 187.597,247.850 187.370,246.586 186.212 C 243.198 183.110,230.101 180.134,215.926 179.248 C 212.696 179.046,209.862 178.762,209.627 178.617 C 209.166 178.332,208.985 88.625,209.443 87.433 C 209.656 86.877,211.020 86.800,220.643 86.800 L 231.600 86.800 231.600 75.800 L 231.600 64.800 200.200 64.800 L 168.800 64.800 168.800 75.800 M216.800 75.800 L 216.800 78.800 200.200 78.800 L 183.600 78.800 183.600 75.800 L 183.600 72.800 200.200 72.800 L 216.800 72.800 216.800 75.800 " stroke="none" fill="#3c74b4" fill-rule="evenodd"></path><path id="path4" d="M200.194 64.699 L 231.587 64.805 231.703 75.502 L 231.819 86.200 231.810 75.400 L 231.800 64.600 200.300 64.596 L 168.800 64.593 200.194 64.699 M183.737 75.800 C 183.739 77.560,183.809 78.230,183.893 77.289 C 183.978 76.349,183.977 74.909,183.891 74.089 C 183.805 73.270,183.736 74.040,183.737 75.800 M190.898 132.673 L 190.800 178.747 189.700 178.913 L 188.600 179.080 189.713 179.140 C 191.303 179.226,191.215 182.199,191.101 132.400 L 190.997 86.600 190.898 132.673 M209.393 133.000 C 209.393 158.080,209.441 168.399,209.500 155.931 C 209.559 143.463,209.559 122.943,209.500 110.331 C 209.441 97.719,209.393 107.920,209.393 133.000 M212.500 179.087 C 213.105 179.178,214.095 179.178,214.700 179.087 C 215.305 178.995,214.810 178.920,213.600 178.920 C 212.390 178.920,211.895 178.995,212.500 179.087 M172.600 180.800 C 171.530 181.072,171.508 181.105,172.400 181.105 C 172.950 181.105,173.940 180.968,174.600 180.800 C 175.670 180.528,175.692 180.495,174.800 180.495 C 174.250 180.495,173.260 180.632,172.600 180.800 M228.433 180.549 C 228.525 180.631,229.500 180.801,230.600 180.928 C 231.949 181.084,232.274 181.049,231.600 180.820 C 230.601 180.480,228.118 180.268,228.433 180.549 M239.800 182.955 C 240.460 183.255,241.450 183.611,242.000 183.746 C 242.803 183.943,242.764 183.884,241.800 183.445 C 241.140 183.145,240.150 182.789,239.600 182.654 C 238.797 182.457,238.836 182.516,239.800 182.955 M155.849 188.215 C 155.712 188.573,155.615 189.166,155.634 189.533 C 155.653 189.900,155.827 189.669,156.021 189.020 C 156.398 187.758,156.265 187.133,155.849 188.215 M248.681 199.200 C 248.804 200.850,248.971 202.275,249.052 202.367 C 249.331 202.681,249.154 199.401,248.806 197.800 C 248.548 196.614,248.516 196.976,248.681 199.200 M153.415 204.615 C 153.261 205.854,153.217 206.950,153.319 207.052 C 153.420 207.154,153.630 206.223,153.785 204.985 C 153.939 203.746,153.983 202.650,153.881 202.548 C 153.780 202.446,153.570 203.377,153.415 204.615 M249.293 204.400 C 249.293 205.170,249.375 205.485,249.476 205.100 C 249.576 204.715,249.576 204.085,249.476 203.700 C 249.375 203.315,249.293 203.630,249.293 204.400 M249.693 208.000 C 249.693 208.770,249.775 209.085,249.876 208.700 C 249.976 208.315,249.976 207.685,249.876 207.300 C 249.775 206.915,249.693 207.230,249.693 208.000 M150.893 223.200 C 150.893 223.970,150.975 224.285,151.076 223.900 C 151.176 223.515,151.176 222.885,151.076 222.500 C 150.975 222.115,150.893 222.430,150.893 223.200 M251.309 224.000 C 251.309 224.990,251.387 225.395,251.482 224.900 C 251.578 224.405,251.578 223.595,251.482 223.100 C 251.387 222.605,251.309 223.010,251.309 224.000 M150.493 226.000 C 150.493 226.770,150.575 227.085,150.676 226.700 C 150.776 226.315,150.776 225.685,150.676 225.300 C 150.575 224.915,150.493 225.230,150.493 226.000 M251.850 229.887 C 251.967 232.135,252.213 234.475,252.397 235.087 C 252.895 236.746,252.834 235.290,252.193 230.200 L 251.638 225.800 251.850 229.887 M150.064 228.650 C 150.029 229.338,149.911 230.507,149.801 231.250 L 149.603 232.600 150.024 231.258 C 150.255 230.520,150.373 229.350,150.286 228.658 L 150.128 227.400 150.064 228.650 M252.893 239.200 C 252.893 239.970,252.975 240.285,253.076 239.900 C 253.176 239.515,253.176 238.885,253.076 238.500 C 252.975 238.115,252.893 238.430,252.893 239.200 M148.424 240.000 C 148.403 240.330,148.032 243.030,147.600 246.000 C 146.780 251.644,146.681 252.652,146.981 252.353 C 147.303 252.031,148.874 240.418,148.665 239.901 C 148.526 239.556,148.451 239.586,148.424 240.000 M253.309 243.200 C 253.309 244.190,253.387 244.595,253.482 244.100 C 253.578 243.605,253.578 242.795,253.482 242.300 C 253.387 241.805,253.309 242.210,253.309 243.200 M224.905 245.487 C 225.514 245.579,226.414 245.577,226.905 245.482 C 227.397 245.387,226.900 245.311,225.800 245.314 C 224.700 245.317,224.297 245.395,224.905 245.487 M233.607 245.615 C 234.381 245.799,235.881 246.320,236.941 246.774 C 239.171 247.731,239.705 247.824,238.589 247.061 C 237.513 246.327,234.045 245.191,233.029 245.240 C 232.555 245.263,232.803 245.424,233.607 245.615 M216.400 246.320 C 214.310 246.667,212.870 247.008,213.200 247.077 C 213.530 247.145,216.050 246.857,218.800 246.436 C 221.550 246.015,222.990 245.675,222.000 245.679 C 221.010 245.684,218.490 245.972,216.400 246.320 M253.600 246.691 C 253.600 247.758,254.337 255.326,254.643 257.400 L 254.879 259.000 252.940 257.159 C 249.437 253.834,244.400 249.925,244.400 250.531 C 244.400 250.618,245.615 251.596,247.100 252.704 C 248.585 253.812,250.962 255.810,252.382 257.144 C 255.590 260.160,255.437 260.464,254.600 252.715 C 253.954 246.733,253.600 244.602,253.600 246.691 M239.800 247.891 C 240.550 248.598,244.108 250.498,243.825 250.041 C 243.670 249.790,242.611 249.119,241.472 248.549 C 240.332 247.980,239.580 247.684,239.800 247.891 M201.600 248.825 C 197.413 249.611,195.190 250.123,195.336 250.270 C 195.513 250.447,204.909 248.816,205.167 248.564 C 205.420 248.315,203.615 248.447,201.600 248.825 M186.849 251.618 C 183.434 252.301,181.794 252.829,183.233 252.781 C 184.391 252.743,192.562 250.881,191.600 250.874 C 191.050 250.870,188.912 251.205,186.849 251.618 M173.242 254.425 C 170.095 255.135,168.736 255.625,170.033 255.581 C 171.159 255.543,178.575 253.667,177.600 253.667 C 177.050 253.667,175.089 254.008,173.242 254.425 M146.093 257.600 C 146.093 258.370,146.175 258.685,146.276 258.300 C 146.376 257.915,146.376 257.285,146.276 256.900 C 146.175 256.515,146.093 256.830,146.093 257.600 M162.331 257.133 C 161.083 257.500,159.913 257.964,159.731 258.164 C 159.537 258.377,159.695 258.421,160.113 258.271 C 160.505 258.130,161.945 257.726,163.313 257.372 C 164.681 257.018,165.875 256.655,165.967 256.564 C 166.325 256.211,164.490 256.497,162.331 257.133 M155.400 259.530 C 153.159 260.490,148.000 263.175,148.000 263.381 C 148.000 263.532,157.135 259.286,157.567 258.936 C 157.983 258.597,156.856 258.906,155.400 259.530 M145.571 259.900 C 145.555 260.285,145.407 261.547,145.242 262.704 C 145.077 263.862,145.023 264.890,145.122 264.989 C 145.328 265.195,147.600 264.083,147.600 263.776 C 147.600 263.664,147.252 263.758,146.826 263.986 C 145.580 264.653,145.523 264.529,145.815 261.800 C 145.968 260.370,145.982 259.200,145.847 259.200 C 145.711 259.200,145.587 259.515,145.571 259.900 M72.000 273.800 C 71.450 274.114,71.180 274.372,71.400 274.372 C 71.620 274.372,72.250 274.114,72.800 273.800 C 73.350 273.486,73.620 273.228,73.400 273.228 C 73.180 273.228,72.550 273.486,72.000 273.800 M66.400 276.200 C 65.850 276.514,65.580 276.772,65.800 276.772 C 66.020 276.772,66.650 276.514,67.200 276.200 C 67.750 275.886,68.020 275.628,67.800 275.628 C 67.580 275.628,66.950 275.886,66.400 276.200 M48.800 283.800 C 48.250 284.114,47.980 284.372,48.200 284.372 C 48.420 284.372,49.050 284.114,49.600 283.800 C 50.150 283.486,50.420 283.228,50.200 283.228 C 49.980 283.228,49.350 283.486,48.800 283.800 M40.400 287.400 C 39.850 287.714,39.580 287.972,39.800 287.972 C 40.020 287.972,40.650 287.714,41.200 287.400 C 41.750 287.086,42.020 286.828,41.800 286.828 C 41.580 286.828,40.950 287.086,40.400 287.400 M34.800 289.800 C 34.250 290.114,33.980 290.372,34.200 290.372 C 34.420 290.372,35.050 290.114,35.600 289.800 C 36.150 289.486,36.420 289.228,36.200 289.228 C 35.980 289.228,35.350 289.486,34.800 289.800 M351.600 290.000 C 352.370 290.422,353.180 290.766,353.400 290.766 C 353.620 290.766,353.170 290.422,352.400 290.000 C 351.630 289.578,350.820 289.234,350.600 289.234 C 350.380 289.234,350.830 289.578,351.600 290.000 M29.200 292.200 C 28.650 292.514,28.380 292.772,28.600 292.772 C 28.820 292.772,29.450 292.514,30.000 292.200 C 30.550 291.886,30.820 291.628,30.600 291.628 C 30.380 291.628,29.750 291.886,29.200 292.200 M356.400 292.200 C 356.950 292.514,357.580 292.772,357.800 292.772 C 358.020 292.772,357.750 292.514,357.200 292.200 C 356.650 291.886,356.020 291.628,355.800 291.628 C 355.580 291.628,355.850 291.886,356.400 292.200 M363.600 295.400 C 364.150 295.714,364.780 295.972,365.000 295.972 C 365.220 295.972,364.950 295.714,364.400 295.400 C 363.850 295.086,363.220 294.828,363.000 294.828 C 362.780 294.828,363.050 295.086,363.600 295.400 M368.400 297.389 C 368.400 297.706,371.300 298.833,371.528 298.605 C 371.641 298.492,371.635 298.400,371.516 298.400 C 371.396 298.400,370.646 298.128,369.849 297.795 C 369.052 297.462,368.400 297.279,368.400 297.389 " stroke="none" fill="#7484a4" fill-rule="evenodd"></path></g></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Genome-Wide Association Studies
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              What is a GWAS?
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="brightness" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/hds-sandbox/GWAS_course" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    An introduction to GWAS
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href=".." class="md-tabs__link">
      Introduction
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Notebooks/GWAS1-WhatIsAGWAS/" class="md-tabs__link">
        Course Notebooks
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
                
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" hidden>
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Genome-Wide Association Studies" class="md-nav__button md-logo" aria-label="Genome-Wide Association Studies" data-md-component="logo">
      
  
  <svg id="svg" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="400" height="400" viewBox="0, 0, 400,400"><g id="svgg"><path id="path0" d="M252.159 229.900 C 252.285 230.725,252.660 234.190,252.993 237.600 C 253.325 241.010,253.986 247.267,254.461 251.505 C 254.935 255.743,255.242 259.291,255.143 259.390 C 255.044 259.489,253.801 258.480,252.382 257.148 C 245.622 250.806,237.506 246.093,232.113 245.377 C 229.077 244.975,216.726 246.493,204.400 248.784 C 168.099 255.530,160.505 257.424,151.304 262.026 C 144.390 265.484,144.854 265.417,145.220 262.904 C 145.373 261.857,145.965 257.670,146.535 253.600 C 147.106 249.530,147.761 244.940,147.991 243.400 C 149.347 234.312,149.976 229.442,149.814 229.281 C 148.602 228.068,126.791 239.440,112.000 248.997 C 111.230 249.494,106.910 252.199,102.400 255.008 C 93.435 260.590,73.920 273.186,73.600 273.596 C 73.490 273.737,72.545 274.435,71.499 275.146 C 63.066 280.885,52.296 290.509,48.688 295.530 L 47.176 297.635 49.088 298.460 C 50.140 298.914,52.350 299.878,54.000 300.603 C 55.650 301.327,59.790 303.126,63.200 304.600 C 66.610 306.074,70.750 307.874,72.400 308.600 C 74.050 309.326,78.190 311.126,81.600 312.600 C 85.010 314.074,89.150 315.872,90.800 316.596 C 100.451 320.827,107.469 323.890,114.200 326.806 C 118.380 328.618,124.590 331.317,128.000 332.806 C 141.853 338.852,153.544 343.938,163.400 348.206 C 178.480 354.735,191.643 360.462,195.991 362.387 C 200.692 364.467,199.085 364.830,215.400 358.000 C 222.880 354.868,230.485 351.697,232.300 350.953 C 234.115 350.209,237.085 348.992,238.900 348.248 C 240.715 347.504,248.050 344.464,255.200 341.491 C 280.764 330.863,296.635 324.275,298.800 323.392 C 300.010 322.898,307.390 319.841,315.200 316.597 C 323.010 313.354,330.390 310.294,331.600 309.798 C 344.055 304.688,356.228 299.532,356.517 299.243 C 356.982 298.778,356.300 297.507,354.242 295.002 C 347.499 286.797,336.884 278.831,311.284 262.766 C 302.348 257.158,296.621 253.481,288.600 248.199 C 273.968 238.565,266.911 234.595,256.539 230.165 C 251.568 228.041,251.878 228.060,252.159 229.900 " stroke="none" fill="#fbc34b" fill-rule="evenodd"></path><path id="path1" d="M64.400 277.358 C 59.670 279.400,53.370 282.112,50.400 283.385 C 47.430 284.658,43.920 286.181,42.600 286.771 C 41.280 287.361,36.805 289.274,32.655 291.022 C 23.862 294.725,23.422 295.100,24.975 297.560 C 25.539 298.453,26.000 299.343,26.000 299.538 C 26.000 299.733,26.330 300.321,26.733 300.846 C 27.136 301.371,27.901 302.602,28.433 303.583 C 28.965 304.563,29.715 305.688,30.100 306.083 C 30.485 306.477,30.800 307.044,30.800 307.342 C 30.800 307.639,31.250 308.369,31.800 308.962 C 32.350 309.556,32.800 310.126,32.800 310.229 C 32.800 310.529,34.932 313.725,35.700 314.575 C 36.085 315.002,36.400 315.461,36.400 315.596 C 36.400 316.080,42.942 324.669,45.225 327.183 C 45.789 327.804,46.533 328.743,46.878 329.271 C 48.447 331.664,62.148 345.666,67.358 350.200 C 68.875 351.520,70.526 352.983,71.027 353.450 C 71.527 353.918,72.505 354.705,73.200 355.200 C 73.895 355.695,74.944 356.526,75.531 357.047 C 78.066 359.293,78.439 359.600,78.641 359.600 C 78.914 359.600,81.504 361.532,82.796 362.700 C 83.344 363.195,84.000 363.600,84.254 363.600 C 84.508 363.600,85.140 364.047,85.658 364.593 C 86.176 365.139,86.748 365.589,86.928 365.593 C 87.109 365.597,88.392 366.410,89.779 367.400 C 91.166 368.390,92.482 369.200,92.702 369.200 C 92.922 369.200,93.333 369.478,93.616 369.819 C 93.898 370.159,95.090 370.892,96.265 371.448 C 97.439 372.004,98.400 372.592,98.400 372.756 C 98.400 372.919,99.525 373.581,100.900 374.226 C 102.275 374.872,103.595 375.580,103.833 375.800 C 104.255 376.190,114.591 381.378,119.400 383.615 C 120.720 384.229,122.790 385.043,124.000 385.424 C 125.210 385.806,126.298 386.268,126.419 386.451 C 126.653 386.808,132.654 388.800,133.495 388.800 C 133.773 388.800,134.000 388.949,134.000 389.132 C 134.000 389.314,135.845 389.974,138.100 390.599 C 140.355 391.224,142.290 391.850,142.400 391.990 C 142.624 392.275,149.658 394.000,150.596 394.000 C 150.928 394.000,151.200 394.144,151.200 394.320 C 151.200 394.638,158.369 396.165,162.200 396.663 C 163.300 396.806,164.290 397.055,164.400 397.216 C 164.609 397.524,171.117 398.407,179.600 399.279 C 186.331 399.971,213.692 399.976,220.400 399.286 C 228.777 398.425,235.285 397.550,235.960 397.194 C 236.312 397.008,239.390 396.373,242.800 395.782 C 246.210 395.192,249.090 394.583,249.200 394.429 C 249.310 394.275,251.200 393.730,253.400 393.216 C 255.600 392.702,257.490 392.165,257.600 392.021 C 257.710 391.878,259.689 391.259,261.997 390.647 C 264.305 390.035,266.285 389.392,266.397 389.218 C 266.509 389.044,268.220 388.410,270.200 387.809 C 272.180 387.208,273.890 386.588,274.000 386.433 C 274.110 386.277,275.370 385.730,276.800 385.218 C 278.230 384.706,279.766 384.063,280.214 383.791 C 280.662 383.518,282.282 382.796,283.814 382.187 C 285.346 381.578,286.690 380.949,286.800 380.790 C 286.910 380.631,287.990 380.082,289.200 379.571 C 290.410 379.060,291.490 378.519,291.600 378.368 C 291.710 378.217,292.970 377.527,294.400 376.835 C 297.895 375.143,301.296 373.278,301.574 372.901 C 301.698 372.732,302.790 372.065,304.000 371.418 C 305.210 370.772,306.380 370.096,306.600 369.917 C 306.820 369.737,308.530 368.607,310.400 367.404 C 312.270 366.201,314.166 364.853,314.614 364.409 C 315.062 363.964,315.621 363.600,315.856 363.600 C 316.091 363.600,316.731 363.195,317.278 362.700 C 318.586 361.516,320.697 360.000,321.038 360.000 C 321.187 360.000,321.703 359.595,322.185 359.100 C 322.667 358.605,324.018 357.480,325.187 356.600 C 327.283 355.022,328.333 354.150,331.926 351.000 C 337.457 346.151,344.569 339.153,348.558 334.635 C 349.901 333.113,351.360 331.477,351.800 330.998 C 353.342 329.318,355.014 327.294,356.600 325.187 C 357.480 324.018,358.605 322.667,359.100 322.185 C 359.595 321.703,360.000 321.187,360.000 321.038 C 360.000 320.697,361.516 318.586,362.700 317.278 C 363.195 316.731,363.600 316.091,363.600 315.856 C 363.600 315.621,363.964 315.062,364.409 314.614 C 364.853 314.166,366.201 312.270,367.404 310.400 C 368.607 308.530,369.737 306.820,369.917 306.600 C 370.096 306.380,370.773 305.210,371.421 304.000 C 372.070 302.790,372.915 301.451,373.300 301.025 C 374.331 299.883,374.160 299.331,372.700 299.093 C 371.985 298.977,370.410 298.430,369.200 297.877 C 367.990 297.325,363.516 295.321,359.259 293.424 C 355.001 291.526,350.823 289.620,349.975 289.187 C 347.743 288.048,347.766 288.156,350.306 290.700 C 354.686 295.089,357.206 298.554,356.517 299.243 C 356.228 299.532,344.055 304.688,331.600 309.798 C 330.390 310.294,323.010 313.354,315.200 316.597 C 307.390 319.841,300.010 322.898,298.800 323.392 C 296.635 324.275,280.764 330.863,255.200 341.491 C 248.050 344.464,240.715 347.504,238.900 348.248 C 237.085 348.992,234.115 350.209,232.300 350.953 C 230.485 351.697,222.880 354.868,215.400 358.000 C 199.085 364.830,200.692 364.467,195.991 362.387 C 191.643 360.462,178.480 354.735,163.400 348.206 C 153.544 343.938,141.853 338.852,128.000 332.806 C 124.590 331.317,118.380 328.618,114.200 326.806 C 107.469 323.890,100.451 320.827,90.800 316.596 C 89.150 315.872,85.010 314.074,81.600 312.600 C 78.190 311.126,74.050 309.326,72.400 308.600 C 70.750 307.874,66.610 306.074,63.200 304.600 C 59.790 303.126,55.650 301.327,54.000 300.603 C 52.350 299.878,50.140 298.914,49.088 298.460 L 47.176 297.635 48.688 295.530 C 52.276 290.537,60.995 282.685,70.542 275.850 C 72.270 274.612,73.530 273.610,73.342 273.622 C 73.154 273.635,69.130 275.316,64.400 277.358 " stroke="none" fill="#a41c24" fill-rule="evenodd"></path><path id="path2" d="M186.180 0.418 C 179.665 0.790,167.868 2.020,167.600 2.356 C 167.490 2.494,164.340 3.121,160.600 3.749 C 156.860 4.377,153.710 5.014,153.600 5.165 C 153.490 5.316,151.060 5.959,148.200 6.594 C 145.340 7.229,142.910 7.875,142.800 8.029 C 142.690 8.183,140.807 8.785,138.616 9.368 C 136.424 9.950,134.534 10.556,134.416 10.716 C 134.297 10.875,132.580 11.524,130.600 12.156 C 128.620 12.789,126.645 13.552,126.211 13.853 C 125.777 14.154,125.147 14.402,124.811 14.405 C 124.475 14.408,123.175 14.930,121.922 15.566 C 120.668 16.201,118.731 17.024,117.616 17.395 C 116.501 17.765,115.367 18.286,115.095 18.552 C 114.823 18.819,113.385 19.551,111.900 20.178 C 110.415 20.806,109.200 21.473,109.200 21.660 C 109.200 21.847,109.019 22.000,108.798 22.000 C 107.992 22.000,99.396 26.558,98.471 27.476 C 98.181 27.764,97.753 28.000,97.520 28.000 C 97.287 28.000,96.085 28.677,94.849 29.505 C 93.612 30.333,92.105 31.263,91.500 31.573 C 90.895 31.882,90.400 32.285,90.400 32.468 C 90.400 32.650,90.192 32.800,89.938 32.800 C 89.684 32.800,88.886 33.250,88.164 33.800 C 87.443 34.350,86.761 34.800,86.647 34.800 C 86.534 34.800,85.956 35.250,85.362 35.800 C 84.769 36.350,84.050 36.800,83.764 36.800 C 83.479 36.800,83.010 37.131,82.723 37.537 C 82.435 37.942,81.833 38.468,81.384 38.706 C 80.516 39.166,74.292 43.950,72.078 45.860 C 71.350 46.487,70.090 47.570,69.278 48.267 C 68.465 48.963,66.559 50.718,65.042 52.167 C 63.524 53.615,62.174 54.802,62.042 54.805 C 61.742 54.812,54.800 61.776,54.800 62.070 C 54.800 62.187,53.615 63.524,52.167 65.042 C 50.718 66.559,48.963 68.465,48.267 69.278 C 47.570 70.090,46.487 71.350,45.860 72.078 C 43.950 74.292,39.166 80.516,38.706 81.384 C 38.468 81.833,37.942 82.435,37.537 82.723 C 37.131 83.010,36.800 83.479,36.800 83.764 C 36.800 84.050,36.350 84.769,35.800 85.362 C 35.250 85.956,34.800 86.534,34.800 86.647 C 34.800 86.761,34.350 87.443,33.800 88.164 C 33.250 88.886,32.800 89.684,32.800 89.938 C 32.800 90.192,32.650 90.400,32.468 90.400 C 32.285 90.400,31.882 90.895,31.573 91.500 C 31.263 92.105,30.333 93.612,29.505 94.849 C 28.677 96.085,28.000 97.287,28.000 97.520 C 28.000 97.753,27.764 98.181,27.476 98.471 C 26.558 99.396,22.000 107.992,22.000 108.798 C 22.000 109.019,21.847 109.200,21.660 109.200 C 21.473 109.200,20.806 110.415,20.178 111.900 C 19.551 113.385,18.819 114.823,18.552 115.095 C 18.286 115.367,17.765 116.501,17.395 117.616 C 17.024 118.731,16.201 120.668,15.566 121.922 C 14.930 123.175,14.408 124.475,14.405 124.811 C 14.402 125.147,14.144 125.788,13.830 126.235 C 13.517 126.682,12.780 128.567,12.192 130.424 C 11.603 132.281,10.790 134.520,10.384 135.400 C 9.978 136.280,9.359 138.253,9.009 139.785 C 8.659 141.317,8.232 142.667,8.060 142.785 C 7.889 142.903,7.229 145.340,6.594 148.200 C 5.959 151.060,5.316 153.490,5.165 153.600 C 5.014 153.710,4.377 156.860,3.749 160.600 C 3.121 164.340,2.495 167.490,2.358 167.600 C 1.571 168.231,0.007 189.790,0.007 200.000 C 0.007 210.683,1.565 231.850,2.391 232.400 C 2.556 232.510,2.821 233.680,2.978 235.000 C 3.442 238.887,4.828 245.769,5.229 246.180 C 5.433 246.389,5.600 246.904,5.600 247.324 C 5.600 248.431,7.584 256.496,7.976 256.985 C 8.159 257.213,8.702 258.930,9.183 260.800 C 9.664 262.670,10.300 264.740,10.597 265.400 C 10.894 266.060,11.616 268.025,12.203 269.766 C 12.789 271.508,13.497 273.308,13.777 273.766 C 14.057 274.225,14.598 275.635,14.980 276.900 C 15.362 278.165,15.826 279.200,16.010 279.200 C 16.195 279.200,16.461 279.605,16.600 280.100 C 17.187 282.177,18.519 285.200,18.846 285.200 C 19.041 285.200,19.200 285.479,19.200 285.821 C 19.200 286.162,19.950 287.917,20.866 289.721 C 21.782 291.524,22.809 293.585,23.149 294.300 C 23.802 295.674,24.400 296.007,24.400 294.995 C 24.400 294.409,25.852 293.648,31.600 291.223 C 39.017 288.093,51.549 282.711,52.600 282.203 C 53.260 281.884,56.320 280.554,59.400 279.247 C 74.416 272.878,75.117 272.519,84.766 266.245 C 89.955 262.872,97.890 257.814,102.400 255.006 C 106.910 252.199,111.230 249.494,112.000 248.997 C 125.615 240.200,136.697 234.209,146.400 230.402 C 148.270 229.668,149.839 229.043,149.888 229.013 C 150.027 228.926,151.108 221.731,152.384 212.400 C 153.031 207.670,153.829 202.090,154.157 200.000 C 154.485 197.910,154.964 194.459,155.222 192.331 C 155.743 188.032,156.477 186.431,158.496 185.193 C 162.136 182.961,171.320 180.695,180.400 179.789 C 183.810 179.448,187.545 179.073,188.700 178.956 L 190.800 178.742 190.800 132.771 L 190.800 86.800 179.800 86.800 L 168.800 86.800 168.800 75.598 L 168.800 64.395 200.300 64.498 L 231.800 64.600 231.800 75.600 L 231.800 86.600 220.700 86.706 L 209.600 86.813 209.600 132.745 L 209.600 178.678 214.300 178.945 C 227.127 179.673,238.312 181.719,243.800 184.341 C 247.685 186.197,247.791 186.390,248.337 192.600 C 249.205 202.470,251.933 228.059,252.139 228.260 C 252.172 228.293,254.106 229.132,256.435 230.125 C 266.875 234.573,273.916 238.530,288.600 248.199 C 296.621 253.481,302.348 257.158,311.284 262.766 C 327.352 272.850,337.672 279.886,344.112 285.151 C 346.019 286.710,348.619 288.421,349.890 288.955 C 351.160 289.488,355.260 291.310,359.000 293.003 C 369.128 297.589,372.933 299.200,373.600 299.184 C 374.293 299.169,375.984 296.636,377.206 293.785 C 377.671 292.702,378.184 291.722,378.346 291.608 C 378.509 291.494,379.060 290.410,379.571 289.200 C 380.082 287.990,380.631 286.910,380.790 286.800 C 380.949 286.690,381.573 285.340,382.177 283.800 C 382.782 282.260,383.494 280.640,383.761 280.200 C 384.029 279.760,384.575 278.500,384.977 277.400 C 385.378 276.300,386.108 274.500,386.599 273.400 C 387.089 272.300,387.803 270.320,388.185 269.000 C 388.567 267.680,389.013 266.510,389.175 266.400 C 389.489 266.187,391.607 258.946,391.603 258.100 C 391.601 257.825,391.746 257.600,391.925 257.600 C 392.104 257.600,392.672 255.755,393.187 253.500 C 393.701 251.245,394.256 249.220,394.420 249.000 C 394.708 248.613,396.116 241.659,396.648 238.000 C 396.792 237.010,397.048 236.020,397.217 235.800 C 397.542 235.379,398.424 228.802,399.284 220.400 C 399.972 213.671,399.970 186.302,399.281 179.600 C 398.377 170.806,397.562 165.011,397.176 164.620 C 396.969 164.411,396.800 163.827,396.800 163.323 C 396.800 161.939,394.779 151.859,394.410 151.400 C 394.233 151.180,393.778 149.560,393.401 147.800 C 393.023 146.040,392.616 144.150,392.497 143.600 C 392.378 143.050,392.160 142.510,392.013 142.400 C 391.866 142.290,391.241 140.400,390.624 138.200 C 390.007 136.000,389.388 134.110,389.248 134.000 C 389.108 133.890,388.477 132.180,387.844 130.200 C 387.212 128.220,386.582 126.510,386.444 126.400 C 386.306 126.290,385.852 125.210,385.435 124.000 C 385.018 122.790,384.053 120.450,383.290 118.800 C 380.697 113.191,378.645 109.001,378.393 108.800 C 378.254 108.690,377.794 107.735,377.371 106.679 C 376.947 105.622,376.330 104.452,376.000 104.079 C 375.670 103.705,374.964 102.455,374.432 101.300 C 373.899 100.145,373.134 98.902,372.732 98.538 C 372.329 98.174,372.000 97.691,372.000 97.466 C 372.000 97.037,369.783 93.407,367.147 89.520 C 366.296 88.265,365.600 87.059,365.600 86.839 C 365.600 86.620,365.349 86.296,365.043 86.120 C 364.737 85.944,364.152 85.234,363.743 84.542 C 362.884 83.089,357.291 75.595,356.621 75.000 C 356.373 74.780,355.750 74.001,355.236 73.269 C 354.721 72.536,353.918 71.536,353.450 71.047 C 352.983 70.557,351.520 68.906,350.200 67.378 C 346.096 62.628,332.942 49.711,329.831 47.377 C 328.968 46.730,327.765 45.750,327.158 45.200 C 324.645 42.923,316.075 36.400,315.596 36.400 C 315.461 36.400,315.002 36.085,314.575 35.700 C 313.447 34.681,310.485 32.800,310.009 32.800 C 309.784 32.800,309.600 32.623,309.600 32.407 C 309.600 32.190,308.700 31.557,307.600 31.000 C 306.500 30.443,305.600 29.810,305.600 29.593 C 305.600 29.377,305.428 29.200,305.218 29.200 C 305.007 29.200,303.882 28.635,302.718 27.944 C 295.509 23.669,286.873 19.200,285.821 19.200 C 285.479 19.200,285.200 19.041,285.200 18.846 C 285.200 18.519,282.177 17.187,280.100 16.600 C 279.605 16.461,279.200 16.195,279.200 16.010 C 279.200 15.826,278.165 15.362,276.900 14.980 C 275.635 14.598,274.225 14.057,273.766 13.777 C 273.308 13.497,271.508 12.789,269.766 12.203 C 268.025 11.616,266.060 10.894,265.400 10.597 C 264.740 10.300,262.670 9.664,260.800 9.183 C 258.930 8.702,257.213 8.159,256.985 7.976 C 256.496 7.584,248.431 5.600,247.324 5.600 C 246.904 5.600,246.389 5.433,246.180 5.229 C 245.769 4.828,238.887 3.442,235.000 2.978 C 233.680 2.821,232.510 2.556,232.400 2.391 C 231.399 0.886,200.840 -0.418,186.180 0.418 M184.000 75.800 L 184.000 78.809 200.300 78.704 L 216.600 78.600 216.600 75.800 L 216.600 73.000 200.300 72.896 L 184.000 72.791 184.000 75.800 " stroke="none" fill="#b3dbf3" fill-rule="evenodd"></path><path id="path3" d="M168.800 75.800 L 168.800 86.800 180.000 86.800 L 191.200 86.800 191.200 132.813 C 191.200 169.280,191.096 178.855,190.700 178.968 C 190.425 179.047,187.616 179.323,184.457 179.582 C 172.546 180.558,160.814 183.407,157.798 186.055 C 156.211 187.448,155.898 188.596,154.807 197.004 C 154.351 200.522,153.542 206.280,153.011 209.800 C 152.480 213.320,151.671 219.170,151.215 222.800 C 150.759 226.430,150.207 230.390,149.988 231.600 C 149.769 232.810,149.228 236.770,148.786 240.400 C 148.344 244.030,147.728 248.440,147.417 250.200 C 146.604 254.793,145.494 264.027,145.725 264.258 C 145.833 264.367,147.767 263.452,150.021 262.227 C 152.276 261.001,155.219 259.620,156.561 259.158 C 157.902 258.696,159.360 258.155,159.800 257.957 C 162.057 256.941,175.745 253.831,188.600 251.413 C 191.460 250.875,195.060 250.158,196.600 249.819 C 198.140 249.480,202.190 248.750,205.600 248.195 C 209.010 247.641,213.780 246.829,216.200 246.391 C 221.890 245.360,232.822 244.890,235.017 245.582 C 240.193 247.213,247.240 251.771,252.931 257.168 L 254.863 259.000 254.611 257.200 C 254.473 256.210,253.913 250.810,253.366 245.200 C 252.820 239.590,252.190 233.290,251.966 231.200 C 251.742 229.110,251.026 222.090,250.376 215.600 C 249.725 209.110,248.924 201.370,248.595 198.400 C 248.267 195.430,247.982 191.749,247.962 190.220 C 247.928 187.597,247.850 187.370,246.586 186.212 C 243.198 183.110,230.101 180.134,215.926 179.248 C 212.696 179.046,209.862 178.762,209.627 178.617 C 209.166 178.332,208.985 88.625,209.443 87.433 C 209.656 86.877,211.020 86.800,220.643 86.800 L 231.600 86.800 231.600 75.800 L 231.600 64.800 200.200 64.800 L 168.800 64.800 168.800 75.800 M216.800 75.800 L 216.800 78.800 200.200 78.800 L 183.600 78.800 183.600 75.800 L 183.600 72.800 200.200 72.800 L 216.800 72.800 216.800 75.800 " stroke="none" fill="#3c74b4" fill-rule="evenodd"></path><path id="path4" d="M200.194 64.699 L 231.587 64.805 231.703 75.502 L 231.819 86.200 231.810 75.400 L 231.800 64.600 200.300 64.596 L 168.800 64.593 200.194 64.699 M183.737 75.800 C 183.739 77.560,183.809 78.230,183.893 77.289 C 183.978 76.349,183.977 74.909,183.891 74.089 C 183.805 73.270,183.736 74.040,183.737 75.800 M190.898 132.673 L 190.800 178.747 189.700 178.913 L 188.600 179.080 189.713 179.140 C 191.303 179.226,191.215 182.199,191.101 132.400 L 190.997 86.600 190.898 132.673 M209.393 133.000 C 209.393 158.080,209.441 168.399,209.500 155.931 C 209.559 143.463,209.559 122.943,209.500 110.331 C 209.441 97.719,209.393 107.920,209.393 133.000 M212.500 179.087 C 213.105 179.178,214.095 179.178,214.700 179.087 C 215.305 178.995,214.810 178.920,213.600 178.920 C 212.390 178.920,211.895 178.995,212.500 179.087 M172.600 180.800 C 171.530 181.072,171.508 181.105,172.400 181.105 C 172.950 181.105,173.940 180.968,174.600 180.800 C 175.670 180.528,175.692 180.495,174.800 180.495 C 174.250 180.495,173.260 180.632,172.600 180.800 M228.433 180.549 C 228.525 180.631,229.500 180.801,230.600 180.928 C 231.949 181.084,232.274 181.049,231.600 180.820 C 230.601 180.480,228.118 180.268,228.433 180.549 M239.800 182.955 C 240.460 183.255,241.450 183.611,242.000 183.746 C 242.803 183.943,242.764 183.884,241.800 183.445 C 241.140 183.145,240.150 182.789,239.600 182.654 C 238.797 182.457,238.836 182.516,239.800 182.955 M155.849 188.215 C 155.712 188.573,155.615 189.166,155.634 189.533 C 155.653 189.900,155.827 189.669,156.021 189.020 C 156.398 187.758,156.265 187.133,155.849 188.215 M248.681 199.200 C 248.804 200.850,248.971 202.275,249.052 202.367 C 249.331 202.681,249.154 199.401,248.806 197.800 C 248.548 196.614,248.516 196.976,248.681 199.200 M153.415 204.615 C 153.261 205.854,153.217 206.950,153.319 207.052 C 153.420 207.154,153.630 206.223,153.785 204.985 C 153.939 203.746,153.983 202.650,153.881 202.548 C 153.780 202.446,153.570 203.377,153.415 204.615 M249.293 204.400 C 249.293 205.170,249.375 205.485,249.476 205.100 C 249.576 204.715,249.576 204.085,249.476 203.700 C 249.375 203.315,249.293 203.630,249.293 204.400 M249.693 208.000 C 249.693 208.770,249.775 209.085,249.876 208.700 C 249.976 208.315,249.976 207.685,249.876 207.300 C 249.775 206.915,249.693 207.230,249.693 208.000 M150.893 223.200 C 150.893 223.970,150.975 224.285,151.076 223.900 C 151.176 223.515,151.176 222.885,151.076 222.500 C 150.975 222.115,150.893 222.430,150.893 223.200 M251.309 224.000 C 251.309 224.990,251.387 225.395,251.482 224.900 C 251.578 224.405,251.578 223.595,251.482 223.100 C 251.387 222.605,251.309 223.010,251.309 224.000 M150.493 226.000 C 150.493 226.770,150.575 227.085,150.676 226.700 C 150.776 226.315,150.776 225.685,150.676 225.300 C 150.575 224.915,150.493 225.230,150.493 226.000 M251.850 229.887 C 251.967 232.135,252.213 234.475,252.397 235.087 C 252.895 236.746,252.834 235.290,252.193 230.200 L 251.638 225.800 251.850 229.887 M150.064 228.650 C 150.029 229.338,149.911 230.507,149.801 231.250 L 149.603 232.600 150.024 231.258 C 150.255 230.520,150.373 229.350,150.286 228.658 L 150.128 227.400 150.064 228.650 M252.893 239.200 C 252.893 239.970,252.975 240.285,253.076 239.900 C 253.176 239.515,253.176 238.885,253.076 238.500 C 252.975 238.115,252.893 238.430,252.893 239.200 M148.424 240.000 C 148.403 240.330,148.032 243.030,147.600 246.000 C 146.780 251.644,146.681 252.652,146.981 252.353 C 147.303 252.031,148.874 240.418,148.665 239.901 C 148.526 239.556,148.451 239.586,148.424 240.000 M253.309 243.200 C 253.309 244.190,253.387 244.595,253.482 244.100 C 253.578 243.605,253.578 242.795,253.482 242.300 C 253.387 241.805,253.309 242.210,253.309 243.200 M224.905 245.487 C 225.514 245.579,226.414 245.577,226.905 245.482 C 227.397 245.387,226.900 245.311,225.800 245.314 C 224.700 245.317,224.297 245.395,224.905 245.487 M233.607 245.615 C 234.381 245.799,235.881 246.320,236.941 246.774 C 239.171 247.731,239.705 247.824,238.589 247.061 C 237.513 246.327,234.045 245.191,233.029 245.240 C 232.555 245.263,232.803 245.424,233.607 245.615 M216.400 246.320 C 214.310 246.667,212.870 247.008,213.200 247.077 C 213.530 247.145,216.050 246.857,218.800 246.436 C 221.550 246.015,222.990 245.675,222.000 245.679 C 221.010 245.684,218.490 245.972,216.400 246.320 M253.600 246.691 C 253.600 247.758,254.337 255.326,254.643 257.400 L 254.879 259.000 252.940 257.159 C 249.437 253.834,244.400 249.925,244.400 250.531 C 244.400 250.618,245.615 251.596,247.100 252.704 C 248.585 253.812,250.962 255.810,252.382 257.144 C 255.590 260.160,255.437 260.464,254.600 252.715 C 253.954 246.733,253.600 244.602,253.600 246.691 M239.800 247.891 C 240.550 248.598,244.108 250.498,243.825 250.041 C 243.670 249.790,242.611 249.119,241.472 248.549 C 240.332 247.980,239.580 247.684,239.800 247.891 M201.600 248.825 C 197.413 249.611,195.190 250.123,195.336 250.270 C 195.513 250.447,204.909 248.816,205.167 248.564 C 205.420 248.315,203.615 248.447,201.600 248.825 M186.849 251.618 C 183.434 252.301,181.794 252.829,183.233 252.781 C 184.391 252.743,192.562 250.881,191.600 250.874 C 191.050 250.870,188.912 251.205,186.849 251.618 M173.242 254.425 C 170.095 255.135,168.736 255.625,170.033 255.581 C 171.159 255.543,178.575 253.667,177.600 253.667 C 177.050 253.667,175.089 254.008,173.242 254.425 M146.093 257.600 C 146.093 258.370,146.175 258.685,146.276 258.300 C 146.376 257.915,146.376 257.285,146.276 256.900 C 146.175 256.515,146.093 256.830,146.093 257.600 M162.331 257.133 C 161.083 257.500,159.913 257.964,159.731 258.164 C 159.537 258.377,159.695 258.421,160.113 258.271 C 160.505 258.130,161.945 257.726,163.313 257.372 C 164.681 257.018,165.875 256.655,165.967 256.564 C 166.325 256.211,164.490 256.497,162.331 257.133 M155.400 259.530 C 153.159 260.490,148.000 263.175,148.000 263.381 C 148.000 263.532,157.135 259.286,157.567 258.936 C 157.983 258.597,156.856 258.906,155.400 259.530 M145.571 259.900 C 145.555 260.285,145.407 261.547,145.242 262.704 C 145.077 263.862,145.023 264.890,145.122 264.989 C 145.328 265.195,147.600 264.083,147.600 263.776 C 147.600 263.664,147.252 263.758,146.826 263.986 C 145.580 264.653,145.523 264.529,145.815 261.800 C 145.968 260.370,145.982 259.200,145.847 259.200 C 145.711 259.200,145.587 259.515,145.571 259.900 M72.000 273.800 C 71.450 274.114,71.180 274.372,71.400 274.372 C 71.620 274.372,72.250 274.114,72.800 273.800 C 73.350 273.486,73.620 273.228,73.400 273.228 C 73.180 273.228,72.550 273.486,72.000 273.800 M66.400 276.200 C 65.850 276.514,65.580 276.772,65.800 276.772 C 66.020 276.772,66.650 276.514,67.200 276.200 C 67.750 275.886,68.020 275.628,67.800 275.628 C 67.580 275.628,66.950 275.886,66.400 276.200 M48.800 283.800 C 48.250 284.114,47.980 284.372,48.200 284.372 C 48.420 284.372,49.050 284.114,49.600 283.800 C 50.150 283.486,50.420 283.228,50.200 283.228 C 49.980 283.228,49.350 283.486,48.800 283.800 M40.400 287.400 C 39.850 287.714,39.580 287.972,39.800 287.972 C 40.020 287.972,40.650 287.714,41.200 287.400 C 41.750 287.086,42.020 286.828,41.800 286.828 C 41.580 286.828,40.950 287.086,40.400 287.400 M34.800 289.800 C 34.250 290.114,33.980 290.372,34.200 290.372 C 34.420 290.372,35.050 290.114,35.600 289.800 C 36.150 289.486,36.420 289.228,36.200 289.228 C 35.980 289.228,35.350 289.486,34.800 289.800 M351.600 290.000 C 352.370 290.422,353.180 290.766,353.400 290.766 C 353.620 290.766,353.170 290.422,352.400 290.000 C 351.630 289.578,350.820 289.234,350.600 289.234 C 350.380 289.234,350.830 289.578,351.600 290.000 M29.200 292.200 C 28.650 292.514,28.380 292.772,28.600 292.772 C 28.820 292.772,29.450 292.514,30.000 292.200 C 30.550 291.886,30.820 291.628,30.600 291.628 C 30.380 291.628,29.750 291.886,29.200 292.200 M356.400 292.200 C 356.950 292.514,357.580 292.772,357.800 292.772 C 358.020 292.772,357.750 292.514,357.200 292.200 C 356.650 291.886,356.020 291.628,355.800 291.628 C 355.580 291.628,355.850 291.886,356.400 292.200 M363.600 295.400 C 364.150 295.714,364.780 295.972,365.000 295.972 C 365.220 295.972,364.950 295.714,364.400 295.400 C 363.850 295.086,363.220 294.828,363.000 294.828 C 362.780 294.828,363.050 295.086,363.600 295.400 M368.400 297.389 C 368.400 297.706,371.300 298.833,371.528 298.605 C 371.641 298.492,371.635 298.400,371.516 298.400 C 371.396 298.400,370.646 298.128,369.849 297.795 C 369.052 297.462,368.400 297.279,368.400 297.389 " stroke="none" fill="#7484a4" fill-rule="evenodd"></path></g></svg>

    </a>
    Genome-Wide Association Studies
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/hds-sandbox/GWAS_course" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    An introduction to GWAS
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" >
      
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_2">
          Course Notebooks
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Course Notebooks" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Course Notebooks
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS1-WhatIsAGWAS/" class="md-nav__link">
        1 What is a GWAS?
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS2-DataCollectionA/" class="md-nav__link">
        2 Data Collection
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS3-QualityControlA/" class="md-nav__link">
        3 Quality Control: Initial Steps
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS4-QualityControlB/" class="md-nav__link">
        4 Quality Control: Relatedness &amp; Population Stratification
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS5-AssociationTesting/" class="md-nav__link">
        5 Association Testing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS6-PRSAnalysis/" class="md-nav__link">
        6 PRS Analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Notebooks/GWAS7-OtherToolsAndTesting/" class="md-nav__link">
        7 Other Tools &amp; Conclusion
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
                
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#why-do-we-do-gwas-slides-2-6" class="md-nav__link">
    Why do we do GWAS? (slides 2-6)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ethical-aspects" class="md-nav__link">
    Ethical aspects
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contents-of-this-course-slides-24-25" class="md-nav__link">
    Contents of this course (slides 24-25)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#11-genetic-variation-slides-7-12" class="md-nav__link">
    1.1 Genetic variation (slides 7-12)
  </a>
  
    <nav class="md-nav" aria-label="1.1 Genetic variation (slides 7-12)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#111-genetic-variants" class="md-nav__link">
    1.1.1 Genetic variants
  </a>
  
    <nav class="md-nav" aria-label="1.1.1 Genetic variants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#some-catalogues-of-genetic-variation" class="md-nav__link">
    Some catalogues of genetic variation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#112-genotypes-and-hardy-weinberg-equilibrium" class="md-nav__link">
    1.1.2 Genotypes and Hardy-Weinberg equilibrium
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#12-what-is-a-genome-wide-association-study" class="md-nav__link">
    1.2 What is a genome-wide association study?
  </a>
  
    <nav class="md-nav" aria-label="1.2 What is a genome-wide association study?">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#121-quantitative-traits" class="md-nav__link">
    1.2.1 Quantitative traits
  </a>
  
    <nav class="md-nav" aria-label="1.2.1 Quantitative traits">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#additive-model" class="md-nav__link">
    Additive model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#full-model" class="md-nav__link">
    Full model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#current-approach-to-quantitative-phenotypes-in-gwas" class="md-nav__link">
    Current approach to quantitative phenotypes in GWAS
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#122-binary-phenotypes" class="md-nav__link">
    1.2.2 Binary phenotypes
  </a>
  
    <nav class="md-nav" aria-label="1.2.2 Binary phenotypes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#relative-risk-and-odds-ratio" class="md-nav__link">
    Relative risk and odds ratio
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#inference-for-or-based-on-counts" class="md-nav__link">
    Inference for OR based on counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#logistic-regression" class="md-nav__link">
    Logistic regression
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ascertained-case-control-studies" class="md-nav__link">
    Ascertained case-control studies
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gwas-software" class="md-nav__link">
    GWAS software
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                

  <a href="https://github.com/hds-sandbox/GWAS_course/edit/master/docs/WhatIsAGWAS.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>What is a GWAS?</h1>

<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>knitr::opts_chunk$set(echo = TRUE)
</code></pre></div>
<p>This document is licensed under a 
<a href="https://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.</p>
<p>The slide set referred to in this document is "GWAS 1".</p>
<p>This course is about <strong>statistical/computational ideas and methods</strong> that are used in 
<strong>genome-wide association studies (GWAS)</strong>. (Note that on these notes 'GWAS' is both singular and plural while other texts may use 'GWASs' or 'GWASes' for plural.) 
More generally, these same methods are useful
for analyzing large data sets of many other types, but, on this course, 
we will approach these from the angle of GWAS.</p>
<p>A GWAS quantifies statistical association between genetic variation
and phenotypes. A <strong>phenotype</strong>, also called a <strong>trait</strong>, 
can be any measured or observed property 
of an individual. Examples of phenotypes are quantitative traits like 
standing height or concentration of cholesterol particles in circulation, or
binary traits like diagnoses of multiple sclerosis or schizophrenia.</p>
<h4 id="why-do-we-do-gwas-slides-2-6">Why do we do GWAS? (slides 2-6)<a class="headerlink" href="#why-do-we-do-gwas-slides-2-6" title="Permanent link">&para;</a></h4>
<p>We do GWAS because a statistical association between 
a particular physical region of the 
genome and the phenotype</p>
<ul>
<li>can point to <strong>biological mechanisms</strong> affecting the phenotype,</li>
<li>can allow <strong>prediction</strong> of the phenotype from genomic information.</li>
</ul>
<p>These results may further benefit</p>
<ul>
<li><strong>medicine</strong> by leading to molecular or environmental interventions against harmful phenotypes,</li>
<li><strong>biotechnology</strong> by improving the ways we utilize microbes, plants or animals,</li>
<li><strong>forensics</strong> by more accurate identification of an individual from a DNA sample,</li>
<li>biogeographic <strong>ancestry inference</strong> of individuals, populations and species,</li>
<li>our understanding of the role of <strong>natural selection</strong> and other <strong>evolutionary forces</strong> in the living world.</li>
</ul>
<p>The genome of an individual remain (nearly) constant throughout the individual's
lifetime. This is a truly remarkable property compared to, e.g., other molecular
sources of information (such as metabolomics, metagenomics, transcriptomics, proteomics or epigenomics)
or environmental factors that may vary widely across time. 
Therefore, the genome seems an ideal starting point for scientific research: 
it needs to be measured only once for an individual and there is no 
<em>reverse causation</em> from the phenotype to genome 
(with cancer as an important exception).</p>
<h4 id="ethical-aspects">Ethical aspects<a class="headerlink" href="#ethical-aspects" title="Permanent link">&para;</a></h4>
<p>As with any powerful technique, the utilization of results from GWAS also raises
many new and difficult ethical questions 
and the legislation of utilization of genome information is 
under active development around the world.
For example, we need concrete answers to</p>
<ul>
<li>who can access an individual's genetic information
and for which purpose: individual himself/herself?, researchers?, medical professionals?, insurance companies?, employers?, 
school system?, everyone? </li>
<li>what kind of information should be returned back to an individual: 
genetic risk for a disease for which some preventive measures exist vs. a disease
with no actionable measures known?, 
genetic prediction of sensitive traits such as IQ?, 
genetic ancestry or family information
that does not match prior expectations of the individual, for example, 
due to false paternity? </li>
<li>what kind of information, if any, should be returned to a relative of
a tested individual given that the relative may also have some of the same genetic variants?</li>
<li>when is gene editing allowed: to cure severe disease?, to prevent a severe mutation
to be passed on to next generation?, to design the next generation to have some favorable genetic variants?</li>
</ul>
<p>When working with human genome data, we should always keep it clear in mind
that there are such profound questions related to these data,
and that the data we handle will likely 
turn out to be more powerful than most of us can imagine today.
Human genome data are never 'just data' but include highly personal information,
and they need to be handled with high respect and care.
Access to genetic data requires a written agreement
between the researcher and the data provider about how and for which purpose
the data can be used.</p>
<h4 id="contents-of-this-course-slides-24-25">Contents of this course (slides 24-25)<a class="headerlink" href="#contents-of-this-course-slides-24-25" title="Permanent link">&para;</a></h4>
<p>The plan is to discuss the following topics (in varying level of detail):</p>
<ol>
<li>
<p>What is a GWAS?</p>
</li>
<li>
<p>Statistics of GWAS (regression coefficients, P-values, statistical power, Bayes factors)</p>
</li>
<li>
<p>Genetic relatedness and population structure </p>
</li>
<li>
<p>Confounding and covariates in GWAS</p>
</li>
<li>
<p>Haplotypes, linkage disequilibrum, imputation, fine-mapping</p>
</li>
<li>
<p>Linear mixed models and heritability</p>
</li>
<li>
<p>Summary statistics and meta-analysis</p>
</li>
<li>
<p>Mendelian randomization</p>
</li>
<li>
<p>Success and critisism of GWAS</p>
</li>
<li>
<p>Human genetics research at <a href="https://www.fimm.fi/en/research/human-genomics-and-preventive-health-care">FIMM</a></p>
</li>
</ol>
<h2 id="11-genetic-variation-slides-7-12">1.1 Genetic variation (slides 7-12)<a class="headerlink" href="#11-genetic-variation-slides-7-12" title="Permanent link">&para;</a></h2>
<p>We all carry two nuclear genomes (i.e. genomes located in cell nucleus), 
one inherited from each of our two parents.
Additionally, we have a small mitochondrial genome, 
<a href="http://www.pnas.org/content/early/2018/11/21/1810946115">assumed to be</a> 
inherited exclusively 
from the mother, but on this course the term 'genome' refers to the nuclear genome.</p>
<p>Human genome is 3.2 billion <strong>nucleotides</strong> (or <strong>base pairs</strong> or <strong>DNA letters</strong> A,C,G,T) 
long sequence (<a href="https://www.yourgenome.org/facts/what-is-dna">see yourgenome.org</a>), 
that is divided into separate physical pieces called <strong>chromosomes</strong> (<a href="https://www.yourgenome.org/facts/what-is-a-chromosome">see yourgenome.org</a>).
There are 22 <strong>autosomal</strong> (non-sex related) chromosomes
and two <strong>sex chromosomes</strong> (X chromosome and Y chromosome). 
Normally, humans have two copies of each
autosome and individuals with one copy of X and one of Y are males 
whereas individuals who have two copies of X are females.
Abnormal number of chromosomes (called <strong>aneuploidies</strong>) typically 
cause severe consequences
or an early death if present in all cells of an individual.
The most common non-lethal exception is the Down syndrome (3 copies of chr 21). 
<strong>Mosaicism</strong>, where some cells have abnormal chromosome numbers also exist
and are often present in cancer cells. </p>
<p>There are three types of pairings that come up when we analyse genomes. </p>
<ul>
<li>
<p>First, the DNA is most of the time
a double-stranded molecule whose two strands (i.e. the two DNA molecules) are glued together 
by the chemical base pairings A-T and C-G.
This base pairing is a key to the copying mechanism of the DNA that is needed 
before any cell division (<a href="https://www.yourgenome.org/video/dna-replication">see yourgenome.org</a>) and 
the DNA molecules that are connected through base pairing 
have the exact same information, just written in the reverse letters 
i.e. A &lt;-&gt; T and C &lt;-&gt; G. 
To make distinction between the two DNA molecules,
it has been agreed that one of the two DNA strands is called 
the <strong>forward strand</strong> (or <strong>positive strand</strong>)
and the other the <strong>reverse strand</strong> (or <strong>negative strand</strong>).
Thus, e.g., when + strand contains base A, the corresponding base on - strand is T and vice versa. </p>
</li>
<li>
<p>Second, the two <strong>homologous</strong> choromosomes of an individual
(e.g. paternal chr 13 with maternal chr 13, or in a male, maternal X and paternal Y) 
can be thought of as a pair. 
Thus, we say that the human genome consists 
of 22 autosomes + X + Y, but each individual
has two copies of each homologous chromosome, 
so has 46 unique chromosomes that are divided into 23 pairs of homologous chromosomes.</p>
</li>
<li>
<p>Third, before any <strong>cell division</strong> each of the 46 unique chromosomes of an individual
copies itself and the two copies (called sister chromatids) 
are paired with each other physically to make
an X-like shape that is often used to show chromosomes in pictures.
Such picture actually has 92 chromosomes since each unique chromosome 
is duplicated in it (but we typically say that there are 46 replicated chromosomes rather than 
that there are 92 chromosomes).
This pairing after copying is important in cell division so that the resulting cells 
will get the correct set of choromosomes. In mitosis (ordinary cell division),
each of the two new cells has one set of the 46 unique chromosomes.
In meiosis, the gametes (sperm and eggs) are formed to have 
only one copy of each homologous chromosome and thus have 23 unique chromosomes. 
During meiosis, the process of <strong>recombination</strong> shuffles the homologous
copies of the paternal and maternal chromosomes in such a way that each of the 
offspring's chromosomes will be a mixture of its grandparental chromosome segments.</p>
</li>
</ul>
<p>Terms</p>
<ul>
<li>
<p><strong>Gene.</strong>
The most obvious way how genetic variation can affect phenotypes is through
variation in how <strong>genes</strong> function. Genes are segements of DNA that code for proteins
(<a href="https://www.yourgenome.org/video/from-dna-to-protein">see yourgenome.org</a>) and variation
in the physical structure of the protein or in the time and place where the protein is made
can have phenotypic consequences. Therefore, we are very 
interested in how genetic variation can affect the function of genes,
and a lot of this is still unknown.
Protein coding genes cover less than 2% of the whole human genome,
but the remaining 98% affects the regulation of genes in many ways.</p>
</li>
<li>
<p><strong>Locus (pl. loci).</strong>
A continuous region of the genome is called a locus (plural loci).
It can be of any size (e.g. a single nucleotide site of length 1 bp or
a region of 10 milion base pairs, 10 Mbp).
<strong>GWAS loci</strong> are regions that include a clear statistical association
with the phenotype of interest.</p>
</li>
</ul>
<h3 id="111-genetic-variants">1.1.1 Genetic variants<a class="headerlink" href="#111-genetic-variants" title="Permanent link">&para;</a></h3>
<p>At any one position of the genome (e.g. nucleotide site at position 13,475,383 of chromosome 1, 
denoted by chr1:13,475,383) 
variation can exist between the genomes in the population. For example,
my paternal chromosome can have a base A and maternal chromosome can have a base G 
(on the +strand of the DNA) at that position. (slide 10)
Such a one-nucleotide variation is called a <strong>single-nucleotide variant (SNV)</strong> and
the two versions are called <strong>alleles</strong>. So in the example case, 
I would be carrying both an allele A and an allele G at that SNV, whereas 
you might be carrying two copies of allele A at the same SNV.
My <strong>genotype</strong> would be AG and yours AA.
An individual having different alleles on his/her two genomes is <strong>heterozygous</strong>
at that locus, and an individual having two copies of the same allele is 
<strong>homozygous</strong> at that locus.
If neither of the alleles
is very rare in the population, say, the <strong>minor allele frequency (MAF)</strong> 
is &gt; 1% in the population, 
the variant is called a polymorphism, <strong>single-nucleotide polymorphism (SNP)</strong>.
There are over 10 million SNPs in the human genome.
More complex genetic variation (slide 9) include structural variation (SV) such as 
copy number variants (CNVs), that include duplications or deletions of genomic regions,
or rearrangements of the genome, such as inversions or translocations of DNA segments
(<a href="https://www.yourgenome.org/facts/what-types-of-mutation-are-there">see yourgenome.org</a>).</p>
<p>A predefined set of 500,000 - 1,000,000 
SNPs can be measured reliably and fairly cheaply (&lt; 50 euros/sample) by DNA microarrays,
which has been the single most important factor making GWAS possible (slide 11; <a href="https://www.mun.ca/biology/scarr/DNA_Chips.html">Illustration</a>).
On this course, we consider SNPs as the canonical type of genetic variation.
Typically, the SNPs are <strong>biallelic</strong>, i.e., there are only two alleles present in the population
and this is what we assume in the following. 
In principle, however, all four possible alleles of a SNP 
could be present in the population.</p>
<p><strong>Ambiguous SNPs.</strong> If the two alleles of a SNP are either C,G or A,T we call the SNP
<em>ambiguous</em> because the strand information must be available (and correct)
in order to make sense of the genotypes at this SNP.
This is because allele C on + strand would be called allele G on - strand 
and if this SNP is reported with respect to different strands in different
studies, the results get mixed up. 
The same problem does not happen with the other SNPs, e.g., 
a SNP with alleles A,C, because 
this SNP contains alleles T,G on the opposite strand
and we could unambiguously match A to T and C to G between the studies.
Note that we can resolve most ambiguous SNPs reliably based on the allele frequencies
as long as the minor allele frequency is not close to 50%.
If we are combining several studies, we should always start by 
plotting the allele frequencies between the studies
after the alleles should be matching each other in order to see that the frequencies
indeed match across the studies.</p>
<h4 id="some-catalogues-of-genetic-variation">Some catalogues of genetic variation<a class="headerlink" href="#some-catalogues-of-genetic-variation" title="Permanent link">&para;</a></h4>
<p>A large part of the genetics research over the last 30 years have been driven by
international projects aiming to catalogue genetic variation in public domain.</p>
<ul>
<li>
<p><a href="https://www.yourgenome.org/stories/what-was-the-human-genome-project-for">The Human Genome Project</a> 
1990-2003 established a first draft of a human genome sequence. </p>
</li>
<li>
<p><a href="https://www.genome.gov/10001688/international-hapmap-project/">The HapMap project</a> 
2002-2009 studied the correlation structure of the common SNPs. </p>
</li>
<li>
<p><a href="http://www.internationalgenome.org/about">The 1000 Genomes project</a> 2008-2015,
expanded HapMap to genome sequence information across the globe and
currently remains a widely-used reference for global allele frequency information.
1000G project was able to characterize well common variation in different <em>populations</em>, but 
missed many rare variants of single individuals 
because the costs of very accurate sequencing were too high.
The tremendous impact of the 1000G project stems from the fact that everyone can download
the individual level genome data of the 1000G samples from the project's website
and use it in their own research.</p>
</li>
<li>
<p><a href="http://exac.broadinstitute.org/">Exome Aggregation Consortium (ExAC)</a> 2014-2016 
concentrated only on the protein coding
parts of the genome, so called <em>exons</em>, that make up less than 2% of the genome
and was able to provide accurate sequence data for the exomes of over 60,000 individuals.
This effort has been particularly important for medical interpretation of rare
variants seen in clinics that diagnose patients with severe disease.
ExAC provides summary level information through browser and downloads
but individual level data cannot be downloaded.</p>
</li>
<li>
<p><a href="http://gnomad.broadinstitute.org/">Genome Aggregation Database (gnomAD)</a>
is expanding the ExAC database and 
also includes additional whole genome sequencing information. 
It is the current state-of-the-art among the public genome variation databases.</p>
</li>
</ul>
<h3 id="112-genotypes-and-hardy-weinberg-equilibrium">1.1.2 Genotypes and Hardy-Weinberg equilibrium<a class="headerlink" href="#112-genotypes-and-hardy-weinberg-equilibrium" title="Permanent link">&para;</a></h3>
<p>Let's consider one SNP in the population. The SNP has two alleles that could be called
by their nucleotides, but with quantitative analyses in mind, 
we name the alleles in such a way that the <strong>minor allele</strong>
(the one that is less common in the population)
is called allele 1 and the <strong>major allele</strong> (the one that is more common in the population)
is called allele 0. 
(There is no general rule that GWAS results are reported as 
allele 1 corresponding to the minor allele, and even if there was, 
the minor allele could differ between two data sets/populations, so consistency across studies
needs always to be checked.) 
Let's denote the <strong>minor allele frequency (MAF)</strong> by <span class="arithmatex">\(f\)</span>.
Since each individual has two copies of the genome, there are individuals with 
three possible genetic types (called <strong>genotypes</strong>) at this SNP.
We denote each genotype by the number of copies of allele 1 that 
the genotype contains (i.e. genotype can be 0, 1 or 2).
If we assume that, at this SNP, the two alleles in one individual are sampled at random
from the population, then the genotype frequencies in the population follow
the binomial distribution <span class="arithmatex">\(\textrm{Bin}(2,f)\)</span>:</p>
<table>
<thead>
<tr>
<th>genotype</th>
<th>expected frequency from Bin(2,f)</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td><span class="arithmatex">\((1-f)^2\)</span></td>
</tr>
<tr>
<td>1</td>
<td><span class="arithmatex">\(2\,f(1-f)\)</span></td>
</tr>
<tr>
<td>2</td>
<td><span class="arithmatex">\(f^2\)</span></td>
</tr>
</tbody>
</table>
<p>These frequencies are called the <strong>Hardy-Weinberg equilibrium</strong> (HWE) genotype frequencies 
(or HW proportions) and 
they define the theoretical equilibrium genotype frequencies given the value of <span class="arithmatex">\(f\)</span> 
in an ideal <em>randomly mating</em> population
without <em>selection</em>, <em>migration</em>, or <em>genetic drift</em> 
(=statistical fluctuations due to finite population size).
In practice, most variants in human populations do approximately follow HWE.
Clear deviations from HWE could point to, for example, 
recent <em>population structure</em> (e.g. two populations have admixed), 
<em>assortative mating</em> (individuals tend to mate with partners of their kind)
or natural selection (e.g. genotype 1 is very advantageous whereas 
genotype 2 is lethal and hence completely absent from the population).
On the other hand, technical problems in <em>genotype calling</em> (i.e. in determining genotypes
from the intensity measures from a genotyping chip) can also cause deviations
from HWE, either because of bad quality data
or because the variation is not a biallelic SNP and has more than two alleles
(slide 12). Therefore,
often variants which do not follow HW frequencies are excluded from many GWAS analyses
as part of the <a href="https://www.nature.com/articles/nprot.2010.116"><em>quality control</em> (QC) procedure</a>.</p>
<p><strong>Testing HWE.</strong> To test for (deviations from) HWE a 1-df <a href="https://en.wikipedia.org/wiki/Hardy%E2%80%93Weinberg_principle#Significance_tests_for_deviation">chi-square test</a> can be used
where the expected counts are derived under HWE given the allele frequencies. </p>
<p>Suppose that we have observed genotype counts <span class="arithmatex">\(n_0,n_1,n_2\)</span> for genotypes
0,1 and 2, respectively, and <span class="arithmatex">\(N = n_0 + n_1 + n_2.\)</span> Our estimate for
population frequency of allele 1 is <span class="arithmatex">\(\widehat{f} = (n_1+2\,n_2)/(2\,N)\)</span>,
and the expected genotype counts under HWE are <span class="arithmatex">\(h_0 = N(1-\widehat{f})^2\)</span>,
<span class="arithmatex">\(h_1 = 2 N \widehat{f}(1-\widehat{f})\)</span> and <span class="arithmatex">\(h_2 = N \widehat{f}^2\)</span>.
The test statistic measures the deviation between the observed counts and the
expected counts:
$$ t_{HWE} = \sum_{i=0}^2 \frac{(n_i-h_i)^2}{h_i}.$$
If HWE holds, then <span class="arithmatex">\(t_{HWE}\)</span> follows approximately a
Chi-square distribution with 1 degree of freedom,
which is used for deriving a P-value.
If theoretical chi-square distribution is used, 
the test is asymptotic and hence not necessarily valid for small sample size 
or very rare variants, and a test statistic distribution 
based on permutations should be used, e.g., by R 
package <a href="https://cran.r-project.org/web/packages/HardyWeinberg/index.html">HardyWeinberg</a>.</p>
<p><strong>Example 1.1.</strong>
Look up the SNP rs429358 from the Ensmbl browser <a href="https://www.ensembl.org/">https://www.ensembl.org/</a>.
Choose Human and type 'rs429358'; you'll see the variant's chr (19), position 
(44,908,684 in the version GRCh38 of the Human Genome coordinates mentioned on top left) 
and the two alleles, T and C, of which C is predicted to be the 'ancestral' that is, the 
older allele, and C is also the minor allele with average MAF of 15% across human populations.
Next click 'Population Genetics' to see allele and genotype frequencies in different human populations. (Familiarize with given populations by hovering mouse above them.)
Scrolling down, in the 1000 Genomes project Phase 3 (1000G) Finnish data, 
the minor allele C has frequency <span class="arithmatex">\(37/198 \approx 18.7\%\)</span> and 
the observed genotype counts are 66 (TT), 29 (TC) and 4 (CC) individuals. 
Let's use these values to visualize the genotype distribution
and apply the standard test for HWE.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>geno = c(66,29,4)
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>n = sum(geno) #number of individuals
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>f = sum(geno*c(0,1,2))/(2*n) #(66*0 + 29*1 + 4*2) / (2*(66+29+4))
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>f #MAF
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>hwe.prop = c((1-f)^2, 2*f*(1-f), f^2) #these would be the genotype freqs under HWE
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>rbind(obs = geno/n, hwe = hwe.prop) #print the observed genotype freqs and the HWE.
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>#For testing HWE we use chi-square test even though counts are quite small in last cell:
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>hwe.test = sum((geno - n*hwe.prop)^2/(n*hwe.prop)) #HWE test statistic
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>hwe.p = pchisq(hwe.test, df = 1, lower = FALSE) #P-value from the test
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>barplot(geno, main = paste(&quot;rs429358 FIN in 1000G Phase3; HWE P=&quot;,signif(hwe.p,3)),
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>        names = c(0,1,2), xlab = &quot;genotype&quot;, col = &quot;skyblue&quot;)
</code></pre></div>
<p>To get familiar with how to generate realistic genotype data,
let's also make example 
genotype data for <span class="arithmatex">\(n = 1000\)</span> additional 
Finns at this variant, both by sampling from the genotype frequencies
and by sampling from the allele frequencies (assuming HWE).
Since this variant seems to follow HWE, we do not expect qualitative differences
between the two simulation approaches.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>set.seed(19) #setting seed guarantees the same simulation results every time this code is run
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>n = 1000
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>sample.from.geno = sample(c(0,1,2), prob = geno, size = n, replace = T) #sample from genotype frequencies
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a># replace = TRUE means sampling with replacement, that is, 
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a># each genotype can be sampled many times, always with the same probabilities given in &#39;prob&#39;
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>tab = table(sample.from.geno) #table counts how many times each value is present
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>counts.from.geno = rep(0,3) #How many carriers of each genotype
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>counts.from.geno[ 1 + as.numeric( names(tab) )] = as.numeric(tab) #works even if some count is 0
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>#To sample from HWE frequencies, we could use:
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>#sample.from.hwe = sample(c(0,1,2), prob = c((1-f)^2, 2*f*(1-f), f^2), size = n, replace = T)
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>#but a simpler way is to sample n genotypes directly from Bin(2,f) distribution:
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>sample.from.hwe = rbinom(n, size = 2, p = f)
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>counts.from.hwe = rep(0,3) #Let&#39;s count how many carriers of each genotype
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>for(ii in 0:2){ #this is another way to do the counting compared to table() above
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>  counts.from.hwe[ii+1] = sum(sample.from.hwe == ii)}
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a>rbind(geno = counts.from.geno/n, hwe = counts.from.hwe/n)
<a id="__codelineno-2-19" name="__codelineno-2-19" href="#__codelineno-2-19"></a>
<a id="__codelineno-2-20" name="__codelineno-2-20" href="#__codelineno-2-20"></a>barplot(cbind(counts.from.geno/n, counts.from.hwe/n), 
<a id="__codelineno-2-21" name="__codelineno-2-21" href="#__codelineno-2-21"></a>        names = c(&quot;geno&quot;,&quot;HWE&quot;), beside = F, horiz = T)
</code></pre></div>
<p>They <em>look</em> pretty similar but 
to do statistical inference we should also quantify the
uncertainty of the estimates (e.g. by 95% intervals). 
For small counts, a Bayesian credible interval  called 
<em>Jeffreys interval</em> is preferred
over the standard 95% confidence interval, 
whereas for larger counts the two approaches agree.
Details of the two approaches are <a href="https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval">here</a>.</p>
<p>Let's make Jeffreys intervals for the estimates of each genotype frequency
in both of the data sets.
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>interval.from.geno = matrix(NA, ncol = 2, nrow = 3) #empty matrix
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>interval.from.hwe = matrix(NA, ncol = 2, nrow = 3)
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>for(ii in 1:3){ #find intervals while looping over 3 genotypes
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>  interval.from.geno[ii,] = qbeta(c(0.025,0.975), counts.from.geno[ii]+0.5, n-counts.from.geno[ii]+0.5)
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>  interval.from.hwe[ii,] = qbeta(c(0.025,0.975), counts.from.hwe[ii]+0.5, n-counts.from.hwe[ii]+0.5)
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>}
</code></pre></div></p>
<p>Now we can print out the observed genotype frequency (1<sup>st</sup> col) and its 95% interval (2<sup>nd</sup> and 3<sup>rd</sup> cols)
for both data sets and compare whether the estimates seem similar
given the uncertainty described by the intervals:
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>cbind(geno.est = counts.from.geno/n,interval.from.geno,
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>      hwe.est = counts.from.hwe/n,interval.from.hwe ) 
</code></pre></div>
All estimates are within other data set's 95% Jeffreys credible interval and 
we have no reason to suspect frequency differences between the two data sets.</p>
<p>A standard two-sample chi-square test can also be carried out to 
quantify the frequency difference as a P-value:
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>chisq.test(rbind(counts.from.geno,counts.from.hwe)) #tests whether rows have same distribution
</code></pre></div>
Unsurprisingly, 
this does not indicate any frequency difference between the two data sets
as P-value is large (0.5361).
We'll talk more about interpretation of P-values later.</p>
<h2 id="12-what-is-a-genome-wide-association-study">1.2 What is a genome-wide association study?<a class="headerlink" href="#12-what-is-a-genome-wide-association-study" title="Permanent link">&para;</a></h2>
<p>Let's look at some recent examples of GWAS.
Two main types of GWAS are studying quantitative traits or disease phenotypes.</p>
<p><strong>Example 1.2. QT-GWAS (slides 14-18)</strong> <a href="https://www.nature.com/articles/nature14177">GWAS on body-mass index (BMI)</a> by Locke et al. (2015)
combined data of 339,000 individuals from 125 studies around
the world to study the association of SNPs and BMI. 
It highlighted 97 regions of the genome
with convincing statistical association with BMI.
Pathway analyses provided support for a role of the central nervous system in
obesity susceptibility and implicated new genes and pathways related to 
synaptic function, glutamate signalling, insulin secretion/action, 
energy metabolism, lipid biology and adipogenesis.</p>
<p><strong>Example 1.3. Disease GWAS (slides 20-22)</strong> <a href="https://www.nature.com/articles/ng.3598">GWAS on migraine</a> by Gormley et al. (2016) 
combined genetic data on 60,000 cases (individuals with migraine)
and 315,000 controls (individuals with no known migraine) originating from
22 studies. Genetic data was available on millions of
genetic variants. At each variant, the genotype distribution between cases and controls were
compared. 38 regions of the genome showed a convincing statistical association with migraine.
Downstream analyses combined the genes into pathways and cell types
and highlighted enrichment of signals near genes that are active in vascular system. </p>
<p>Terms:</p>
<ul>
<li><strong>Monogenic</strong> phenotype is determined by a single gene/locus.</li>
<li><strong>Oligogenic</strong> phenotype is influenced by a handful of genes/loci.</li>
<li><strong>Polygenic</strong> phenotype is influenced by many genes/loci.</li>
<li><strong>Complex trait</strong> is a (quantitative) phenotype that is not monogenic. Typically polygenic and also influenced by many environmental factors.</li>
<li><strong>Common disease</strong> is a disease/condition that is common in the population (say, prevalence of 0.1% or more). Examples:  MS-disease (prevalence in the order of 0.1%), schizophrenia (~1%) or Type 2 diabetes (~10%).</li>
<li><strong>Common variant</strong> has frequency of at least 1% (also 5% is used as the threshold).</li>
<li><strong>Low-frequency variant</strong> has frequency of at least 0.1% and lower than a common variant.</li>
<li><strong>Rare variant</strong> has frequency lower than a low-frequency variant.</li>
</ul>
<p>GWAS have shown us that, very generally, 
complex traits and common diseases are highly polygenic,
and many common variants with only small effects influence these phenotypes.
We don't yet know which are the exact causal variants for each phenotype because
of the correlation structure among genetic variants 
(this is the fine-mapping problem we'll look later). 
We also don't yet know very accurately
how rare variants affect each phenotype because that requires very large 
sample sizes interrogated by <em>genome sequencing</em> techniques, not only by SNP arrays.</p>
<h3 id="121-quantitative-traits">1.2.1 Quantitative traits<a class="headerlink" href="#121-quantitative-traits" title="Permanent link">&para;</a></h3>
<p>Let's mimick the data we see on slide 3.
The phenotype is LDL-cholesterol level and we assume that the trait distributions
of individuals with 0, 1 or 2 copies of allele T at SNP rs11591147
are Normal distributions with SD=1 and with means of 
0.02, -0.40 and -2.00, respectively.
Allele T frequency is 4% in Finland. 
Let's simulate <span class="arithmatex">\(n=10,000\)</span> individuals and boxplot them by genotype.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>n = 10000
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>f = 0.04
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>mu = c(0.02, -0.40, -2.00) #mean of each genotype
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>sigma = c(1, 1, 1) #SD of each genotype
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>x = rbinom(n, size = 2, p = f) #genotypes for &#39;n&#39; individuals assuming HWE
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>table(x)/n #(always check that simulated data is ok before starting to work with it!)
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>y = rep(NA,n) #make empty phenotype vector
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>for(ii in 0:2){ #go through each genotype group: 0, 1, 2.
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>  y[x == ii] = rnorm(sum(x == ii), mu[1+ii], sigma[1+ii]) } #generate trait for group ii
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>boxplot(y ~ x, main = &quot;Simulated rs11591147 in Finns&quot;, ylab = &quot;LDL&quot;, 
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>        xlab = &quot;Copies of T&quot;, col = &quot;limegreen&quot;)
</code></pre></div>
<p>We see that the phenotype varies with genotype in such a way that each additional 
copy of allele T decreases the level of LDL.</p>
<h5 id="additive-model">Additive model<a class="headerlink" href="#additive-model" title="Permanent link">&para;</a></h5>
<p>The simplest way to analyze these data statistically is to 
use an <strong>additive model</strong>, that makes the assumption that the means of the groups
depend additively on the number of allele 1 in the genotype, and 
that the SDs of the genotype groups are constant. 
Thus, we fit a linear model <span class="arithmatex">\(y = \mu + x \beta + \varepsilon\)</span>, where
<span class="arithmatex">\(y\)</span> is the phenotype, <span class="arithmatex">\(x\)</span> is the genotype (0,1 or 2) and parameters to be estimated are </p>
<ul>
<li><span class="arithmatex">\(\mu\)</span>, the mean of genotype 0 and </li>
<li><span class="arithmatex">\(\beta\)</span>, the effect of each copy of allele 1 on the mean phenotype.</li>
</ul>
<p>The error terms <span class="arithmatex">\(\varepsilon\)</span>
are assumed to have an identical Normal distribution <span class="arithmatex">\(N(0,\sigma^2)\)</span> where
<span class="arithmatex">\(\sigma^2\)</span> is not known and will be estimated from the data. 
Let's fit this linear model in R using <code>lm()</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>lm.fit = lm(y ~ x)
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>summary(lm.fit)
</code></pre></div>
<p>The <code>summary(lm.fit)</code> command produced </p>
<ul>
<li>parameter estimates (or Coefficients) <span class="arithmatex">\(\widehat{\mu}\)</span> and <span class="arithmatex">\(\widehat{\beta}\)</span>, </li>
<li>their standard errors (SE) (estimates for square root of the sampling variance of the parameter estimates), </li>
<li>t-statistic (estimate/SE) and </li>
<li>P-value under the null hypothesis that the parameter is 0 and errors are uncorrelated and have  distribution <span class="arithmatex">\(N(0,\sigma^2)\)</span>. </li>
</ul>
<p>Under the assumptions of linear model, 
sampling distribution of t-statistic is <span class="arithmatex">\(t\)</span>-distribution and hence q% confidence 
intervals are determined as <span class="arithmatex">\(\widehat{\beta}\pm a\times \textrm{SE}\)</span>, 
where <span class="arithmatex">\(a\)</span> is the q/2% quantile of <span class="arithmatex">\(t\)</span>-distribution with <span class="arithmatex">\(n-2\)</span> degrees of freedom. 
When <span class="arithmatex">\(\sigma^2\)</span> is known, the <span class="arithmatex">\(t\)</span>-distribution is replaced by a Gaussian, 
and same is approximately true when <span class="arithmatex">\(n\)</span> becomes large, 
even if estimate <span class="arithmatex">\(\widehat{\sigma^2}\)</span> is used in computing SE. 
In these cases, we often talk about z-statistic instead of t-statistic. 
In GWAS analyses, we typically have thousands of samples 
and use z-scores and the Normal approximation by default.</p>
<p>The last paragraph in the output tells about the full model fit. 
We can measure how much variation in <span class="arithmatex">\(y\)</span> is left unexplained by the model 
by computing <strong>residual sum of squares</strong> (RSS):
$$
RSS = \sum_{i=1}^n \left(y_i - \widehat{\mu} - x_{i} \widehat{\beta}\right)^2
$$
<span class="arithmatex">\(R^2\)</span> is the proportion of variance explained by the linear model, 
i.e., <span class="arithmatex">\(R^2 = 1-\frac{RSS}{n-1} / \widehat{\textrm{Var}}(y)\)</span>.
Adjusted version penalizes for additional predictors and is defined here
as <span class="arithmatex">\(R_{adj}^2 = 1-\frac{RSS}{n-2} / \widehat{\textrm{Var}}(y)\)</span>. 
Note that if there is only the
intercept parameter <span class="arithmatex">\(\mu\)</span> in the model, 
then <span class="arithmatex">\(R^2=R_{adj}^2=0\)</span>, and if the model explains data perfectly (<span class="arithmatex">\(RSS=0\)</span>), 
then <span class="arithmatex">\(R^2=R_{adj}^2=1\)</span>. 
In other cases <span class="arithmatex">\(R^2\)</span> values are between 0 and 1 
and larger values mean more variance explained by the model.</p>
<p>However, <span class="arithmatex">\(R^2\)</span> should not be the only value used to judge how suitable 
the model is for the data. One should also plot the data and the model 
fit in different ways to assess this question. (Although, of course not
for all variants in GWAS, but for the most interesting ones.) 
For this simple linear model, a scatter plot and a regression line is a good way to
assess whether there seem to be deviations from the additivity assumption.
Additionally, the differences in residual variation between the groups 
could indicate interaction effecs between the genetic variant and 
some other genetic or environmental variable.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>plot( x + runif(n, -0.05, 0.05), y, xlab = &quot;genotype&quot;, ylab = &quot;LDL&quot;, xaxt = &quot;n&quot;, 
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>      pch = 3, cex = 0.5, col = &quot;gray&quot;) 
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>#runif() adds some jitter to x so that all points are not on top of each other
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>axis(1, at = 0:2, labels = 0:2)
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>points(0:2, c(mean(y[x==0]), mean(y[x==1]), mean(y[x==2])), col = &quot;red&quot;, pch = &quot;X&quot;, cex = 1.3)
<a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>abline(lm.fit, col = &quot;orange&quot;, lwd = 2)
<a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a>legend(&quot;topright&quot;, pch = &quot;X&quot;, legend =&quot;group means&quot;, col = &quot;red&quot;)
</code></pre></div>
<p><strong>Conclusion</strong>: We see a statistically highly significant association between the 
genotype and phenotype where a copy of allele T decreases LDL levels by 0.45 units. 
This variant explains about
1.5% of the variation in LDL-cholesterol levels. We also see that individuals homozygous
for allele T (genotype 2) have on average lower levels of LDL than the model predicts, 
which indicates a deviation from the additivity assumption.
Let's next fit a full 2-parameter model to quantify this deviation.</p>
<h4 id="full-model">Full model<a class="headerlink" href="#full-model" title="Permanent link">&para;</a></h4>
<p>Let's add a new parameter <span class="arithmatex">\(\gamma\)</span> to the model that describes the residual effect
for group 2 after the additive effect is accounted for. 
The model is <span class="arithmatex">\(y = \mu + x\beta + z\gamma + \varepsilon,\)</span>
where <span class="arithmatex">\(z\)</span> is indicator of genotype 2, i.e., <span class="arithmatex">\(z_i=1\)</span> if individual <span class="arithmatex">\(i\)</span> has genotype 2
and otherwise <span class="arithmatex">\(z_i=0\)</span>.
This is the full model
where each genotype group has its own mean (genotype 0: <span class="arithmatex">\(\mu\)</span>; 
genotype 1: <span class="arithmatex">\(\mu + \beta\)</span>
and genotype 2: <span class="arithmatex">\(\mu + 2\beta + \gamma\)</span>).</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>z = as.numeric( x == 2 ) #z is indicator for genotype group 2
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>lm.full = lm( y ~ x + z )
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>summary(lm.full)
</code></pre></div>
<p>It seems that also the new variable is useful (large effect and small P-value).
Now the interpretation of coefficients is that genotype 1 has avg. phenotype
of -0.40 and genotype 2 has avg. phenotype -0.398*2-1.206 = -2.00.</p>
<p>Note also that the full model above gives the same model fit and is simply 
a different parameterization of
the linear regression that treats the genotype as a factor with three levels. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>lm.full2 = lm( y ~ as.factor(x) )
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>summary(lm.full2)
</code></pre></div>
<p>We should use this latter parameterisation of the full model if we are interested
in SE of the genetic effect of genotype 2, because that information is not
given by the first parameterisation of the full model. If we instead are
interested to quantify how much the data show deviation from the additivity
then the first parameterisation is the most suitable.</p>
<p>(The latter parameterisation is also the same model as the 
traditional Analysis of Variance ANOVA but 
we are working within the regression model framework for reasons that become clear
once we get to confounders and covariates.)</p>
<p>All these models make the same assumption that SD is constant across the 
genotype groups. </p>
<h4 id="current-approach-to-quantitative-phenotypes-in-gwas">Current approach to quantitative phenotypes in GWAS<a class="headerlink" href="#current-approach-to-quantitative-phenotypes-in-gwas" title="Permanent link">&para;</a></h4>
<p>Almost all GWAS analyze quantitative traits using the additive model, i.e.,
a linear regression model with a single parameter for genetic effect.
The full model is typically only used for a small group of interesting variants
identified by the additive model
to check them manually for possible deviations from the additive effects.
The main reason for this is that the additive model is usually almost as
powerful to find associations as the full model even when deviations from additivity are present
in the data, since typically one of the genotype groups is much smaller than the other two and
hence does not affect much the statistical model fit. 
Additionally, our current understanding is
that most associations follow well the additive model and
the additive model has more power than the full model, for these cases.
(But note that our current understanding may be biased in favor of
the additive model since we do not usually look very carefully
for non-additive effects.)
It would seem useful to run both the additive and full model in GWAS, 
but this is often not
done because there are millions of variants to be analyzed 
and hence there are a lot of work and output 
already when only the simplest model is applied.</p>
<p><strong>Quantile normalisation (QN).</strong> 
Often in large GWAS the quantitative phenotype is forced to follow a
Normal distribution by quantile normalisation (also <em>inverse-Normal transfromation</em>).
This way the effect of phenotypic outliers is diminished while keeping the
ranking of the trait values constant. 
This also harmonizes the trait distributions across multiple cohorts by
forcing them to look similar.</p>
<p>To apply QN to a set of <span class="arithmatex">\(n\)</span> trait values, 
we first regress out from the trait values the central covariates 
(such as sex and age) using a linear model. 
Then we order the residuals of the regression in ascending order 
<span class="arithmatex">\(\widehat{r}_{(1)}\leq \ldots \leq \widehat{r}_{(n)}\)</span>.
Now the QN'ed trait values for the sample is taken from
the inverse of the cumulative distribution function of the Normal distribution
at <span class="arithmatex">\(n\)</span> equally spaced values between 0 and 1, and the resulting values
<span class="arithmatex">\(q_1 \leq \ldots \leq q_n\)</span> are matched to the individuals so that the value
<span class="arithmatex">\(q_i\)</span> becomes the trait value for individual who corresponds to residual
<span class="arithmatex">\(\widehat{r}_{(i)}\)</span>. Advantage of QN is robustness to outliers and to systematic
differences in measurements between studies. A disadvantage is that we
lose some information of the phenotype distribution, which could be useful
if it was modeled properly.</p>
<p>Let's do QN for 100 males and 100 females where the phenotype 
in males follows <span class="arithmatex">\(2+\textrm{Gamma}(\textrm{shape}=1.5,\textrm{scale}=1.5)\)</span> and 
in females <span class="arithmatex">\(6+\textrm{Gamma}(\textrm{shape}=1.5,\textrm{scale}=1.5)\)</span>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>n = 200 #males + females
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>fem = rep( c(0,1), each = n/2) #who is female
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a>y = 2 + rgamma(n, shape = 1.5, scale = 1.5) #males have shift of 2
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>y[fem == 1] = 4 + y[fem == 1] #females have shift of 6 = 2 + 4
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>hist(y, breaks = 30, col = &quot;khaki&quot;) #shows some outliers compared to mixture of 2 Normals
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a>
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a>#regress out sex and take residuals
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a>lm.fit = lm(y ~ fem)
<a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a>r = residuals(lm.fit)
<a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a>
<a id="__codelineno-11-11" name="__codelineno-11-11" href="#__codelineno-11-11"></a>#find QN&#39;ed trait values from qnorm = inverse of cumulative distribution of Normal
<a id="__codelineno-11-12" name="__codelineno-11-12" href="#__codelineno-11-12"></a>inv.normalise &lt;- function(x) { #this would also tolerate NAs
<a id="__codelineno-11-13" name="__codelineno-11-13" href="#__codelineno-11-13"></a>  return( qnorm( (rank(x, na.last = &quot;keep&quot;) - 0.5) / sum(!is.na(x))))}
<a id="__codelineno-11-14" name="__codelineno-11-14" href="#__codelineno-11-14"></a>q = inv.normalise(r)
<a id="__codelineno-11-15" name="__codelineno-11-15" href="#__codelineno-11-15"></a>#Let&#39;s plot y and q (after scaling to mean=0, var=1) 
<a id="__codelineno-11-16" name="__codelineno-11-16" href="#__codelineno-11-16"></a>par(mfrow = c(1,2))
<a id="__codelineno-11-17" name="__codelineno-11-17" href="#__codelineno-11-17"></a>plot(density(scale(y)), col = &quot;black&quot;, lwd = 2, xlim = c(-4,4), ylim = c(0,0.5), 
<a id="__codelineno-11-18" name="__codelineno-11-18" href="#__codelineno-11-18"></a>     xlab = &quot;trait&quot;, main = &quot;&quot; )
<a id="__codelineno-11-19" name="__codelineno-11-19" href="#__codelineno-11-19"></a>lines(density(scale(q)), col = &quot;darkgreen&quot;, lwd = 2)
<a id="__codelineno-11-20" name="__codelineno-11-20" href="#__codelineno-11-20"></a>plot(y, q, col = c(&quot;cyan&quot;,&quot;gold&quot;)[1+fem])
<a id="__codelineno-11-21" name="__codelineno-11-21" href="#__codelineno-11-21"></a>legend(&quot;bottomright&quot;, col = c(&quot;cyan&quot;,&quot;gold&quot;), pch = 1, leg = c(&quot;male&quot;,&quot;female&quot;) )
</code></pre></div>
<p>We see that (right plot) the outliers on the original scale are still 
the most extreme values in QN-scale on y-axis, 
but they are now close to the other values and not anymore outliers. 
The QN'ed trait, obviously, looks perfectly Normal (left plot)
while in the original trait we see outliers on right tail.
We also see
the effect of sex as the two modal distribution of <span class="arithmatex">\(y\)</span>, 
which was evened out by the regression step <strong>before</strong> QN was applied.</p>
<h3 id="122-binary-phenotypes">1.2.2 Binary phenotypes<a class="headerlink" href="#122-binary-phenotypes" title="Permanent link">&para;</a></h3>
<p>How do we look for the association between a genotype and a binary disease status?
Here the phenotype is 0-1 indicator whether an individual is a <strong>case</strong> (y=1; has the disease)
or a <strong>control</strong> (y=0; does not have a disease).
If we take a practical approach, we could just apply the linear regression model 
to binary outcome variable but it is not conceptually satisying 
because the error terms cannot be assumed normally distributed. 
Hence we will use a generalized linear regression model called <strong>logistic regression</strong>.
To derive the model, let's first define the effect size parameter for binary outcomes.</p>
<h4 id="relative-risk-and-odds-ratio">Relative risk and odds ratio<a class="headerlink" href="#relative-risk-and-odds-ratio" title="Permanent link">&para;</a></h4>
<p>To measure whether a genotype affects disease status, we consider the 
<strong>relative risk</strong> (RR) parameter for genotype 1
<span class="arithmatex">\(<span class="arithmatex">\(\lambda_1 = \frac{\textrm{Pr}(Y=1|X=1)}{\textrm{Pr}(Y=1|X=0)}.\)</span>\)</span>
Thus <span class="arithmatex">\(\lambda_1\)</span> tells how many times larger the risk of getting the disease is
for individuals with genotype 1 compared to individuals with genotype 0.
If </p>
<ul>
<li><span class="arithmatex">\(\lambda_1 = 1\)</span>, there is no association to disease,</li>
<li><span class="arithmatex">\(\lambda_1 &gt; 1\)</span>, genotype 1 confers risk to disease,</li>
<li><span class="arithmatex">\(\lambda_1 &lt; 1\)</span>, genotype 1 confers protection to disease.</li>
</ul>
<p>Similarly we define relative risk <span class="arithmatex">\(\lambda_2\)</span> as the factor by which genotype 2 
multiplies the disease risk of genotype 0.</p>
<p>Relative risk is a simple to define, but it turns out not to be very
easy to estimate in the regression setting, particularly in the typical GWAS setting.
However, if we modify the parameter slightly from comparing risks to comparing odds,
we will get a measure that can be estimated in practice.
The odds corresponding to a probability value <span class="arithmatex">\(p\)</span> 
are <span class="arithmatex">\(p/(1-p)\)</span>, i.e. they tell how many times as probable the event is to occur than
it is to not occur. If probability of a disease is 50%, then the odds of disease are 1,
if <span class="arithmatex">\(p=1\%\)</span> then odds are <span class="arithmatex">\(1/99 = 0.0101\)</span> and if <span class="arithmatex">\(p=99\%\)</span> then odds are 99.
Just like with RR we measured the <em>relative</em> increase in risk between two genotypes,
with odds we also use relative odds, also known as <strong>odds ratio</strong>, between genotypes.
So the odds ratio for genotype 1 is
<span class="arithmatex">\(<span class="arithmatex">\(\textrm{OR}_1 = \frac{\textrm{Pr}(Y=1|X=1)}{\textrm{Pr}(Y=0|X=1)} :  \frac{\textrm{Pr}(Y=1|X=0)}{\textrm{Pr}(Y=0|X=0)} =
\frac{\textrm{Pr}(Y=1|X=1) \textrm{Pr}(Y=0|X=0)}{\textrm{Pr}(Y=1|X=0)\textrm{Pr}(Y=0|X=1)}.\)</span>\)</span>
By Bayes formula we
can swap the roles of <span class="arithmatex">\(Y\)</span> and <span class="arithmatex">\(X\)</span> in the above formula, e.g., by writing
<span class="arithmatex">\(\textrm{Pr}(Y=1|X=1) = \textrm{Pr}(X=1|Y=1) \textrm{Pr}(Y=1) / \textrm{Pr}(X=1)\)</span>.
By applying this to all 4 terms, we get
<span class="arithmatex">\(<span class="arithmatex">\(\textrm{OR}_1 = \frac{\textrm{Pr}(X=1|Y=1) \textrm{Pr}(Y=1) \textrm{Pr}(X=0)
\textrm{Pr}(X=0|Y=0) \textrm{Pr}(Y=0) \textrm{Pr}(X=1)} {\textrm{Pr}(X=0|Y=1) \textrm{Pr}(Y=1)\textrm{Pr}(X=1) 
\textrm{Pr}(X=1|Y=0) \textrm{Pr}(Y=0) \textrm{Pr}(X=0)} = 
\frac{\textrm{Pr}(X=1|Y=1) 
\textrm{Pr}(X=0|Y=0) } {\textrm{Pr}(X=0|Y=1)  
\textrm{Pr}(X=1|Y=0) }.\)</span>\)</span>
This shows that we can estimate our target odds ratio for the disease between 
different genotypes
equally well by collecting individuals based on disease status and observing
their genotype distributions. This is the important property why odds ratios
are possible to measure in a typical disease GWAS that collects individuals
by disease status and then measures the genotypes.</p>
<p><strong>Example 1.4. OR and risk.</strong> 
Suppose that we know that there is a strong risk variant for Alzheimer's disease (AD)
with OR=3.0 (for one copy of the risk allele). Suppose that non-carriers of the 
risk allele have a lifetime risk of 15% of AD. How large is the lifetime risk 
for carriers of one or two copies of the risk variant?</p>
<p>Let <span class="arithmatex">\(p_0=0.15\)</span> be the risk in non-carriers. We know that odds of carriers of one copy
are <span class="arithmatex">\(p_1/(1-p_1) = 3.0\cdot p_0/(1-p_0) = 3.0\cdot 0.15/0.85 = 0.5294\)</span>. Hence
the risk is <span class="arithmatex">\(p_1 = 0.5294/(1+0.5294) = 0.346\)</span>. That is, 35%.</p>
<p>For carriers of two copies,
<span class="arithmatex">\(p_2/(1-p_2) = 3.0\cdot p_1/(1-p_1) = 3.0\cdot 0.5294 = 1.5882\)</span> and
<span class="arithmatex">\(p_2 = 1.5882 / (1 + 1.5882) = 0.61363\)</span>. That is, 61%.</p>
<h4 id="inference-for-or-based-on-counts">Inference for OR based on counts<a class="headerlink" href="#inference-for-or-based-on-counts" title="Permanent link">&para;</a></h4>
<p>In practice, GWAS analyses for disease studies are done using regression models since
they have a possibility to account for confounding covariates.
However, sometimes it is also useful to be able to quickly estimate OR parameters
and their uncertainty using count data without access to regression models.
But remember that such estimates cannot be adjusted for covariates 
and therefore can be biased by confounding factors.</p>
<p>Suppose we have observed the following genotype counts:</p>
<table>
<thead>
<tr>
<th>group</th>
<th>genotype 0</th>
<th>genotype 1</th>
<th>genotype 2</th>
</tr>
</thead>
<tbody>
<tr>
<td>cases</td>
<td><span class="arithmatex">\(S_0\)</span></td>
<td><span class="arithmatex">\(S_1\)</span></td>
<td><span class="arithmatex">\(S_2\)</span></td>
</tr>
<tr>
<td>controls</td>
<td><span class="arithmatex">\(R_0\)</span></td>
<td><span class="arithmatex">\(R_1\)</span></td>
<td><span class="arithmatex">\(R_2\)</span></td>
</tr>
</tbody>
</table>
<p>Then the estimates for odds-ratio between genotype 1 and genotype 0, its logarithm
and the standard error of the logOR are
<span class="arithmatex">\(<span class="arithmatex">\(\widehat{\textrm{OR}}_1 = \frac{S_1 R_0}{S_0 R_1},\qquad \log\left(\widehat{\textrm{OR}}_1 \right) = 
\log\left(\frac{S_1 R_0}{S_0 R_1}\right) \qquad \textrm{ and } \qquad
\textrm{SE}\left( \log\left(\widehat{\textrm{OR}}_1 \right) \right) = \sqrt{\frac{1}{R_1} + \frac{1}{R_0} + \frac{1}{S_1} + \frac{1}{S_0}}.\)</span>\)</span>
In particular, standard error can only be calculated for logOR and not for OR because
only the sampling distribution of logOR is approximately Normal.</p>
<p>The 95% confidence interval for logOR is, naturally, 
<span class="arithmatex">\((\textrm{logOR}-1.96\cdot \textrm{SE},\, \textrm{logOR}+1.96\cdot \textrm{SE})\)</span>,<br>
and 95% CI for OR is <span class="arithmatex">\((\exp(\textrm{logOR}-1.96\cdot \textrm{SE}),\, \exp(\textrm{logOR}+1.96\cdot \textrm{SE}))\)</span>.<br>
Thus, the endpoints of the 95%CI must always be computed on the log-odds scale 
and then transformed to the OR scale.</p>
<p>Similar inference can be done for the <span class="arithmatex">\(\textrm{OR}_2\)</span> parameter measuring the odds-ratio
between genotypes 2 and 0 by substituting the counts of genotype 1 with the counts of genotype 2 in the formulas above.</p>
<p>If any counts are very small, or even 0, the SE is not reliable. 
One can add a value of 0.5 to each of the observed counts to
get an OR estimate even in these case, but one shouldn't rely on the SE estimate.</p>
<h4 id="logistic-regression">Logistic regression<a class="headerlink" href="#logistic-regression" title="Permanent link">&para;</a></h4>
<p>Logistic regression model takes the place of linear regression as the 
basic GWAS model when the phenotype is binary. 
It explains the logarithm of the odds of the disease by
the genotype. The simplest model is the <strong>additive</strong> model:
<span class="arithmatex">\(<span class="arithmatex">\(\log\left( \frac{\textrm{Pr}(Y=1|X=x)}{\textrm{Pr}(Y=0|X=x)} \right)
= \mu + x\beta.\)</span>\)</span>
Thus, <span class="arithmatex">\(\mu\)</span> is the logarithm of odds ('log-odds') for genotype 0 and <span class="arithmatex">\(\beta\)</span> is the log of
odds ratio (logOR) between genotype 1 and 0 (and <span class="arithmatex">\(\exp(\beta)\)</span> is the corresponding odds ratio).
Similarly <span class="arithmatex">\(2\beta\)</span> is the logOR between genotypes 2 and 0.
This model is additive on the log-odds scale and hence multiplicative on the odds scale.
Due to this duality, it is sometimes called additive model and sometimes
called multiplicative model, which is a source of confusion. 
In these notes, it is called the additive model.
In R, such a logistic regression model can be fitted by <code>glm(y~x, family="binomial")</code>.</p>
<p>To try out logistic regression, we should learn how to
simulate some case-control data according to
the logistic regression model.</p>
<p><strong>Example 1.5.</strong>
Let's assume that our risk allele <span class="arithmatex">\(A\)</span> has frequency 13% in controls,
and that it is in HWE in controls. If the risk model is additive on log-odds scale
with odds-ratio 1.43 per each copy of allele A, what are the genotype frequencies
in cases?</p>
<p>Let's denote case frequencies by <span class="arithmatex">\(f_0, f_1, f_2\)</span> and control frequencies
by <span class="arithmatex">\(q_0, q_1, q_2.\)</span>
From formulas above, we get that 
<span class="arithmatex">\(<span class="arithmatex">\(f_1 = \textrm{Pr}(A=1|Y=1) = \textrm{OR}_1 \frac{\textrm{Pr}(A=1|Y=0)} {\textrm{Pr}(A=0|Y=0) } \textrm{Pr}(A=0|Y=1) = \textrm{OR}_1 \frac{q_1 f_0}{q_0},\)</span>\)</span>
<span class="arithmatex">\(<span class="arithmatex">\(f_2 = \textrm{Pr}(A=2|Y=1) = \textrm{OR}_2 \frac{\textrm{Pr}(A=2|Y=0)} {\textrm{Pr}(A=0|Y=0) } \textrm{Pr}(A=0|Y=1) = \textrm{OR}_2 \frac{q_2 f_0}{q_0}.\)</span>\)</span>
Since <span class="arithmatex">\(f_0+f_1+f_2=1,\)</span> we get 
<span class="arithmatex">\(<span class="arithmatex">\(f_0 = \left(1 + \textrm{OR}_1 \frac{q_1}{q_0} +\textrm{OR}_2 \frac{q_2 }{q_0}\right)^{-1}.\)</span>\)</span>
Now we can compute the genotype frequencies in cases 
(note <span class="arithmatex">\(\textrm{OR}_2 = \textrm{OR}_1^2\)</span> under the additive model):
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>or = 1.43
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>a.cntrl = 0.13
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>q = c((1-a.cntrl)^2, 2*a.cntrl*(1-a.cntrl), a.cntrl^2) #HWE holds in controls
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a>f.0 = 1/(1 + or*q[2]/q[1] + or^2*q[3]/q[1])
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a>f = c(f.0, or*q[2]/q[1]*f.0, or^2*q[3]/q[1]*f.0)
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a>rbind(controls = q, cases = f) #print out cases and controls
</code></pre></div></p>
<p>Let's generate 2000 cases and controls from these genotype frequencies
and estimate the genetic effect using logistic regression.
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>n = 2000
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>x.cases = sample(c(0,1,2), prob = f, size = n, replace = T)
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a>x.controls = sample(c(0,1,2), prob = q, size = n, replace = T)
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a>x = c(x.cases, x.controls) #genotypes of all samples
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a>y = c(rep(1,n), rep(0,n)) #binary phenotype corresponding to genotypes: 1st cases, then controls
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a>glm.fit = glm(y ~ x, family = &quot;binomial&quot;)
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a>summary(glm.fit)
</code></pre></div>
What is the estimate and 95%CI on odds ratio scale?
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>b = summary(glm.fit)$coeff[2,1] #estimate, beta-hat
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a>se = summary(glm.fit)$coeff[2,2] #standard error
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>exp(c(b, b - 1.96*se, b + 1.96*se)) #endpoints always on logOR scale, then transform to OR scale
</code></pre></div>
Let's compare the result from the additive logistic regression model to
the results that we get from the raw genotype counts between the genotypes 1 and 0
using the formulas above (see "Inference for OR based on counts"):
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>s1 = sum(x.cases == 1); s0 = sum(x.cases == 0); r1 = sum(x.controls == 1); r0 = sum(x.controls == 0)
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>or.1.counts = s1*r0/s0/r1
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>se.1.counts = sqrt(sum( 1/c(s1,s0,r1,r0) ) )
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a>c(or.1.counts, exp(log(or.1.counts) - 1.96*se.1.counts), exp(log(or.1.counts) + 1.96*se.1.counts) )
</code></pre></div>
These are a bit different from the additive model results above. 
The reason is that the additive model above 
also uses the data from individuals with genotype 2 to estimate
the OR parameter. It turns out that if we apply the additive regression model
only to the individuals having either genotype 0 or genotype 1, then we get
essentially the same results
as we get from the raw counts:
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>glm.fit = glm(y[x != 2] ~ x[x != 2], family = &quot;binomial&quot;)
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a>b = summary(glm.fit)$coeff[2,1] #estimate, beta-hat
<a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a>se = summary(glm.fit)$coeff[2,2] #standard error
<a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a>exp(c( b, b - 1.96 * se, b + 1.96 * se )) 
</code></pre></div>
Remember that even though the count based inference and logistic regression inference
based on additive model gives similar results here, they may give very different
results when there are confounding covariates included in the regression model, and
in such situations we trust more in the regression model results.
We come back to this later.</p>
<p>Analogous to the quantitative phenotypes, we can use the full model also
for the binary data:
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>z = as.numeric( x == 2 )
<a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>glm.full = glm( y ~ x + z, family = &quot;binomial&quot;)
<a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a>summary(glm.full)
</code></pre></div>
In this example, we have no deviation from additivity.</p>
<h4 id="ascertained-case-control-studies">Ascertained case-control studies<a class="headerlink" href="#ascertained-case-control-studies" title="Permanent link">&para;</a></h4>
<p>Suppose we are studying MS disease whose prevalence is about 1/1000.
Even if we collect 100,000 individuals from the population, we still 
would get only about 100 cases! In ascertained case-control studies
we enrich cases by collecting a case sample directly from the diseased people 
and similarly we collect controls either from the general population or, even better,
from the individuals we know are disease free. (For diseases with prevalence &lt; 1%
there is little difference between these control ascertainment strategies.)
Thus by a phenotype-based ascertainment 
we may have a GWAS of 10,000 individuals that is divided into 
sample of 5,000 cases and 5,000 controls. This approach gives much
higher power than population collection of 
100 cases and ~100,000 controls. (We will do power analyses next week.)</p>
<p>Can we analyse such ascertained case-control samples using the same logistic regression
model as applied above? The answer is yes. The parameter <span class="arithmatex">\(\beta\)</span> is the logOR
and we showed earlier that this parameter can be estimated also
by ascertaining individuals based on their phenotypes. 
This result also extends to the use of logistic regression.
However, 
the parameter <span class="arithmatex">\(\mu\)</span> that determines the absolute odds of disease for genotype class 0
depends on the sampling strategy, i.e., on the proportion of cases in the data.
Thus, in ascertained data, <span class="arithmatex">\(\mu\)</span> does NOT estimate the population prevalence parameter
of the genotype group 0.
However, we can still apply the logistic regression model to
the ascertained case-control sample 
and estimate the three central association statistics:
genetic effect <span class="arithmatex">\(\beta\)</span>, its uncertainty (standard error), 
and P-value. </p>
<h4 id="gwas-software">GWAS software<a class="headerlink" href="#gwas-software" title="Permanent link">&para;</a></h4>
<p>As current GWAS consider 10,000s of individuals and millions of variants,
those analyses are done with a specific software that read the specific file formats.
Most popular software is PLINK in its recent, efficient version <a href="https://www.cog-genomics.org/plink/2.0/">2.0</a>.
Another widely used software is  <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html">SNPTEST</a>.</p>
<p>In this course, we do not focus on the commands or input file formats
of any particular software, since that would take already all our time
and the software are in constant development.
The goal of this course is to understand why each analysis is done 
and how to interpret the output, in particular from a statistical
point of view. These skills are independent of the GWAS software.</p>


  


  


              
            </article>
            
          </div>
        </div>
        
          <a href="#" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </a>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
      
      
    
    <a href="https://github.com/hds-sandbox" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.1.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
    
    
      
      
    
    <a href="https://twitter.com/ucph_heads" target="_blank" rel="noopener" title="twitter.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.1.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.ecf98df9.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.9c69f0bc.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>