<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Conor O’Hare">
<meta name="author" content="Samuele Soraggi">
<meta name="author" content="Alba Refoyo Martinez">

<title>Genome-Wide Association Studies - Association testing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-EXW13D5861"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-EXW13D5861', { 'anonymize_ip': true});
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../img/logo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Genome-Wide Association Studies</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-access" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Access</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-access">    
        <li>
    <a class="dropdown-item" href="../access/index.html">
 <span class="dropdown-text">General</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../access/UCloud.html">
 <span class="dropdown-text">UCloud</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../access/genomedk.html">
 <span class="dropdown-text">Genome</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tutorial" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Tutorial</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tutorial">    
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS1-WhatIsAGWAS.html">
 <span class="dropdown-text">What is a GWAS?</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS2-DataCollection.html">
 <span class="dropdown-text">Data Collection</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS3-QualityControlA.html">
 <span class="dropdown-text">Quality Control A</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS4-QualityControlB.html">
 <span class="dropdown-text">Quality Control B?</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS5-AssociationTesting.html">
 <span class="dropdown-text">Association Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS6-PRSAnalysis.html">
 <span class="dropdown-text">PRS Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../Notebooks/GWAS7-OtherToolsAndTesting.html">
 <span class="dropdown-text">Other tools and testing</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#biometric-model" id="toc-biometric-model" class="nav-link active" data-scroll-target="#biometric-model"><span class="header-section-number">0.2.1</span> Biometric model</a></li>
  <li><a href="#modelling" id="toc-modelling" class="nav-link" data-scroll-target="#modelling"><span class="header-section-number">1</span> 6.1 Modelling</a>
  <ul class="collapse">
  <li><a href="#the-linearlogistic-model" id="toc-the-linearlogistic-model" class="nav-link" data-scroll-target="#the-linearlogistic-model"><span class="header-section-number">1.0.1</span> The linear/logistic model</a></li>
  <li><a href="#plink-in-association-testing" id="toc-plink-in-association-testing" class="nav-link" data-scroll-target="#plink-in-association-testing"><span class="header-section-number">1.0.2</span> PLINK in association testing</a></li>
  <li><a href="#analysis-with-plink" id="toc-analysis-with-plink" class="nav-link" data-scroll-target="#analysis-with-plink"><span class="header-section-number">1.0.3</span> 6.1.1 Analysis with PLINK</a></li>
  </ul></li>
  <li><a href="#correction-for-multiple-testing" id="toc-correction-for-multiple-testing" class="nav-link" data-scroll-target="#correction-for-multiple-testing"><span class="header-section-number">2</span> 6.2 Correction for multiple testing</a>
  <ul class="collapse">
  <li><a href="#commands" id="toc-commands" class="nav-link" data-scroll-target="#commands"><span class="header-section-number">2.0.1</span> 6.3.1 Commands</a></li>
  <li><a href="#manhattan-and-qq-plots" id="toc-manhattan-and-qq-plots" class="nav-link" data-scroll-target="#manhattan-and-qq-plots"><span class="header-section-number">2.0.2</span> Manhattan and QQ-plots</a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="GWAS5-AssociationTesting.out.ipynb" download="GWAS5-AssociationTesting.out.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Association testing</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Conor O’Hare </p>
             <p>Samuele Soraggi <a href="mailto:samuele@birc.au.dk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-1159-5535" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
             <p>Alba Refoyo Martinez <a href="mailto:alba.martinez@sund.ku.dk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-3674-4007" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-note no-icon callout-titled" title="Important notes for this notebook">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important notes for this notebook
</div>
</div>
<div class="callout-body-container callout-body">
<p>After QC and calculation of MDS components, the data is ready for subsequent association tests. The appropriate statistical test can be selected depending on the expected genetic model of the trait or disease of interest and the nature of the phenotypic trait studied. In this tutorial, we provide scripts for various types of association that are suitable for binary traits (e.g., alcohol-dependent patients vs.&nbsp;healthy controls) or quantitative traits (e.g., the number of alcoholic beverages consumed per week).</p>
<section id="learning-outcomes" class="level2" data-number="0.1">
<h2 data-number="0.1" class="anchored" data-anchor-id="learning-outcomes"><span class="header-section-number">0.1</span> Learning outcomes</h2>
<ul>
<li><strong>Discuss</strong> different types of association tests</li>
<li><strong>Identify</strong> the suitable association test for your data</li>
</ul>
</section>
<section id="how-to-make-this-notebook-work" class="level2" data-number="0.2">
<h2 data-number="0.2" class="anchored" data-anchor-id="how-to-make-this-notebook-work"><span class="header-section-number">0.2</span> How to make this notebook work</h2>
<ul>
<li>In this notebook, we will use both <code>R</code> and <code>bash command line</code> programming languages. Remember to <strong>change the kernel</strong> whenever you transition from one language to the other (<code>Kernel --&gt; Change Kernel</code>) indicated by the languages’ images. We will first run <code>Bash</code> commands.</li>
</ul>
<p><img src="Images/bash.png" alt="Bash" width="80"> Choose the Bash kernel</p>
</section>
</div>
</div>
<section id="biometric-model" class="level3" data-number="0.2.1">
<h3 data-number="0.2.1" class="anchored" data-anchor-id="biometric-model"><span class="header-section-number">0.2.1</span> Biometric model</h3>
<p>The theory of genetic association is based on the biometrical model. Here’s an example (taken from the Biometrical Genetics chapter in <span class="citation" data-cites="neale2013methodology">Neale and Cardon (<a href="#ref-neale2013methodology" role="doc-biblioref">2013</a>)</span>. The biometric model (first established by Fisher) considers each variant having an additive effect on the value of a trait. Consider for a specific variant its possible genotypes. Each of those has a specific effect on the value of a trait (of mean value m). <code>aa</code> contributes with <code>-d</code>, AA with <code>d</code>. <code>d</code> is the measured difference between the homozygotes <code>AA</code> and <code>aa</code>, with <code>m</code> their average effect. <code>h</code> is themeasured effect of the heterozygote <code>Aa</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="Images/biometric.png" class="img-fluid quarto-figure quarto-figure-center figure-img" label="fig-biometricmodel"></p>
<figcaption>A biometric model showing the two parameters <code>d</code> and <code>h</code> describing the effect of a variant’s genotypes on a trait.</figcaption>
</figure>
</div>
<p>Multiple variants have an additive effect on a trait, meaning <code>d</code> and <code>h</code> effects will add across variants associated to a trait.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Example">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Example
</div>
</div>
<div class="callout-body-container callout-body">
<p>Imagine that we are talking about genes that influence adult stature. Let us assume that the normal range for males is from 4 feet 10 inches (1.47m) to 6 feet 8 inches (2m): that is, about <strong>22 inches (0.56m)</strong>. And let us assume that each somatic chromosome has one gene of roughly equivalent effect. Then, roughly speaking, we are thinking in terms of loci for which the <strong>homozygotes contribute ±d=±1/2 inch (from the midpoint)</strong>, depending on whether they are <em>AA</em>, the increasing homozygote, or <em>aa</em>, the decreasing homozygote. In reality, although some loci may contribute greater effects than this, others will almost certainly contribute less; thus, we are talking about the kind of model in which any particular polygene is having an effect that would be difficult to detect by the methods of classical genetics.</p>
</div>
</div>
<p>Typically in GWAS, we use linear models for when the phenotype is continuous (BMI, height, blood pressure) and logistic models for when the phenotype is binary (whether you have the disease or not). Both of these models should incorporate an additional random effect term <span class="math inline">\(e\)</span>, and this is to minimize the effect of covariates. For example, in the above example, the same height genes might affect females in the same way they do males, but the range might be smaller.</p>
</section>
<section id="modelling" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> 6.1 Modelling</h1>
<section id="the-linearlogistic-model" class="level3" data-number="1.0.1">
<h3 data-number="1.0.1" class="anchored" data-anchor-id="the-linearlogistic-model"><span class="header-section-number">1.0.1</span> The linear/logistic model</h3>
<p>The basic linear regression model is written as:</p>
<p><span class="math display">\[y = G\beta_G + X\beta_X + \epsilon\]</span></p>
<p>Here, <span class="math inline">\(y\)</span> is the phenotype vector, <span class="math inline">\(G\)</span> is the genotype/dosage matrix (the allele counts) for the current variant, <span class="math inline">\(X\)</span> is the fixed-covariate matrix, and <span class="math inline">\(e\)</span> is the error term subject to least-squares minimization.</p>
<p>For binary phenotypes, <code>--logistic</code> in PLINK fits a logistic or Firth regression model instead, with the same <span class="math inline">\(G\beta_G + X\beta_X\)</span> terms. More details about PLINK’s association methods can be found <a href="https://www.cog-genomics.org/plink/2.0/assoc">here</a>.</p>
<p>Within PLINK, the association between SNPs and quantitative outcome measures can be tested with the options <code>‐‐assoc</code> and <code>‐‐linear</code>. When PLINK detects a quantitative outcome measure (i.e., values other than 1, 2, 0, or NA), the <code>‐‐assoc</code> option will automatically treat it as such by performing an asymptotic version of the usual Student’s t-test to compare two means. This option does not allow the use of covariates. The <code>‐‐linear</code> option in PLINK performs a linear regression analysis with each SNP as a predictor. Similar to the <code>‐‐logistic</code> option, the <code>‐‐linear</code> option enables the use of covariates and is somewhat slower than the <code>‐‐assoc</code> option.</p>
</section>
<section id="plink-in-association-testing" class="level3" data-number="1.0.2">
<h3 data-number="1.0.2" class="anchored" data-anchor-id="plink-in-association-testing"><span class="header-section-number">1.0.2</span> PLINK in association testing</h3>
<p>PLINK offers one degree of freedom (1 df) allelic tests in which the trait value, or the log‐odds of a binary trait, increases or decreases linearly as a function of the number of risk alleles (minor allele <em>a</em> against major allele <em>A</em>). In addition, non‐additive tests are available: * the genotypic association test (2 df: <em>aa</em> vs <em>Aa</em> vs <em>AA</em>) * the dominant gene action test (1 df: <em>aa</em> &amp; <em>Aa</em> vs <em>AA</em>) * the recessive gene action test, (1 df: aa vs <em>Aa</em> &amp; <em>AA</em>)</p>
<p>However, non‐additive tests are not widely applied, because the statistical power to detect non‐additivity is low in practice. More complex analyses (e.g., Cox regression analysis) can be performed by using R‐based “plug‐in” functions in PLINK.</p>
</section>
<section id="analysis-with-plink" class="level3" data-number="1.0.3">
<h3 data-number="1.0.3" class="anchored" data-anchor-id="analysis-with-plink"><span class="header-section-number">1.0.3</span> 6.1.1 Analysis with PLINK</h3>
<p><img src="Images/bash.png" alt="Bash" width="40"> We link the data folder to easily access it and create a folder for the output files generated in this notebook.</p>
<div id="fc7317ea-43cf-4538-80f5-88d38de2e2c4" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>ln <span class="sc">-</span>sf ..<span class="sc">/</span>Data</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>mkdir <span class="sc">-</span>p Results<span class="sc">/</span>GWAS5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Recall that our dataset contains binary traits: phenotype can take two values 0 or 1. In this tutorial, we will apply both <code>--assoc</code>and <code>--logistic</code> to our data. A reminder that <code>--assoc</code> option does not allow to correct covariates such as principal components (PCs)/MDS components for population stratification, which makes it less suited for association analyses. But it’s worth seeing the results to make this clear.</p>
<p>We start with <code>--assoc</code>:</p>
<div id="959006a4" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>plink <span class="sc">--</span>bfile Results<span class="sc">/</span>GWAS4<span class="sc">/</span>HapMap_3_r3_10 <span class="sc">--</span>assoc <span class="sc">--</span>out Results<span class="sc">/</span>GWAS5<span class="sc">/</span>assoc_results <span class="sc">--</span>silent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s have a look at the association output file:</p>
<div id="ba8e2613" class="cell" data-tags="[]" data-execution_count="8">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">5</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>assoc_results.assoc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CHR         SNP         BP   A1      F_A      F_U   A2        CHISQ            P           OR 
   1   rs3131972     742584    A   0.1944   0.1455    G       0.9281       0.3354        1.418 
   1   rs3131969     744045    A   0.1759      0.1    G        2.647       0.1037        1.921 
   1   rs1048488     750775    C   0.1981   0.1455    T        1.054       0.3045        1.451 
   1  rs12562034     758311    A  0.06481   0.1182    G        1.863       0.1723       0.5171 </code></pre>
</div>
</div>
<p>where the columns represent:</p>
<ul>
<li><code>CHR</code> Chromosome code<br>
</li>
<li><code>SNP</code> Variant identifier<br>
</li>
<li><code>BP</code> Base-pair coordinate<br>
</li>
<li><code>A1</code> Allele 1 (usually minor)<br>
</li>
<li><code>F_A</code> Allele 1 frequency among cases<br>
</li>
<li><code>F_U</code> Allele 1 frequency among controls<br>
</li>
<li><code>A2</code> Allele 2<br>
</li>
<li><code>CHISQ</code> Allelic test chi-square statistic</li>
<li><code>P</code> Allelic test p-value. The p-value indicates the probability that the observed association (or one more extreme) would occur by chance if there is no true association between the SNP and the trait.</li>
<li><code>OR</code> odds(allele 1 | case) / odds(allele 1 | control)</li>
</ul>
<p>Applying <code>--logistic</code>, we can now consider principal components as covariates. We will use the MDS components <code>covar_mds.txt</code> from the previous tutorial (option <code>--covar</code>). We also remove the lines containing NA values in the output, because later will give problems in doing plots.:</p>
<div id="a1242c4f" class="cell" data-tags="[]" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --logistic </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note, we use the option --hide-covar to only show the additive results of the SNPs in the output file.</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>plink <span class="sc">--</span>bfile Results<span class="sc">/</span>GWAS4<span class="sc">/</span>HapMap_3_r3_10 <span class="sc">--</span>covar Results<span class="sc">/</span>GWAS4<span class="sc">/</span>covar_mds.txt <span class="sc">--</span>logistic <span class="sc">--</span>hide<span class="sc">-</span>covar <span class="sc">--</span>out Results<span class="sc">/</span>GWAS5<span class="sc">/</span>logistic_results <span class="sc">--</span>silent</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove NA values, those might give problems generating plots in later steps.</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>awk <span class="st">'!/'</span>NA<span class="st">'/'</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>logistic_results.assoc.logistic <span class="sc">&gt;</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>logistic_results.assoc_2.logistic</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The columns <code>CHR, SNP, BP, A1, OR, P</code> have the same meaning as with <code>--assoc</code> testing. The new ones are</p>
<ul>
<li><code>TEST</code> the type of test performed. It usually includes “ADD” (additive model) and may also include other genetic models depending on the options specified (e.g., dominant, recessive).</li>
<li><code>NMISS</code> the number of non-missing observations (individuals) used in the analysis for that particular SNP.</li>
<li><code>STAT</code> the test t-statistic for the logistic regression test</li>
</ul>
<div id="5ef27076-2a68-4c86-bbf6-08ea46f68361" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">5</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>logistic_results.assoc.logistic</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CHR         SNP         BP   A1       TEST    NMISS         OR         STAT            P 
   1   rs3131972     742584    A        ADD      109      1.782        1.363       0.1729
   1   rs3131969     744045    A        ADD      109      2.195        1.705      0.08811
   1   rs1048488     750775    C        ADD      108      1.801         1.39       0.1644
   1  rs12562034     758311    A        ADD      109     0.4209       -1.526        0.127</code></pre>
</div>
</div>
<p>The results obtained from these GWAS analyses will be visualized in the last step. This will also show if the data set contains any genome-wide significant SNPs.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>in case of a quantitative outcome measure the option <code>--logistic</code> should be replaced by <code>--linear</code>. The use of the <code>--assoc</code> option is also possible for quantitative outcome measures (as metioned previously, this option does not allow the use of covariates).</p>
</div>
</div>
</section>
</section>
<section id="correction-for-multiple-testing" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> 6.2 Correction for multiple testing</h1>
<p>Modern genotyping arrays can genotype up to 4 million markers, hence, multiple testing becomes a problem in our analyses. Why?</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Multiple testing is a problem because it increases the rate of false positives. In easier terms, you have a very small chance to get a falsely significant p-value (false positive) when running a single comparison, but the rate of false positives increases very quickly when running many comparisons, such as millions as it happens with GWAS datasets.</p>
<p>To learn more about statistical testing and false positives, <a href="https://web.stanford.edu/class/bios221/book/06-chap.html">look at this online book chapter</a>.</p>
</div>
</div>
<p>Three widely applied methods for determining genome‐wide significance are the use of Bonferroni correction, Benjamini–Hochberg false discovery rate (FDR), and permutation testing.</p>
<section id="bonferroni-correction" class="level4" data-number="2.0.0.1">
<h4 data-number="2.0.0.1" class="anchored" data-anchor-id="bonferroni-correction"><span class="header-section-number">2.0.0.1</span> Bonferroni correction</h4>
<p>The Bonferroni correction aims to control the probability of having at least one false positive finding due to SNPs not being independent. It calculates the adjusted p-value threshold with the formula <span class="math inline">\(0.05/n\)</span>, where <span class="math inline">\(n\)</span> is the number of SNPs tested, and the adjusted p-value as <span class="math inline">\(p/n\)</span>. However, many SNPs are correlated due to linkage disequilibrium (LD). The Bonferroni method can be conservative, leading to unnecessary exclusions and consequently, to an increase in the proportion of false negative findings.</p>
</section>
<section id="false-discovery-rate" class="level4" data-number="2.0.0.2">
<h4 data-number="2.0.0.2" class="anchored" data-anchor-id="false-discovery-rate"><span class="header-section-number">2.0.0.2</span> False discovery rate</h4>
<p>FDR controls the expected proportion of false positives among all signals with an FDR value below a fixed threshold, assuming that SNPs are independent. Whilst this method is less conservative than Bonferroni correction, controlling for FDR does not imply any notion of statistical significance; it is merely a method to minimize the expected proportion of false positives. Moreover, this method has its limitations as SNPs, and thus p-values, are not independent whereas this is an assumption of the FDR method.</p>
</section>
<section id="permutation-methods" class="level4" data-number="2.0.0.3">
<h4 data-number="2.0.0.3" class="anchored" data-anchor-id="permutation-methods"><span class="header-section-number">2.0.0.3</span> Permutation methods</h4>
<p>The outcome measure labels are randomly permuted multiple times (say, on the order of 1 million) which effectively removes any true association between the outcome measure and the genotype. Then, for all permuted data sets, the statistical tests provide the empirical distribution of the test statistic and the p-values under the null hypothesis of no association. The original test statistic or p-value obtained from the observed data is subsequently compared to the empirical distribution of p-values to determine an empirically adjusted p-value.</p>
</section>
<section id="commands" class="level3" data-number="2.0.1">
<h3 data-number="2.0.1" class="anchored" data-anchor-id="commands"><span class="header-section-number">2.0.1</span> 6.3.1 Commands</h3>
<p>Both the Bonferroni correction and FDR are straighforward to implement using PLINK, with the command <code>--adjust</code>:</p>
<div id="4a8686d5" class="cell" data-tags="[]" data-execution_count="9">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --adjust</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>plink <span class="sc">--</span>bfile Results<span class="sc">/</span>GWAS4<span class="sc">/</span>HapMap_3_r3_10 <span class="sc">-</span>assoc <span class="sc">--</span>adjust <span class="sc">--</span>out Results<span class="sc">/</span>GWAS5<span class="sc">/</span>adjusted_assoc_results <span class="sc">--</span>silent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s look at the output file from PLINK:</p>
<div id="2019b3d9" class="cell" data-tags="[]">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">5</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>adjusted_assoc_results.assoc.adjusted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>:::{.callout-important title=“Stop - Read - Solve”}</p>
<p>The Bonferroni-corrected p-value threshold is calculated using an initial threshold of 0.05. 1. How is this threshold exactly determined? 2. Using that threshold, do any of the SNPs in your dataset seem to be significantly associated with the phenotype?</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Hint">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Hint
</div>
</div>
<div class="callout-body-container callout-body">
<p>Calculate the Bonferroni threshold based on the number of SNPs in your dataset using either Python or R.</p>
<ol type="1">
<li>Find the number of SNPs by running the following command on your .bim file: <code>cat Results/GWAS4/HapMap_3_r3_10.bim | wc -l</code>.</li>
<li>Apply the formula to calculate the threshold: <code>bonferroni_threshold &lt;- alpha / num_snps</code>, being alpha the significant level (0.05) and num_snps, the total number of SNPs in your dataset (num_snps=1073788).</li>
</ol>
</div>
</div>
<div id="42533a52-ca04-422e-a240-44c93adce882" class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write code here and change kernel accordingly </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The Bonferroni correction for all SNPs (<code>BONF</code>) returns a value of 1, meaning that the SNP does not have a significant association with the phenotype given the threshold <span class="math inline">\(0.05/n\)</span>. As stated above, this is an indication that this method could be too conservative and suggests that we have many false negatives (Type II errors).</p>
</div>
</div>
</div>
<p><img src="Images/R.png" alt="R" width="40"> This is not the case for the Benjamini &amp; Hochberg method of FDR (<code>FDR-BH</code>).</p>
<p>In the code below, we will plot the distribution of values to illustrate this.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of the code chunk, please add some code to print a summary of the column using the <code>summary()</code> function. - What is the minimum FDR-adjusted p-value? and the mean FDR-adjusted p-value? - Are there any SNPs significantly associated with the phenotype after applying genome-wide correction?</p>
</div>
</div>
<div id="575384cc-8ed9-4a0c-a2c5-e0bf13211c64" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">suppressWarnings</span>(<span class="fu">library</span>(ggplot2)))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width =</span> <span class="dv">9</span>, <span class="at">repr.plot.height =</span> <span class="dv">4</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>tests <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/adjusted_assoc_results.assoc.adjusted"</span>, <span class="at">header=</span>T)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>hist.relatedness <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(tests, <span class="fu">aes</span>(<span class="at">x=</span>FDR_BH)) <span class="sc">+</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.02</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">fill=</span><span class="st">"tomato"</span>) <span class="sc">+</span> </span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of p-values for FDR correction (B-H correction)"</span>) <span class="sc">+</span> </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"p-value"</span>) <span class="sc">+</span> </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Frequency"</span>) <span class="sc">+</span> </span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>))</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(hist.relatedness)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Write code here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.2872  0.9647  0.9742  0.9710  0.9742  1.0000 </code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-10-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">
<p>In R: <code>summary(tests$FDR_BH)</code> - Min=0.29 - Mean=0.97</p>
<p>The black line indicates that there are very few values with an FDR-adjusted p-value lower than ~0.95, with the lowest being ~0.29 (corresponding to the Min. value printed in the summary table). Thus, in this particular case, we see that no single variant is significant at the 0.05 level after genome-wide correction.</p>
</div>
</div>
<p>Finally, we can have a look at the permutation correction. Since this is a computationally intensive way of testing, we run it on a subset of the data including only chromosome 22. We will also need the option <code>--mperm</code> to define how many permutations we want to do. The subset of the data is done below:</p>
<div id="73e10fe8" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>awk <span class="st">'{ if ($1 == 22) print $2 }'</span> Results<span class="sc">/</span>GWAS4<span class="sc">/</span>HapMap_3_r3_10.bim <span class="sc">&gt;</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>subset_snp_chr_22.txt</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>plink <span class="sc">--</span>bfile Results<span class="sc">/</span>GWAS4<span class="sc">/</span>HapMap_3_r3_10 <span class="sc">--</span>extract Results<span class="sc">/</span>GWAS5<span class="sc">/</span>subset_snp_chr_22.txt <span class="sc">--</span>make<span class="sc">-</span>bed \</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>out Results<span class="sc">/</span>GWAS5<span class="sc">/</span>HapMap_subset_for_perm <span class="sc">--</span>silent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we run permutation testing (we choose 100K permutations to avoid waiting too long)</p>
<div id="98069683-b3c5-40fb-966f-1c2f6ebcb783" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>plink <span class="sc">--</span>bfile Results<span class="sc">/</span>GWAS5<span class="sc">/</span>HapMap_subset_for_perm <span class="sc">--</span>assoc <span class="sc">--</span>mperm <span class="dv">1000000</span> <span class="sc">--</span>threads <span class="dv">4</span> \</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="sc">--</span>out Results<span class="sc">/</span>GWAS5<span class="sc">/</span>perm_result <span class="sc">--</span>silent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The table in output looks like this, with the <code>EMP2</code> column indicating the corrected p-value</p>
<div id="c439644f-def6-45bb-957a-2f4b89eb0a5e" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">5</span> Results<span class="sc">/</span>GWAS5<span class="sc">/</span>perm_result.assoc.mperm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CHR          SNP         EMP1         EMP2 
  22   rs11089128      0.09869            1 
  22    rs7288972       0.3679            1 
  22   rs11167319       0.7516            1 
  22    rs8138488      0.01473            1 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>Are there any significant SNPs?</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p><img src="Images/R.png" alt="R" width="40"> Again, we do not infer any significance with the permutation correction.</p>
<div id="032a8c79-7ce1-4b4c-8c40-708967e5b571" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">suppressWarnings</span>(<span class="fu">library</span>(ggplot2)))</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width =</span> <span class="dv">9</span>, <span class="at">repr.plot.height =</span> <span class="dv">4</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>tests <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/perm_result.assoc.mperm"</span>, <span class="at">header=</span>T)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>hist.relatedness <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(tests, <span class="fu">aes</span>(<span class="at">x=</span>EMP2)) <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.02</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">fill=</span><span class="st">"tomato"</span>) <span class="sc">+</span> </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of p-values for permutation correction"</span>) <span class="sc">+</span> </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"p-value"</span>) <span class="sc">+</span> </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Frequency"</span>) <span class="sc">+</span> </span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>))</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="fu">show</span>(hist.relatedness)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>All in all, when correcting for multiple testing, it is important to keep in mind that different correction measures have different properties, and it is up to the investigator to decide which to use and how to interpret them. In general, <strong>permutation correction is considered to be the gold standard in GWAS analysis</strong>, but it is not always effective when used with specific statistical models (<span class="citation" data-cites="joo_multiple_2016">Joo et al. (<a href="#ref-joo_multiple_2016" role="doc-biblioref">2016</a>)</span> , <span class="citation" data-cites="uffelmann_genome-wide_2021">Uffelmann et al. (<a href="#ref-uffelmann_genome-wide_2021" role="doc-biblioref">2021</a>)</span>), a problematic you should be aware of.</p>
</section>
<section id="manhattan-and-qq-plots" class="level3" data-number="2.0.2">
<h3 data-number="2.0.2" class="anchored" data-anchor-id="manhattan-and-qq-plots"><span class="header-section-number">2.0.2</span> Manhattan and QQ-plots</h3>
<p><img src="Images/R.png" alt="R" width="40"></p>
<p>We can make suggestions for the presence of high-association alleles using a visual approach. The first is called a <strong>Manhattan plot</strong>. The idea is straightforward: for each SNP, we plot it against the negative log of its p-value from the test of association <span class="citation" data-cites="gibson2010hints">(<a href="#ref-gibson2010hints" role="doc-biblioref">Gibson 2010</a>)</span>. We can generate these plots with the <code>R</code> package <code>qqman</code> using the <code>manhattan()</code> function.</p>
<div id="600bbfc8" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Setup to avoid long messages and plot on screen</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">jupyter.plot_mimetypes =</span> <span class="st">'image/png'</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Load GWAS package qqman</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"qqman"</span>))</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Manhattan plot using --logistic results</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>results_log <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/logistic_results.assoc_2.logistic"</span>, <span class="at">head=</span><span class="cn">TRUE</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">manhattan</span>(results_log,<span class="at">chr=</span><span class="st">"CHR"</span>,<span class="at">bp=</span><span class="st">"BP"</span>,<span class="at">p=</span><span class="st">"P"</span>,<span class="at">snp=</span><span class="st">"SNP"</span>, <span class="at">main =</span> <span class="st">"Manhattan plot: logistic"</span>)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Manhattan plot using --assoc</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>results_as <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/assoc_results.assoc"</span>, <span class="at">head=</span><span class="cn">TRUE</span>)</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="fu">manhattan</span>(results_as,<span class="at">chr=</span><span class="st">"CHR"</span>,<span class="at">bp=</span><span class="st">"BP"</span>,<span class="at">p=</span><span class="st">"P"</span>,<span class="at">snp=</span><span class="st">"SNP"</span>, <span class="at">main =</span> <span class="st">"Manhattan plot: assoc"</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-15-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-15-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>The blue line represents the threshold for significance (in the two plots, <span class="math inline">\(10^{-5}\)</span>). We can see that there are no significant SNPs associated with the phenotype when we use the <code>--logistic</code> command (first plot). However, when we use the <code>--assoc</code> command (second plot), we obtain significant SNPs.</p>
<p>Why is there a difference?</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Recall from the beginning of this chapter that the <code>--assoc</code> command does not correct for covariates. So even though we have promising (and hopefully publishable!) results, this form of analysis may be flawed by the underlying population stratification, which is taken into account with the <code>--logistic</code> model.</p>
</div>
</div>
</div>
<p>The second method of visually determining significance is to use a <a href="https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot"><strong>QQ-plot</strong></a>. This plots the expected <span class="math inline">\(-\log_{10}p\)</span> value against the observed <span class="math inline">\(-\log_{10}p\)</span> value. It’s a good way to observe not only outliers that could have significant associations but also peculiarities within our data. For example, if a plot suggests an extreme deviation between the x- and y-axes, then there might be an error with our analyses or data.</p>
<p>We will create these plots using the qq() function from the qqman package in R.</p>
<div id="71434c4b-e28f-4ed5-b81e-0d4935934374" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Setup to avoid long messages and plot on screen</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn=</span><span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">jupyter.plot_mimetypes =</span> <span class="st">'image/png'</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Install and load GWAS package qqman</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">"qqman"</span>)) </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># QQ plot for --logistic</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>results_log <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/logistic_results.assoc_2.logistic"</span>, <span class="at">head=</span><span class="cn">TRUE</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="fu">qq</span>(results_log<span class="sc">$</span>P, <span class="at">main =</span> <span class="st">"Q-Q plot of GWAS p-values (log) using --logistic"</span>)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co"># QQ plot for --assoc</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>results_as <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Results/GWAS5/assoc_results.assoc"</span>, <span class="at">head=</span><span class="cn">TRUE</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="fu">qq</span>(results_as<span class="sc">$</span>P, <span class="at">main =</span> <span class="st">"Q-Q plot of GWAS p-values (log) using --assoc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS5-AssociationTesting_files/figure-html/cell-16-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can observe that in the upper half of the first plot (<code>--logistic</code>), the observed values are lower than the expected values. This pattern is consistent with what we see in the corresponding Manhattan plot. Additionally, examining the Q-Q plot generated from the <code>--assoc</code> command output reveals SNPs where the expected p-values exceed the observed p-values.</p>
<p>It’s important to remember that, while this suggests an association between these SNPs and the studied phenotype, there isn’t sufficient information here to determine the <strong>causal variant</strong>. In fact, there could potentially be multiple causal variants and the causal variants could be in LD with some of the significant variants. Identifying the causal variants would require further investigation using biological methods. However, this analysis has significantly reduced the number of SNPs that need to be studied.</p>



</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-gibson2010hints" class="csl-entry" role="listitem">
Gibson, Greg. 2010. <span>“Hints of Hidden Heritability in GWAS.”</span> <em>Nature Genetics</em> 42 (7): 558–60.
</div>
<div id="ref-joo_multiple_2016" class="csl-entry" role="listitem">
Joo, Jong Wha J., Farhad Hormozdiari, Buhm Han, and Eleazar Eskin. 2016. <span>“Multiple Testing Correction in Linear Mixed Models.”</span> <em>Genome Biology</em> 17 (1): 62. <a href="https://doi.org/10.1186/s13059-016-0903-6">https://doi.org/10.1186/s13059-016-0903-6</a>.
</div>
<div id="ref-neale2013methodology" class="csl-entry" role="listitem">
Neale, MCCL, and Lon R Cardon. 2013. <em>Methodology for Genetic Studies of Twins and Families</em>. Vol. 67. Springer Science &amp; Business Media.
</div>
<div id="ref-uffelmann_genome-wide_2021" class="csl-entry" role="listitem">
Uffelmann, Emil, Qin Qin Huang, Nchangwi Syntia Munung, Jantina de Vries, Yukinori Okada, Alicia R. Martin, Hilary C. Martin, Tuuli Lappalainen, and Danielle Posthuma. 2021. <span>“Genome-Wide Association Studies.”</span> <em>Nature Reviews Methods Primers</em> 1 (1): 1–21. <a href="https://doi.org/10.1038/s43586-021-00056-9">https://doi.org/10.1038/s43586-021-00056-9</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>