<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Quality Control: initial steps – GWAS</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-8ea72dc5fed832574809a9c94082fbbb.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-ea0b82cd5f822071eb0559dd9120d05c.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark-82b651110b069aa65eb395552b20c80d.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-EXW13D5861"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-EXW13D5861', { 'anonymize_ip': true});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="http://hds-sandbox.github.io/" class="navbar-brand navbar-brand-logo">
    <img src="../img/logo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="http://hds-sandbox.github.io/">
    <span class="navbar-title">GWAS</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../access/setup.html"> 
<span class="menu-text">Access</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../develop/general-intro.html" aria-current="page"> 
<span class="menu-text">Tutorials</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/hds-sandbox"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/ucph-heads/"> <i class="bi bi-linkedin" role="img" aria-label="LinkedIn">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../develop/GWAS0-geneticTheory.html">Introduction</a></li><li class="breadcrumb-item"><a href="../develop/GWAS3-QualityControlA.html">Quality Control: initial steps</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/general-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Instructions</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS0-geneticTheory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic Theory Review</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS1-WhatIsAGWAS.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">What is a GWAS?</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS2-DataCollectionA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data collection</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS2-DataCollection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data exploration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS3-QualityControlA.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Quality Control: initial steps</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Quality Control</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS4-QualityControlB.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control: Relatedness &amp; Population Stratification</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Association Testing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS5-AssociationTesting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Association testing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS5b-PopulationStratification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population stratification</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS6-PRSAnalysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polygenic scores I</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS7-PRSII.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polygenic scores II</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../develop/GWAS8-OtherToolsAndTesting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other tools &amp; conclusions</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#individual-and-snp-missingness" id="toc-individual-and-snp-missingness" class="nav-link active" data-scroll-target="#individual-and-snp-missingness">Individual and SNP Missingness</a>
  <ul class="collapse">
  <li><a href="#missingness-per-snp" id="toc-missingness-per-snp" class="nav-link" data-scroll-target="#missingness-per-snp">Missingness per SNP</a></li>
  <li><a href="#missingness-per-individual" id="toc-missingness-per-individual" class="nav-link" data-scroll-target="#missingness-per-individual">Missingness per individual</a></li>
  <li><a href="#first-run-of-the-plink-tool" id="toc-first-run-of-the-plink-tool" class="nav-link" data-scroll-target="#first-run-of-the-plink-tool">First run of the PLINK tool</a></li>
  </ul></li>
  <li><a href="#sex-discrepancy" id="toc-sex-discrepancy" class="nav-link" data-scroll-target="#sex-discrepancy">Sex Discrepancy</a>
  <ul class="collapse">
  <li><a href="#plink-commands" id="toc-plink-commands" class="nav-link" data-scroll-target="#plink-commands">PLINK Commands</a>
  <ul class="collapse">
  <li><a href="#how-can-we-proceed" id="toc-how-can-we-proceed" class="nav-link" data-scroll-target="#how-can-we-proceed">How can we proceed?</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#minor-allele-frequency-maf" id="toc-minor-allele-frequency-maf" class="nav-link" data-scroll-target="#minor-allele-frequency-maf">Minor Allele Frequency (MAF)</a>
  <ul class="collapse">
  <li><a href="#plink-commands-1" id="toc-plink-commands-1" class="nav-link" data-scroll-target="#plink-commands-1">PLINK commands</a></li>
  </ul></li>
  <li><a href="#hardy-weinberg-equilibrium-hwe" id="toc-hardy-weinberg-equilibrium-hwe" class="nav-link" data-scroll-target="#hardy-weinberg-equilibrium-hwe">Hardy-Weinberg Equilibrium (HWE)</a>
  <ul class="collapse">
  <li><a href="#plink-commands-2" id="toc-plink-commands-2" class="nav-link" data-scroll-target="#plink-commands-2">PLINK commands</a></li>
  </ul></li>
  <li><a href="#heterozygosity-rate" id="toc-heterozygosity-rate" class="nav-link" data-scroll-target="#heterozygosity-rate">Heterozygosity Rate</a>
  <ul class="collapse">
  <li><a href="#plink-commands-3" id="toc-plink-commands-3" class="nav-link" data-scroll-target="#plink-commands-3">PLINK commands</a></li>
  </ul></li>
  <li><a href="#bibliography" id="toc-bibliography" class="nav-link" data-scroll-target="#bibliography">Bibliography</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="GWAS3-QualityControlA.out.ipynb" download="GWAS3-QualityControlA.out.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../develop/GWAS0-geneticTheory.html">Introduction</a></li><li class="breadcrumb-item"><a href="../develop/GWAS3-QualityControlA.html">Quality Control: initial steps</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Quality Control: initial steps</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-note no-icon callout-titled" title="Important notes for this notebook">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important notes for this notebook
</div>
</div>
<div class="callout-body-container callout-body">
<p>After gathering your data and genotyping, there are many checks to do to determine the quality of the data. It is crucial to <strong>perform quality control before</strong> carrying out any GWAS, otherwise, there is the risk that some of the associations are spurious.</p>
<section id="learning-outcomes" class="level2">
<h2 class="anchored" data-anchor-id="learning-outcomes">Learning outcomes</h2>
<ul>
<li><strong>Distinguish</strong> the various QC steps</li>
<li><strong>Discuss and choose</strong> thresholds on plink</li>
<li><strong>Implement</strong> basic QC in <code>plink</code> and statistical plots in <code>R</code></li>
<li><strong>Hypothesize</strong> the effects of various <code>plink</code> commands and <strong>verify</strong> your hypothesis using the analysis in <code>R</code></li>
</ul>
</section>
<section id="how-to-make-this-notebook-work" class="level2">
<h2 class="anchored" data-anchor-id="how-to-make-this-notebook-work">How to make this notebook work</h2>
<ul>
<li>In this notebook, we will use both <code>R</code> and <code>bash command line</code> programming languages. Remember to <strong>change the kernel</strong> whenever you transition from one language to the other (<code>Kernel --&gt; Change Kernel</code>) indicated by the languages’ images. We will first run <code>Bash</code> commands.</li>
</ul>
<p><img src="Images/bash.png" alt="Bash" width="80"> Choose the Bash kernel</p>
</section>
</div>
</div>
<p>The next critical step in any GWAS involves <strong>scrutinizing the data for potential issues</strong>. It’s crucial to ensure that the results are not merely artifacts of poor data quality discovered after the analysis.</p>
<p>After gathering your data and genotyping, there are many checks one can do to determine the quality of the data. PLINK provides several summary statistics for quality control (e.g.&nbsp;missing genotype rate, minor allele frequency, and Hardy-Weinberg equilibrium failures) which will serve as thresholds for subsequent analyses.</p>
<p>For this tutorial, we will look at the following topics:</p>
<ol type="1">
<li>Individual Missingness<br>
</li>
<li>Sex discrepancy<br>
</li>
<li>Minor Allele Frequency (MAF)<br>
</li>
<li>Hardy-Weinberg Equilibrium (HWE)<br>
</li>
<li>Heterozygosity Rate</li>
</ol>
<section id="individual-and-snp-missingness" class="level1">
<h1>Individual and SNP Missingness</h1>
<section id="missingness-per-snp" class="level2">
<h2 class="anchored" data-anchor-id="missingness-per-snp">Missingness per SNP</h2>
<p>Overall, the SNP genotyping platform is very reliable and delivers stable results when it comes to determining genotypes. Of course, it is not flawless. One of the most frequent problems is that some of the SNPs are just not well-genotyped in the entire population. These should be removed to improve the overall data quality.</p>
<p>Certainly, we can’t remove every SNP with any missing values, as that would result in losing a significant portion of our data. Instead, we adopt some <strong>thresholds</strong>. However, determining these thresholds isn’t governed by strict rules. You have the flexibility to set them within “reasonable” limits. To find out what constitutes “reasonable,” consulting relevant literature for your species of interest is advisable. For humans, a common starting point is a threshold of 0.2, meaning we exclude SNPs where 20% or more of the genotypes are missing in our population.</p>
</section>
<section id="missingness-per-individual" class="level2">
<h2 class="anchored" data-anchor-id="missingness-per-individual">Missingness per individual</h2>
<p>The reliability of SNP chips is also high when it comes to individual genotypes. In some cases, however, some of the individuals contain a large number of missing SNPs. The reason could be low DNA sample quality, the wrong chip type used (e.g.&nbsp;cattle chip for deer samples), or other technical issues. Regardless of the reason, you should remove the worst offenders from your data set, to not compromise the overall quality of your results.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>Have a look at the following toy example:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>/</th>
<th>SNP1</th>
<th>SNP2</th>
<th>SNP3</th>
<th>SNP4</th>
<th>SNP5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>IND1</td>
<td>22</td>
<td>00</td>
<td>11</td>
<td>12</td>
<td>22</td>
</tr>
<tr class="even">
<td>IND2</td>
<td>22</td>
<td>00</td>
<td>11</td>
<td>12</td>
<td>22</td>
</tr>
<tr class="odd">
<td>IND3</td>
<td>11</td>
<td>12</td>
<td>11</td>
<td>22</td>
<td>21</td>
</tr>
<tr class="even">
<td>IND4</td>
<td>00</td>
<td>00</td>
<td>11</td>
<td>11</td>
<td>00</td>
</tr>
<tr class="odd">
<td>IND5</td>
<td>22</td>
<td>00</td>
<td>11</td>
<td>22</td>
<td>22</td>
</tr>
</tbody>
</table>
<p>Here, we have a data set of five individuals, each of them genotyped for five SNPs. The genotypes themselves are in numerical coding, 11 and 22 being the two homozygous, 12 the heterozygous, and 00 coded as missing. We want to apply two filters: one for the variants, and one for the individuals.</p>
<ul>
<li>First remove variants with &gt;= 40% missingness</li>
<li>Then remove individuals with &gt;= 40% missingness</li>
</ul>
<p>Which variants and individuals are we removing?</p>
</div>
</div>
<p>In practice, you want to <strong>remove the SNPs based on missingness before the individuals</strong>. This is simply because we generally have a lot more SNPs than individuals, and would thus lose less information by removing SNPs than individuals. This will remove “bad” SNPs first, leaving a lower rate of missingness for all the individuals.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Variant-level threshold</strong>: In this example, we would only remove SNP2 since it exceeds this threshold, with 80% missingness (4 out of 5 missing values).</p>
<p><strong>Individual-level threshold</strong>: IND4 is removed from the data set, as there is missing genotype information for 2/4 = 50% SNPs (after filtering out SNP2).</p>
</div>
</div>
</div>
</section>
<section id="first-run-of-the-plink-tool" class="level2">
<h2 class="anchored" data-anchor-id="first-run-of-the-plink-tool">First run of the PLINK tool</h2>
<p>Using PLINK, we can address missing data using two functions; one operates at the variant level and the other at the individual level:</p>
<ul>
<li><code>--geno</code>: This will remove SNPs with a specified proportion of missingness (e.g.&nbsp;<code>--geno 0.01</code> will remove SNPs with 1% missingness).</li>
<li><code>--mind</code>: This will remove individuals with a specified proportion of missingness (e.g.&nbsp;<code>--mind 0.01</code> will remove individuals with 1% missingness).</li>
</ul>
<p>A common threshold for –mind and –geno ranges from 1% to 5% to ensure quality and robustness. Higher thresholds may be acceptable if the study has a large sample size and missingness is not widespread. Those filters must run one at a time in PLINK.</p>
<p>Let’s implement our first QC method in PLINK.</p>
<p><img src="Images/bash.png" alt="Bash" width="40"> We use <code>ln -sf</code> to link the data folder and create a folder for output files (<code>Results/GWAS3</code>)</p>
<div id="5dea98b7-c4dd-4d62-a5e6-b041486b511d" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-sf</span> ../Data</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> Results/GWAS3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>PLINK program uses options implemented using the <code>--</code> symbol to perform different functions or access files. We will use the option <code>--bfile</code> for data (filename without extension), <code>--missing</code> for missing data summary, and <code>--out</code> for output prefixes. This helps us analyze data missingness.</p>
<div id="f206a476" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Data/HapMap_3_r3_1 <span class="at">--missing</span> <span class="at">--out</span> Results/GWAS3/data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to Results/GWAS3/data.log.
Options in effect:
  --bfile Data/HapMap_3_r3_1
  --missing
  --out Results/GWAS3/data

385567 MB RAM detected; reserving 192783 MB for main workspace.
1457897 variants loaded from .bim file.
165 people (80 males, 85 females) loaded from .fam.
112 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 112 founders and 53 nonfounders present.
Calculating allele frequencies... done.
Warning: 225 het. haploid genotypes present (see Results/GWAS3/data.hh ); many
commands treat these as missing.
Total genotyping rate is 0.997378.
--missing: Sample missing data report written to Results/GWAS3/data.imiss, and
variant-based missing data report written to Results/GWAS3/data.lmiss.</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>PLINK prints a lot of information. Make sure you read it.</p>
<ol type="1">
<li>Which file formats are read by PLINK to understand the data?</li>
<li>What is the percentage of valid genotype data?</li>
<li>What is the count of females and males? Is sex information available for all individuals?</li>
<li>What files are generated as output from the command?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>PLINK reads two files in the format <em>.fam</em> and <em>.bim</em> for this task. As explained in the notebook GWAS2, these files contain data for individuals and variants, respectively.</li>
<li>Immediately we can see that the total genotyping rate for our sample is 0.997378.</li>
<li>It counts 80 males and 85 females.</li>
<li>It prints the name of the two output files from this command: <code>data.imiss</code> and <code>data.lmiss</code>, in the directory <code>Results/GWAS3/</code>. These files show the proportion of <strong>missing SNPs per individual</strong> and the proportion of <strong>missing individuals per SNP</strong>, respectively.</li>
</ol>
</div>
</div>
<p>These files include a header and one line per sample or per variant, containing the following fields: * FID: Family ID * IID: Within-family ID * CHR and SNP: chromosome and SNP ID. * MISS_PHENO: Indicates if the phenotype is missing (Y/N) * N_MISS: Number of missing genotype calls (excluding obligatory missings or heterozygous haploids) * N_GENO: Number of potentially valid calls * F_MISS: Missing call rate</p>
<p>Let’s check the first 5 lines of the files to understand their structure. - <strong>Are there missing phenotypes or genotypes in the <code>.imiss</code> file?</strong></p>
<div id="e64cb902-6e5b-472c-a760-2f84a52f2d58" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-5</span> Results/GWAS3/data.imiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    FID       IID MISS_PHENO   N_MISS   N_GENO   F_MISS
   1328   NA06989          N     4203  1457897 0.002883
   1377   NA11891          N    20787  1457897  0.01426
   1349   NA11843          N     1564  1457897 0.001073
   1330   NA12341          N     6218  1457897 0.004265</code></pre>
</div>
</div>
<p>No, the first individuals of the <code>imiss</code> file do not seem to miss phenotypes (column <code>MISS_PHENO</code>) and have small proportion of missing genotypes (column <code>F_MISS</code>).</p>
<ul>
<li><strong>What about at the variant-level (<code>.lmiss</code>)?</strong></li>
</ul>
<div id="4714af05-3e55-4b63-8c0d-1416379b54ed" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-5</span> Results/GWAS3/data.lmiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> CHR         SNP   N_MISS   N_GENO   F_MISS
   1   rs2185539        0      165        0
   1  rs11510103        4      165  0.02424
   1  rs11240767        0      165        0
   1   rs3131972        0      165        0</code></pre>
</div>
</div>
<p>In the <code>lmiss</code> file, we can see that the second variant has missing data (4 missing genotype calls).</p>
<p><img src="Images/R.png" alt="R" width="40"> Switch to the R kernel.</p>
<p>We can visualize the distribution of missing data in individuals and SNPs using histograms in <code>R</code> (change kernel).</p>
<p>The <code>ggplot2</code> package is used here for visualization and various plot elements are combined with the <code>+</code> symbol. Histograms are saved in your results directory as <code>histimiss.png</code> and <code>histlmiss.png</code>. Look below at the <code>R</code> code and the output.</p>
<div id="904776e1-2135-41c4-a356-1bd0b298cf82" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">))</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">options</span><span class="er">(</span><span class="ex">repr.plot.width</span> = 9, repr.plot.height = 4<span class="kw">)</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data into R </span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="ex">indmiss</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="va">file</span><span class="op">=</span><span class="st">"Results/GWAS3/data.imiss"</span>, <span class="va">header</span><span class="op">=</span>TRUE<span class="kw">)</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="ex">snpmiss</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="va">file</span><span class="op">=</span><span class="st">"Results/GWAS3/data.lmiss"</span>, <span class="va">header</span><span class="op">=</span>TRUE<span class="kw">)</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co">#lmiss histogram</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.lmiss</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">snpmiss,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>snpmiss<span class="pp">[</span><span class="ss">,5</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">geom_histogram</span><span class="er">(</span><span class="ex">binwidth</span> = 0.005, col = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Frequency of missingness rates in SNPs"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">xlab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">ylab</span><span class="er">(</span><span class="st">"Missingness in SNPs"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co">#imiss histogram</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.imiss</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">indmiss,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>indmiss<span class="pp">[</span><span class="ss">,6</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    <span class="ex">geom_histogram</span><span class="er">(</span><span class="ex">binwidth</span> = 0.001, col = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Frequency of missingness rates in individuals"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>    <span class="ex">xlab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="ex">ylab</span><span class="er">(</span><span class="st">"Missingness in Individuals"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a><span class="co">#show histograms</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.lmiss</span><span class="kw">)</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.imiss</span><span class="kw">)</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plots silently</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="kw">{</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.lmiss, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histlmiss.png"</span><span class="kw">);</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.imiss, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histimiss.png"</span><span class="kw">);</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a><span class="kw">})</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-6-output-1.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-6-output-2.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>What is the highest percentage of missingness for a SNP?</li>
<li>Are there individuals with no missing data?</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here are a few observations we can make. Firstly, it’s clear that very few individuals have no missing data at all. One interpretation could be that one or a few SNPs are poorly genotyped across all samples. Fortunately, this isn’t the case here, otherwise the first plot would show many more SNPs with high missing frequency. As shown in the SNP-based histogram, the highest percentage of missing data for a SNP is around 0.05 (5%). Overall, the histogram provides insight into how applying a missingness filter will affect the size of your remaining data.</p>
</div>
</div>
<p><img src="Images/bash.png" alt="Bash" width="40"> Switch to the Bash kernel.</p>
<p>Next, we will use PLINK to filter the data using missingness thresholds to ensure data quality.</p>
<p>We choose a 2% missingness threshold for both individuals and SNP, as only a few outliers exceed this rate (this threshold does not need to be the same for SNPs and individuals). Below, we provide the command to remove samples (<code>--mind</code>) and SNPs (<code>--geno</code>) above this threshold. The <code>--out</code> option defines the output prefix, and <code>--make-bed</code> denotes the output format (consisting of the three files <code>.bed</code>, <code>.bim</code>, <code>.fam</code>).</p>
<p>PLINK provides detailed output during execution, so check the printed text to monitor progress.</p>
<div id="5077d592-61cd-4b33-8691-e0a77f06b326" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Delete samples with missing call freq &gt; 2% and SNPs with missingness &gt; 2%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Data/HapMap_3_r3_1 <span class="at">--mind</span> .02 <span class="at">--geno</span> .02 <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to Results/GWAS3/HapMap_3_r3_2.log.
Options in effect:
  --bfile Data/HapMap_3_r3_1
  --geno .02
  --make-bed
  --mind .02
  --out Results/GWAS3/HapMap_3_r3_2

385567 MB RAM detected; reserving 192783 MB for main workspace.
1457897 variants loaded from .bim file.
165 people (80 males, 85 females) loaded from .fam.
112 phenotype values loaded from .fam.
1 person removed due to missing genotype data (--mind).
ID written to Results/GWAS3/HapMap_3_r3_2.irem .
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 112 founders and 52 nonfounders present.
Calculating allele frequencies... done.
Warning: 225 het. haploid genotypes present (see Results/GWAS3/HapMap_3_r3_2.hh
); many commands treat these as missing.
Total genotyping rate in remaining samples is 0.997486.
26686 variants removed due to missing genotype data (--geno).
1431211 variants and 164 people pass filters and QC.
Among remaining phenotypes, 56 are cases and 56 are controls.  (52 phenotypes
are missing.)
--make-bed to Results/GWAS3/HapMap_3_r3_2.bed + Results/GWAS3/HapMap_3_r3_2.bim
+ Results/GWAS3/HapMap_3_r3_2.fam ... done.</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>What happened during filtering?</p>
<ul>
<li>How many individuals and variants were removed?</li>
<li>Can you preview the output files? Which command do you need for each file? Use it to print the first 5 lines of each file.</li>
</ul>
</div>
</div>
<div id="1b3382a6-8c0d-4a9c-b3b9-381562c620ee" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write code here - fam file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="85e9ef99-d528-4092-a78f-64716bd1d67a" class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write code here - bim file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="13a50218-0507-4c1d-877a-3f6f9b0c9870" class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write code here - bed file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Only 1 individual has been removed, while 26686 variants where above the filtering threshold.</p>
</div>
</div>
</div>
<p>We can see a preview of the output files: those are again a set of <code>fam</code>, <code>bim</code> and <code>bed</code> file including only the data passing the filters. For the first two formats we can use <code>head</code>, while for the <code>bed</code> file we need also <code>xxd</code> because of the binary nature of the file.</p>
<div id="95d017ba-4ad1-4787-b1d1-901c99a57acc" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-5</span> Results/GWAS3/HapMap_3_r3_2.fam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1328 NA06989 0 0 2 2
1377 NA11891 0 0 1 2
1349 NA11843 0 0 1 1
1330 NA12341 0 0 2 2
1344 NA10850 0 NA12058 2 -9</code></pre>
</div>
</div>
<div id="14273355-f7d3-46d7-aee3-d0e26680ac72" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-5</span> Results/GWAS3/HapMap_3_r3_2.bim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1   rs2185539   0   556738  T   C
1   rs11240767  0   718814  T   C
1   rs3131972   0   742584  A   G
1   rs3131969   0   744045  A   G
1   rs1048488   0   750775  C   T</code></pre>
</div>
</div>
<div id="ad7f12e6-a432-4a9c-a87b-5b5798584e6b" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">xxd</span> <span class="at">-b</span> Results/GWAS3/HapMap_3_r3_2.bed <span class="kw">|</span> <span class="fu">head</span> <span class="at">-5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>00000000: 01101100 00011011 00000001 11111111 11111111 11111111  l.....
00000006: 11111111 11111111 11111111 11111111 11111111 11111111  ......
0000000c: 11111111 11111111 11111111 11111111 11111111 11111111  ......
00000012: 11111111 11111111 11111111 11111111 11111111 11111111  ......
00000018: 11111111 11111111 11111111 11111111 11111111 11111111  ......</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="sex-discrepancy" class="level1">
<h1>Sex Discrepancy</h1>
<p>One useful check is verifying if the indicated sex is correct. Using PLINK, you can calculate the inbreeding coefficient on the X chromosome under the assumption that it is an autosomal chromosome. This approach is insightful because PLINK treats haploid chromosomes (like the X chromosome in males) as homozygotes due to technical reasons. Consequently, assuming the X chromosome is autosomal makes males appear highly inbred on the X, whereas females do not (since they have two X chromosomes). As a result, the inbreeding coefficient estimates will be close to 1 for males and 0 for females.</p>
<section id="plink-commands" class="level2">
<h2 class="anchored" data-anchor-id="plink-commands">PLINK Commands</h2>
<p>This sex check can be performed in PLINK using the <code>--check-sex</code> option. The results are outputted in the file with extension <code>.sexcheck</code> (in the <code>Results/GWAS4</code> folder) in which the sex is in column <code>PEDSEX</code> (1 means male and 2 means female) and the inbreeding coefficient is in column <code>F</code>).</p>
<p>Generally, males should have an X chromosome homozygosity estimate &gt;0.8 and females should have a value &lt;0.2. So we could simply remove any individuals where the homozygosity estimate (F) does not match their specified sex. Subjects who do not fulfill these requirements are flagged “PROBLEM” by PLINK in the output file.</p>
<p>Note the option <code>--silent</code> to avoid long texts printed out on the screen.</p>
<div id="bc02622c" class="cell" data-tags="[]" data-execution_count="7">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_2 <span class="at">--check-sex</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_2 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Warning: 181 het. haploid genotypes present (see Results/GWAS3/HapMap_3_r3_2.hh
); many commands treat these as missing.</code></pre>
</div>
</div>
<p><img src="Images/R.png" alt="R" width="40"> Switch to the R kernel.</p>
<p>We can visualize the results of the sex check by plotting three histograms for the F values in males, females, and all samples. Let’s look at all sexes and males first.</p>
<div id="8ca49591-e1f1-4766-9dc9-3bf66670ba92" class="cell" data-tags="[]" data-execution_count="9">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inbreeding coefficients for all sexes and for males/females separately</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">)))</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="ex">options</span><span class="er">(</span><span class="ex">repr.plot.width</span> = 9, repr.plot.height = 4<span class="kw">)</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data into R and subset for sexes</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="ex">sex</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="st">"Results/GWAS3/HapMap_3_r3_2.sexcheck"</span><span class="ex">,</span> header=T,as.is=T<span class="kw">)</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="ex">male</span> <span class="op">&lt;</span>- subset<span class="er">(</span><span class="ex">sex,</span> sex<span class="va">$PEDSEX</span>==1<span class="kw">)</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sex inbreeding coeff histogram</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.sex</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">sex,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>sex<span class="pp">[</span><span class="ss">,6</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span>, bins=50<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Values of F for all sexes"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"F"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Inbreeding coeff in males histogram</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.male</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">male,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>male<span class="pp">[</span><span class="ss">,6</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"blue"</span>, bins=50<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Values of F for males"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"F"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.sex</span><span class="kw">)</span></span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.male</span><span class="kw">)</span></span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plots </span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.sex, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histsex.png"</span><span class="kw">)</span></span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.male, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histmale.png"</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-15-output-1.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 7 in image
Saving 7 x 7 in image</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-15-output-3.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><ol type="A">
<li>Do the plots look as you would expect?</li>
</ol></li>
<li><ol start="2" type="A">
<li>Modify the code above to plot the F values for females and save the plot as <code>Results/GWAS3/histfemale.png</code></li>
</ol></li>
<li><ol start="3" type="A">
<li>What can you conclude about the distribution of inbreeding coefficients between males and females? (pay attention to the axis ranges)</li>
</ol></li>
</ul>
</div>
</div>
<div id="15e474cd-6109-47dc-8843-02e8c0219935" class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write your code here - female F values histogram </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="A">
<li><p>Yes.</p></li>
<li><p>Code for the histogram below.</p></li>
<li><p>The inbreeding coefficient plot indicates that there is one woman with a sex discrepancy (F value of 0.99). When using other datasets often a few discrepancies might be found.</p></li>
</ol>
</div>
</div>
<div id="b346f7ac-a8d2-447b-963d-6807e21a85a8" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># B) </span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># subset females </span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="ex">female</span> <span class="op">&lt;</span>- subset<span class="er">(</span><span class="ex">sex,</span> sex<span class="va">$PEDSEX</span>==2<span class="kw">)</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Inbreeding coeff in females histogram</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.female</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">female,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>female<span class="pp">[</span><span class="ss">,6</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"blue"</span>, bins=50<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Values of F for females"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"F"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Show hist</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.female</span><span class="kw">)</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plot </span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.female, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histfemale.png"</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 7 in image</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-17-output-2.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="how-can-we-proceed" class="level3">
<h3 class="anchored" data-anchor-id="how-can-we-proceed">How can we proceed?</h3>
<p>We have two options when dealing with sex discrepancy. One is to simply remove any individual with sex discrepancy. In our case, this would involve removing the female with the F value of 0.99.</p>
<p><img src="Images/bash.png" alt="bash" width="40"> Switch to the Bash kernel.</p>
<p>Removal can be done with the command below. Use <code>grep</code> to find lines in the sexcheck file that contain <code>PROBLEM</code>. Then, pipe (send) the output to the command <code>awk</code> (using the so-called pipe symbol <code>|</code>). <code>awk</code> will extract the first two columns from each line identified by the <code>grep</code> command and redirect them to the file <code>sex_discrepancy.txt</code> using the symbol <code>&gt;</code> (redirecting the output away from the screen/terminal).</p>
<div id="57f544c6" class="cell" data-tags="[]" data-execution_count="3">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"PROBLEM"</span> Results/GWAS3/HapMap_3_r3_2.sexcheck <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{print$1,$2}'</span><span class="op">&gt;</span> Results/GWAS3/sex_discrepancy.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Show the content of the file using <code>cat</code>:</p>
<div id="3e151a47-8fc2-4931-823f-40e2732b076b" class="cell" data-tags="[]" data-execution_count="4">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> Results/GWAS3/sex_discrepancy.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1349 NA10854</code></pre>
</div>
</div>
<p>For long files, you can simply count the number of lines in the file with <code>wc</code> (word count command)</p>
<div id="a9ca1c0b" class="cell" data-tags="[]" data-execution_count="5">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> Results/GWAS3/sex_discrepancy.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1 Results/GWAS3/sex_discrepancy.txt</code></pre>
</div>
</div>
<p>The file can be provided to PLINK for removal of sex-discrepant individuals using the option <code>--remove</code>:</p>
<div id="5c10a26c" class="cell" data-tags="[]" data-execution_count="6">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_2 <span class="at">--remove</span> Results/GWAS3/sex_discrepancy.txt <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_3 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Warning: 181 het. haploid genotypes present (see Results/GWAS3/HapMap_3_r3_3.hh
); many commands treat these as missing.</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Alternative approach">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Alternative approach
</div>
</div>
<div class="callout-body-container callout-body">
<p>The other approach supported by PLINK is to impute the sex codes based on the SNP data, which is done with the command <code>--impute-sex</code> as shown below.</p>
<pre class="{bash}"><code>plink --bfile Results/GWAS3/HapMap_3_r3_2 --impute-sex --make-bed --out Results/GWAS3/HapMap_3_r3_3_alt
</code></pre>
<p>PLINK automatically detects and imputes incorrect sex information, requiring just an additional option to be specified. However, we will not be executing this command. Instead, we will retain only the dataset where incorrect sexes have already been filtered out: <code>Results/GWAS3/HapMap_3_r3_3</code>.</p>
</div>
</div>
</section>
</section>
</section>
<section id="minor-allele-frequency-maf" class="level1">
<h1>Minor Allele Frequency (MAF)</h1>
<p>Excluding SNPs based on minor allele frequency (MAF) is somewhat controversial. In a sense, it has little to do with quality control – there is no reason to think there are any errors in the data. The main justification is statistical:</p>
<ul>
<li>If MAF is low, and the SNP is rare, then the power is low (i.e.&nbsp;don’t spend multiple testing corrections on tests that are unlikely to find anything anyway).</li>
<li>Some statistical methods perform badly with low MAF (e.g.&nbsp;the chi-squared-test).</li>
</ul>
<p>An appropriate cutoff definitely depends on sample size – the larger the sample, the greater your ability to include rare SNPs. Typically, researchers utilize thresholds of 0.1 or 0.05 (<span class="citation" data-cites="kanaka_concepts_2023">Kanaka et al. (<a href="#ref-kanaka_concepts_2023" role="doc-biblioref">2023</a>)</span>).</p>
<section id="plink-commands-1" class="level2">
<h2 class="anchored" data-anchor-id="plink-commands-1">PLINK commands</h2>
<p>Filtering data based on the minor allele frequencies is done in a similar way to previous commands. If you want to get rid only of the fixed SNPs, you specify a MAF threshold of 0, which can be done by the command <code>--maf 0</code></p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>One should <strong>limit MAF analysis to only autosomal chromosomes</strong>, meaning you need to generate a subset of the data containing only autosomal chromosomes as done below.</p>
</div>
</div>
<p>First, we extract the SNP identifiers from chromosomes 1 to 22 with <code>awk</code>, and redirect the output to the file <code>snp_1_22.txt</code>. Then, PLINK can extract those SNPs using the option <code>--extract</code>:</p>
<div id="80174ebb" class="cell" data-tags="[]" data-execution_count="8">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get a file with autosomal variants</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{ if ($1 &gt;= 1 &amp;&amp; $1 &lt;= 22) print $2 }'</span> Results/GWAS3/HapMap_3_r3_3.bim <span class="op">&gt;</span> Results/GWAS3/snp_1_22.txt</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter data based on the list of SNPs</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_3 <span class="at">--extract</span> Results/GWAS3/snp_1_22.txt <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_4 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>How can you verify if only the autosomal regions are present in the output? Retrieve a list of chromosomes present in one of the files with the prefix <code>Results/GWAS3/HapMap_3_r3_4</code>.</p>
</div>
</div>
<div id="90aaf03e-e293-48ee-8a48-e9915fb6bc30" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># write your code</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>SNPs are saved in the output <code>bim</code> file and the first column contains the chromosome names. So, we just need to extract the first column, and check the unique names (as extra, we count how many times each chromosome occurs):</p>
</div>
</div>
</div>
<div id="9bd82caf-88b2-4140-b302-bf3dce3ae505" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 Results/GWAS3/HapMap_3_r3_4.bim <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 117486 1
 117527 2
  97357 3
  86412 4
  88839 5
  92059 6
  75941 7
  75794 8
  64194 9
  74392 10
  71825 11
  69411 12
  52348 13
  45631 14
  42297 15
  45099 16
  38736 17
  41147 18
  26439 19
  36631 20
  19427 21
  20314 22</code></pre>
</div>
</div>
<hr>
<p>Now that we have a set of files containing only autosomal chromosomes, let’s obtain summary statistics for the minor allele frequency and plot the values on a histogram. This can be done using the <code>--freq</code> option in PLINK.</p>
<p><strong>Try to write the PLINK command below before reviewing the command provided.</strong></p>
<div id="f0316fff-9e1d-46d7-9f64-01fa8b2e083f" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Your PLINK command</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3

  plink &lt;input flag(s)...&gt; [command flag(s)...] [other flag(s)...]
  plink --help [flag name(s)...]

Commands include --make-bed, --recode, --flip-scan, --merge-list,
--write-snplist, --list-duplicate-vars, --freqx, --missing, --test-mishap,
--hardy, --mendel, --ibc, --impute-sex, --indep-pairphase, --r2, --show-tags,
--blocks, --distance, --genome, --homozyg, --make-rel, --make-grm-gz,
--rel-cutoff, --cluster, --pca, --neighbour, --ibs-test, --regress-distance,
--model, --bd, --gxe, --logistic, --dosage, --lasso, --test-missing,
--make-perm-pheno, --tdt, --qfam, --annotate, --clump, --gene-report,
--meta-analysis, --epistasis, --fast-epistasis, and --score.

"plink --help | more" describes all functions (warning: long).</code></pre>
</div>
</div>
<div id="d7fe20f4" class="cell" data-tags="[]" data-execution_count="12">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a plot of the MAF distribution.</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_4 <span class="at">--freq</span> <span class="at">--out</span> Results/GWAS3/MAF_check <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="Images/R.png" alt="R" width="40"> Switch to the R kernel.</p>
<p>Let’s plot the MAF histogram using <code>R</code>. We will have a second plot, which is a zoom-in version with MAF up to 0.1 to better observe the low MAF values. Note: The zoom-in version will throw some warnings as data points are removed, but don’t worry about it.</p>
<div id="c8e23649-69c5-4bc1-a4c3-5f47c113f947" class="cell" data-tags="[]" data-execution_count="16">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MAF plot only for autosomal SNPs. Note the zoomed interval (0,0.1)</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">)))</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="ex">options</span><span class="er">(</span><span class="ex">repr.plot.width</span> = 9, repr.plot.height = 4<span class="kw">)</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data into R </span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="ex">maf_freq</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="st">"Results/GWAS3/MAF_check.frq"</span><span class="ex">,</span> header =TRUE, as.is=T<span class="kw">)</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="ex">summary</span><span class="er">(</span><span class="ex">maf_freq</span><span class="va">$MAF</span><span class="kw">)</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="co"># maf_freq histogram</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.maf</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">maf_freq,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>maf_freq<span class="pp">[</span><span class="ss">,5</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span>, bins=50<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"MAF distribution"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"MAF"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ylab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a><span class="co"># zoom-in into X-axis</span></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.maf.zoom</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">maf_freq,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>maf_freq<span class="pp">[</span><span class="ss">,5</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span>, bins = 10<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Zoom into MAF distribution up to value 0.1"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"MAF"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ylab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlim</span><span class="er">(</span><span class="ex">0,</span> 0.1<span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.maf</span><span class="kw">)</span></span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.maf.zoom</span><span class="kw">)</span></span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plot</span></span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.maf, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histmaf.png"</span><span class="kw">))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00000 0.05856 0.18750 0.20315 0.33480 0.50000 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Removed 929934 rows containing non-finite outside the scale range
(`stat_bin()`).”
Warning message:
“Removed 2 rows containing missing values or values outside the scale range
(`geom_bar()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-27-output-3.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-27-output-4.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-27-output-5.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><img src="Images/bash.png" alt="bash" width="40"> Switch to the Bash kernel.</p>
<p>As stated above, your MAF threshold depends on sample size, though a conventional MAF threshold for a regular GWAS is between 0.01 and 0.05. Here, to ensure the inclusion of only SNPs we will apply a MAF threshold of 0.05 (that is, remove SNPs where the MAF is 5% or less). The threshold is given with the option <code>--maf</code>:</p>
<div id="465be532" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove SNPs with a low MAF frequency.</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_4 <span class="at">--maf</span> 0.05 <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_5 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>Determine the number of SNPs remaining after applying quality control measures.</strong> Hint: Check the <code>.bim</code> file from the output with the prefix <code>Results/GWAS3/HapMap_3_r3_5</code>.</li>
<li><strong>Repeat the process for a MAF threshold of 0.01. How many additional SNPs would be retrieved?</strong> Do not overwrite the output from the command above! Please, rename the output prefix (e.g.&nbsp;<code>Results/GWAS3/HapMap_3_r3_5_alt</code>)</li>
</ul>
</div>
</div>
<div id="c06831c1-f8b7-4461-a259-e79b621ea3a1" class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write your code here - maf &gt; 0.05</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2c7d1f11-6365-44d1-b78e-ff12ec48abdf" class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write your code here - maf &gt; 0.01</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Check the number of lines of the <code>.bim</code> file using <code>wc -l</code> and modify the plink command from above:</p>
</div>
</div>
<div id="c3466a68-a536-4265-b80a-9c9474424410" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> Results/GWAS3/HapMap_3_r3_5.bim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1073788 Results/GWAS3/HapMap_3_r3_5.bim</code></pre>
</div>
</div>
<div id="25787f46-8b93-4fd3-ac89-c234362eefbd" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plink filtering step</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_4 <span class="at">--maf</span> 0.01 <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_5_alt <span class="at">--silent</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co"># get # variants</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> Results/GWAS3/HapMap_3_r3_5_alt.bim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1182154 Results/GWAS3/HapMap_3_r3_5_alt.bim</code></pre>
</div>
</div>
<p>There would be 1,073,788 variants left with the more lenient cutoff and 1,182,154 with the stringent one. The different is ~100,000 SNPS</p>
<hr>
<p>It is worth noting that no matter what the sample size is, monomorphic SNPs (i.e., SNPs that show no genetic variation whatsoever in the sample) are usually problematic and should always be removed. Some code crashes when monomorphic SNPs are included; even if this wasn’t the case, these SNPs cannot possibly be informative in a genome-wide association study.</p>
</section>
</section>
<section id="hardy-weinberg-equilibrium-hwe" class="level1">
<h1>Hardy-Weinberg Equilibrium (HWE)</h1>
<p>The Hardy-Weinberg rule from population genetics states that genetic variation (thus, allele and genotype frequencies) in a population will remain constant unless certain disturbing factors are introduced. his implies that once we know the allele frequencies for <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span>, the genotype frequencies will be defined as <span class="math inline">\(p^2\)</span>, <span class="math inline">\(2pq\)</span>, and <span class="math inline">\(q^2\)</span>.</p>
<p>Let’s say the frequency of allele A (<span class="math inline">\(p\)</span> in the equation) is 0.4, and that of allele B (<span class="math inline">\(q\)</span>) is 0.6. This means for the H-W scenario the genotype frequencies will be 0.16 for AA, 0.48 for AB, and 0.36 for BB. In a population of e.g.&nbsp;1000 individuals with the mentioned allele frequencies we expect to see 160 AA, 480 AB, and 360 BB individuals. Of course, we rarely see exact H-W distributions in real populations. The question then becomes, what is the difference between the expected H-W and observed proportions? There are typically two reasons why a SNP is not in HWE:</p>
<ul>
<li>Genotyping error for this SNP</li>
<li>Mating is not random</li>
</ul>
<p>In reality, mating is not random, which complicates excluding SNPs based on HWE. It is generally recommended to exclude an SNP only when HWE is significantly violated, such as when the p-value for a statistical test (e.g., assessing whether the data follow a binomial distribution) is extremely low, like p-value p&lt;10e−10</p>
<div class="callout callout-style-default callout-note callout-titled" title="HWE and binomial distribution">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
HWE and binomial distribution
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Why is HWE connected to the binomial distribution? What is the ground theory behind HWE?</strong></p>
<p>You can find a clear explanation of HWE at (<span class="citation" data-cites="lachance_hardyweinberg_2016">Lachance (<a href="#ref-lachance_hardyweinberg_2016" role="doc-biblioref">2016</a>)</span>).</p>
</div>
</div>
<section id="plink-commands-2" class="level2">
<h2 class="anchored" data-anchor-id="plink-commands-2">PLINK commands</h2>
<p>We can use the option <code>--hardy</code> in PLINK to generate H-W p-values (as well as observed and expected heterozygosity). Then, use <code>awk</code> to select SNPs (column 9 of the file) with HWE p-value &lt; 0.0001, indicating a strong deviation from HWE.</p>
<div id="67839627" class="cell" data-scrolled="true" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_5 <span class="at">--hardy</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_5 <span class="at">--silent</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{ if ($9 &lt;0.00001) print $0 }'</span> Results/GWAS3/HapMap_3_r3_5.hwe <span class="op">&gt;</span> Results/GWAS3/HapMap_3_r3_5.deviating.hwe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>What can we conclude about the data from the generated files?</strong></p>
<p><img src="Images/R.png" alt="R" width="40"> Switch to the R kernel.</p>
<p>Let’s plot a histogram of the HWE p-values and zoom in for the deviating p-values.To do this, we need to read the PLINK output and examine the table with the low HWE p-values. The header of the PLINK output includes the following fields:</p>
<ul>
<li>CHR: Chromosome number.</li>
<li>SNP: SNP identifier (rsID).</li>
<li>TEST: Type of HWE test performed. This will usually be “UNAFF” for the test on controls if your data has case/control status.</li>
<li>A1: First allele (reference allele).</li>
<li>A2: Second allele (alternate allele).</li>
<li>GENO: Genotype counts in the format “HOM1/HET/HOM2”, where HOM1 is the count of homozygous for the first allele, HET is the count of heterozygous, and HOM2 is the - count of homozygous for the second allele.</li>
<li>O(HET): Observed heterozygote frequency.</li>
<li>E(HET): Expected heterozygote frequency.</li>
<li>P: Hardy-Weinberg equilibrium exact test p-value.</li>
</ul>
<p>We will modify the values in the <code>TEST</code> column to have more readable names for the plots and save the updated values in the <code>Phenotype</code> column.</p>
<div id="abe983c3-2cde-453c-9f32-118e2fcca040" class="cell" data-tags="[]" data-execution_count="3">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HWE tables</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">dplyr</span><span class="kw">)))</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data into R using dplyr library </span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="ex">hwe</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="va">file</span><span class="op">=</span><span class="st">"Results/GWAS3/HapMap_3_r3_5.hwe"</span>, <span class="va">header</span><span class="op">=</span>TRUE<span class="kw">)</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="ex">hwe_zoom</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="va">file</span><span class="op">=</span><span class="st">"Results/GWAS3/HapMap_3_r3_5.deviating.hwe"</span>, <span class="va">header</span><span class="op">=</span>FALSE<span class="kw">)</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column names </span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="ex">colnames</span><span class="er">(</span><span class="ex">hwe_zoom</span><span class="kw">)</span> <span class="op">&lt;</span>- <span class="ex">colnames</span><span class="er">(</span><span class="ex">hwe</span><span class="kw">)</span></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Modify colnames </span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="ex">hwe</span><span class="va">$Phenotype</span> <span class="op">&lt;</span>- recode<span class="er">(</span><span class="ex">hwe</span><span class="va">$TEST</span><span class="ex">,</span> <span class="st">"ALL"</span>=<span class="st">"All"</span>, <span class="st">"UNAFF"</span>=<span class="st">"Control"</span>, <span class="st">"AFF"</span>=<span class="st">"Non-Control"</span><span class="kw">)</span></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="ex">hwe_zoom</span><span class="va">$Phenotype</span> <span class="op">&lt;</span>- recode<span class="er">(</span><span class="ex">hwe_zoom</span><span class="va">$TEST</span><span class="ex">,</span> <span class="st">"ALL"</span>=<span class="st">"All"</span>, <span class="st">"UNAFF"</span>=<span class="st">"Control"</span>, <span class="st">"AFF"</span>=<span class="st">"Non-Control"</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we print the first rows of the two generated tables</p>
<div id="74418d2d-22de-4162-8144-5e5360f211be" class="cell" data-tags="[]" data-execution_count="4">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span><span class="er">(</span><span class="ex">hwe</span><span class="kw">)</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span><span class="er">(</span><span class="ex">hwe_zoom</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 10</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">CHR</th>
<th data-quarto-table-cell-role="th" scope="col">SNP</th>
<th data-quarto-table-cell-role="th" scope="col">TEST</th>
<th data-quarto-table-cell-role="th" scope="col">A1</th>
<th data-quarto-table-cell-role="th" scope="col">A2</th>
<th data-quarto-table-cell-role="th" scope="col">GENO</th>
<th data-quarto-table-cell-role="th" scope="col">O.HET.</th>
<th data-quarto-table-cell-role="th" scope="col">E.HET.</th>
<th data-quarto-table-cell-role="th" scope="col">P</th>
<th data-quarto-table-cell-role="th" scope="col">Phenotype</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>1</td>
<td>rs3131972</td>
<td>ALL</td>
<td>A</td>
<td>G</td>
<td>2/33/77</td>
<td>0.2946</td>
<td>0.2758</td>
<td>0.7324</td>
<td>All</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>1</td>
<td>rs3131972</td>
<td>AFF</td>
<td>A</td>
<td>G</td>
<td>1/19/36</td>
<td>0.3393</td>
<td>0.3047</td>
<td>0.6670</td>
<td>Non-Control</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>1</td>
<td>rs3131972</td>
<td>UNAFF</td>
<td>A</td>
<td>G</td>
<td>1/14/41</td>
<td>0.2500</td>
<td>0.2449</td>
<td>1.0000</td>
<td>Control</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>1</td>
<td>rs3131969</td>
<td>ALL</td>
<td>A</td>
<td>G</td>
<td>2/26/84</td>
<td>0.2321</td>
<td>0.2320</td>
<td>1.0000</td>
<td>All</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>1</td>
<td>rs3131969</td>
<td>AFF</td>
<td>A</td>
<td>G</td>
<td>1/17/38</td>
<td>0.3036</td>
<td>0.2817</td>
<td>1.0000</td>
<td>Non-Control</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>1</td>
<td>rs3131969</td>
<td>UNAFF</td>
<td>A</td>
<td>G</td>
<td>1/9/46</td>
<td>0.1607</td>
<td>0.1771</td>
<td>0.4189</td>
<td>Control</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 10</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">CHR</th>
<th data-quarto-table-cell-role="th" scope="col">SNP</th>
<th data-quarto-table-cell-role="th" scope="col">TEST</th>
<th data-quarto-table-cell-role="th" scope="col">A1</th>
<th data-quarto-table-cell-role="th" scope="col">A2</th>
<th data-quarto-table-cell-role="th" scope="col">GENO</th>
<th data-quarto-table-cell-role="th" scope="col">O.HET.</th>
<th data-quarto-table-cell-role="th" scope="col">E.HET.</th>
<th data-quarto-table-cell-role="th" scope="col">P</th>
<th data-quarto-table-cell-role="th" scope="col">Phenotype</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>3</td>
<td>rs7623291</td>
<td>ALL</td>
<td>T</td>
<td>C</td>
<td>22/28/62</td>
<td>0.2500</td>
<td>0.4362</td>
<td>8.938e-06</td>
<td>All</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>7</td>
<td>rs34238522</td>
<td>ALL</td>
<td>C</td>
<td>T</td>
<td>0/64/48</td>
<td>0.5714</td>
<td>0.4082</td>
<td>3.515e-06</td>
<td>All</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>8</td>
<td>rs3102841</td>
<td>ALL</td>
<td>C</td>
<td>A</td>
<td>8/78/23</td>
<td>0.7156</td>
<td>0.4905</td>
<td>1.899e-06</td>
<td>All</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>9</td>
<td>rs354831</td>
<td>ALL</td>
<td>C</td>
<td>T</td>
<td>12/18/82</td>
<td>0.1607</td>
<td>0.3047</td>
<td>6.339e-06</td>
<td>All</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>9</td>
<td>rs10990625</td>
<td>ALL</td>
<td>C</td>
<td>T</td>
<td>23/28/61</td>
<td>0.2500</td>
<td>0.4424</td>
<td>9.391e-06</td>
<td>All</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>9</td>
<td>rs10990625</td>
<td>AFF</td>
<td>C</td>
<td>T</td>
<td>15/8/33</td>
<td>0.1429</td>
<td>0.4483</td>
<td>3.574e-07</td>
<td>Non-Control</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, we can plot the histograms. We isolate each “phenotype” to see if there are any significant differences in p-value distribution between them.</p>
<div id="48e8396f-5b57-472c-b347-2b3a7879c277" class="cell" data-scrolled="true" data-tags="[]" data-execution_count="11">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HWE p-values calculated with PLINK and zoom for the SNPs deviating from HWE. We can spot some deviations from HWE in the zoomed plot. Note that the p+values for the phenotype `All` is not the merging of the barplots from the other two phenotypes!</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">)))</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="ex">options</span><span class="er">(</span><span class="ex">repr.plot.width</span> = 9, repr.plot.height = 4<span class="kw">)</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="co"># maf_freq histogram</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.hwe</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">hwe,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>hwe<span class="pp">[</span><span class="ss">,9</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="fu">col</span> = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span>, bins=50<span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"HWE distribution"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"HWE p-value"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ylab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">facet_wrap</span><span class="er">(</span><span class="ex">~Phenotype</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span> </span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">)</span><span class="ex">,strip.text</span> = element_text<span class="er">(</span><span class="fu">size</span> = 14<span class="kw">))</span></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a><span class="co"># maf_freq histogram</span></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.hwe_below_threshold</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">hwe_zoom,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>hwe_zoom<span class="pp">[</span><span class="ss">,9</span><span class="pp">]</span><span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="ex">binwidth</span> = 0.0000015, col = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"HWE distribution for strongly deviating SNPs only"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"HWE p-value"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ylab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">facet_wrap</span><span class="er">(</span><span class="ex">~Phenotype</span><span class="kw">)</span> <span class="ex">+</span></span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>10.5<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">)</span><span class="ex">,strip.text</span> = element_text<span class="er">(</span><span class="fu">size</span> = 14<span class="kw">))</span></span>
<span id="cb55-24"><a href="#cb55-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-25"><a href="#cb55-25" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.hwe</span><span class="kw">)</span></span>
<span id="cb55-26"><a href="#cb55-26" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.hwe_below_threshold</span><span class="kw">)</span></span>
<span id="cb55-27"><a href="#cb55-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-28"><a href="#cb55-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plots </span></span>
<span id="cb55-29"><a href="#cb55-29" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="kw">{</span></span>
<span id="cb55-30"><a href="#cb55-30" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.hwe, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histhwe.png"</span><span class="kw">);</span></span>
<span id="cb55-31"><a href="#cb55-31" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.hwe_below_threshold, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/histhwe_below_threshold.png"</span><span class="kw">);})</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-36-output-1.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-36-output-2.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><img src="Images/bash.png" alt="bash" width="40"> Switch to the Bash kernel.</p>
<p>Almost all of our SNPs are in HWE, as determined by testing whether the alleles follow a binomial distribution. However, in our zoomed-in plot, we can spot a few extreme deviations. We can remove these outliers using the option <code>--hwe</code> in PLINK with the specified threshold. Note that <code>--hwe</code> filters only on controls, so if you want to apply the threshold to non-controls as well, add the <code>include-nonctrl</code> in the command. In the first step below, we filter controls using a threshold of <code>1e-6</code>, followed by filtering all phenotypes with a threshold <code>1e-10</code>:</p>
<div id="f64ffce1" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter applied to controls </span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_5 <span class="at">--hwe</span> 1e-6 <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_5_onlycontrols <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5151da4b" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># threshold apply to all ind</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_5_onlycontrols <span class="at">--hwe</span> 1e-10 include-nonctrl <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_6 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="heterozygosity-rate" class="level1">
<h1>Heterozygosity Rate</h1>
<p>We filter individuals based on their heterozygosity rate, which is similar to HWE but applied at the individual level. If an individual has an unusually high number of heterozygous calls (A/B) and no homozygous calls (A/A or B/B), or vice versa, it could indicate data issues.</p>
<section id="plink-commands-3" class="level2">
<h2 class="anchored" data-anchor-id="plink-commands-3">PLINK commands</h2>
<p>PLINK does not provide a summary of individual heterozygosity. To assess this, we first prune the dataset by excluding highly correlated SNPs, as they reduce analysis power. We use <code>--indep-pairwise &lt;window size&gt;['kb'] &lt;step size (variant ct)&gt; &lt;r^2 threshold&gt;</code> to prune SNPs, excluding high inversion regions (<code>--exclude</code> with <code>Data/inversion.txt</code>). For more information, click <a href="https://www.cog-genomics.org/plink/1.9/ld">here</a>.</p>
<p>In our case, the parameters for the <code>--indep-pairwise</code> are <code>50 5 0.2</code> representing the window size, shift-window step count, and correlation threshold for pruning. Once we have the pruned list, we will check heterozygosity on the <strong>set of SNPs that are not highly correlated</strong>.</p>
<div id="6cb3ed80" class="cell" data-tags="[]" data-execution_count="3">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_6 <span class="at">--exclude</span> Data/inversion.txt <span class="at">--range</span> <span class="at">--indep-pairwise</span> 50 5 0.2 <span class="at">--out</span> Results/GWAS3/indepSNP <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With this pruned list, we measure the heterozygosity rates of the individuals in the remaining independent SNPs.</p>
<div id="a551b42f" class="cell" data-tags="[]" data-execution_count="4">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_6 <span class="at">--extract</span> Results/GWAS3/indepSNP.prune.in <span class="at">--het</span> <span class="at">--out</span> Results/GWAS3/R_check <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="Images/R.png" alt="R" width="40"> Switch to the R kernel.</p>
<p>Now, let’s plot the distribution of heterozygosity rates. Again, we will first examine the heterozygosity table. The <code>--het</code> option in PLINK generates an output table with the following columns:</p>
<ul>
<li>FID: Individual’s Family ID.</li>
<li>IID: Individual ID within the family.</li>
<li>O(HOM): Observed number of homozygous genotypes.</li>
<li>E(HOM): Expected number of homozygous genotypes under Hardy-Weinberg equilibrium.</li>
<li>N(NM): Number of non-missing autosomal genotypes.</li>
<li>F: Inbreeding coefficient (F), which is calculated as <span class="math inline">\(\tfrac{(E(HOM)−O(HOM))}{N(NM)}\)</span>.</li>
</ul>
<div id="1c961837-0788-43d5-a5b0-6f65e147f7a6" class="cell" data-tags="[]" data-execution_count="1">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="ex">het</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="st">"Results/GWAS3/R_check.het"</span><span class="ex">,</span> head=TRUE<span class="kw">)</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span><span class="er">(</span><span class="ex">het</span><span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 6</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">FID</th>
<th data-quarto-table-cell-role="th" scope="col">IID</th>
<th data-quarto-table-cell-role="th" scope="col">O.HOM.</th>
<th data-quarto-table-cell-role="th" scope="col">E.HOM.</th>
<th data-quarto-table-cell-role="th" scope="col">N.NM.</th>
<th data-quarto-table-cell-role="th" scope="col">F</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>1328</td>
<td>NA06989</td>
<td>66999</td>
<td>67450</td>
<td>103892</td>
<td>-0.0125100</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>1377</td>
<td>NA11891</td>
<td>66840</td>
<td>66820</td>
<td>102941</td>
<td>0.0005092</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>1349</td>
<td>NA11843</td>
<td>67243</td>
<td>67550</td>
<td>104055</td>
<td>-0.0085140</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>1330</td>
<td>NA12341</td>
<td>66648</td>
<td>67390</td>
<td>103809</td>
<td>-0.0203700</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>1344</td>
<td>NA10850</td>
<td>67411</td>
<td>67510</td>
<td>103988</td>
<td>-0.0028170</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>1328</td>
<td>NA06984</td>
<td>66938</td>
<td>67230</td>
<td>103577</td>
<td>-0.0081300</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will calculate the heterozygosity rates manually for plotting, as they are not included by default in the table. The formula is <span class="math inline">\(N(NM)-O(HOM)/N(NM)\)</span></p>
<div id="80d47e15-ca66-44b2-b27a-da477b9d5648" class="cell" data-tags="[]" data-execution_count="3">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot for heterozigosity rates</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">)))</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="ex">options</span><span class="er">(</span><span class="ex">repr.plot.width</span> = 9, repr.plot.height = 4<span class="kw">)</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="ex">het</span><span class="va">$HET_RATE</span> = <span class="er">(</span><span class="ex">het</span><span class="st">$"N.NM."</span> <span class="at">-</span> het<span class="st">$"O.HOM."</span><span class="kw">)</span><span class="ex">/het</span><span class="st">$"N.NM."</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plink.imiss histogram</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="ex">hist.het</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">het,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>HET_RATE<span class="kw">))</span> <span class="ex">+</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">geom_histogram</span><span class="er">(</span><span class="ex">binwidth</span> = 0.001, col = <span class="st">"black"</span>, fill=<span class="st">"tomato"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">labs</span><span class="er">(</span><span class="ex">title</span> = <span class="st">"Heterozygosity Rates"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">xlab</span><span class="er">(</span><span class="st">"Heterozygosity Rate"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">ylab</span><span class="er">(</span><span class="st">"Frequency"</span><span class="kw">)</span> <span class="ex">+</span> </span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw()</span> <span class="ex">+</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>  <span class="ex">theme</span><span class="er">(</span><span class="ex">axis.title=element_text</span><span class="er">(</span><span class="va">size</span><span class="op">=</span>13<span class="kw">)</span><span class="ex">,</span> axis.text=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>10.5<span class="kw">)</span><span class="ex">,</span> plot.title=element_text<span class="er">(</span><span class="va">size</span><span class="op">=</span>15<span class="kw">))</span></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a><span class="ex">show</span><span class="er">(</span><span class="ex">hist.het</span><span class="kw">)</span></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plots </span></span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">ggsave</span><span class="er">(</span><span class="va">plot</span><span class="op">=</span>hist.het, <span class="va">filename</span><span class="op">=</span><span class="st">"Results/GWAS3/heterozygosity.png"</span><span class="kw">))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="GWAS3-QualityControlA_files/figure-html/cell-42-output-1.png" width="540" height="240" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As a rule of thumb, <strong>we remove individuals whose heterozygosity deviates (sd) more than 3 standard deviations from the mean</strong>. We will create a file that subsets the heterozygosity table based on this threshold for use in PLINK removal. Additionally, we’ll add a <code>HET_DST</code> column to the table containing the standardized heterozygosity values.</p>
<div id="963e0a29" class="cell" data-tags="[]" data-execution_count="6">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="ex">suppressMessages</span><span class="er">(</span><span class="ex">suppressWarnings</span><span class="er">(</span><span class="ex">library</span><span class="er">(</span><span class="ex">dplyr</span><span class="kw">)))</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read file </span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="ex">het</span> <span class="op">&lt;</span>- read.table<span class="er">(</span><span class="st">"Results/GWAS3/R_check.het"</span><span class="ex">,</span> head=TRUE<span class="kw">)</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add column with heterozygocity rates</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="ex">het</span><span class="va">$HET_RATE</span> = <span class="er">(</span><span class="ex">het</span><span class="st">$"N.NM."</span> <span class="at">-</span> het<span class="st">$"O.HOM."</span><span class="kw">)</span><span class="ex">/het</span><span class="st">$"N.NM."</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Substract inds that deviate more than sd=3 from the mean </span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a><span class="ex">het_fail</span> <span class="op">&lt;</span>- subset<span class="er">(</span><span class="ex">het,</span> <span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span> <span class="op">&lt;</span> mean<span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">)</span><span class="ex">-3*sd</span><span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">))</span> <span class="kw">|</span> </span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>                    <span class="kw">(</span><span class="ex">het</span><span class="va">$HET_RATE</span> <span class="op">&gt;</span> mean<span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">)</span><span class="ex">+3*sd</span><span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">)))</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column with standardized heterozygosity rates</span></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a><span class="ex">het_fail</span><span class="va">$HET_DST</span> <span class="op">&lt;</span>- <span class="er">(</span><span class="ex">het_fail</span><span class="va">$HET_RATE</span><span class="ex">-mean</span><span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">))</span><span class="ex">/sd</span><span class="er">(</span><span class="ex">het</span><span class="va">$HET_RATE</span><span class="kw">)</span></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a><span class="co"># write file </span></span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a><span class="ex">write.table</span><span class="er">(</span><span class="ex">het_fail,</span> <span class="st">"Results/GWAS3/fail-het-qc.txt"</span>, row.names=FALSE, quote=FALSE<span class="kw">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Stop - Read - Solve">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stop - Read - Solve
</div>
</div>
<div class="callout-body-container callout-body">
<p>The output of the R code is saved as <code>fail-het-qc.txt</code>. Let’s check how many individuals are picked up as having a heterozygosity rate deviating more than 3 SDs from the mean.</p>
<p>Hint: <code>wc -l</code> counts only the lines in the file (incl.&nbsp;header).</p>
<p><img src="Images/bash.png" alt="bash" width="40"> Switch to the Bash kernel.</p>
</div>
</div>
<div id="e63bbd7f-001e-47d3-a888-c57aac1ac518" class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Write your code here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We can count 3 lines (one is the header), meaning there are two individuals to be filtered out. To use the file in PLINK, we need only the first two columns (Family ID and Individual ID) as shown below</p>
<p><code>wc -l Results/GWAS3/fail-het-qc.txt</code></p>
</div>
</div>
</div>
<p>We need the first two columns to create the list of individuals to remove.</p>
<div id="ffe3b148-7c59-4f7b-ad72-87e44359e73a" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> Results/GWAS3/fail-het-qc.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FID IID O.HOM. E.HOM. N.NM. F HET_RATE HET_DST
1330 NA12342 68056 67220 103550 0.0229 0.34277160791888 -3.74231467217876
1459 NA12874 68786 67550 104050 0.03379 0.338913983661701 -5.03987173113217</code></pre>
</div>
</div>
<p>We use <code>awk</code> to print out the first two columns in the file <code>het-fail-ind.txt</code> and the option <code>--remove</code> to filter out the individuals using PLINK.</p>
<div id="d35701f1" class="cell" data-tags="[]" data-execution_count="6">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{print$1, $2}'</span> Results/GWAS3/fail-het-qc.txt <span class="op">&gt;</span> Results/GWAS3/het-fail-ind.txt</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--bfile</span> Results/GWAS3/HapMap_3_r3_6 <span class="at">--remove</span> Results/GWAS3/het-fail-ind.txt <span class="at">--make-bed</span> <span class="at">--out</span> Results/GWAS3/HapMap_3_r3_7 <span class="at">--silent</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After all filtering is done, we are left with 161 individuals and a total of 1073788 SNPs</p>
<div id="9ad08550-3bd3-4048-993b-48ca21b1502e" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use `wc -l` command to verify the number of lines in the `bim` and `.fam` files. </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1073788 Results/GWAS3/HapMap_3_r3_7.bim</code></pre>
</div>
</div>
<p>Look back at the very first run of PLINK. We had 1457897 SNPs and 165 individuals. We filtered out around 1/3 of all SNPs!</p>
<div class="callout callout-style-default callout-important callout-titled" title="Challenge yourself: first quality control on mice data">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Challenge yourself: first quality control on mice data
</div>
</div>
<div class="callout-body-container callout-body">
<p>We continue working with the mice data. Now we want to do a quality control workflow as we did before. Try to repeat all the steps, excluding the sex discrepancies, which cannot be done due to the missing sex chromosomes in the data:</p>
<p>1 - Missingness analysis and filtering, both per SNP and per individual.</p>
<p>2 - Minor Allele Frequency. <strong>Note:</strong> The number of chromosomes is different in this dataset.</p>
<p>3 - Hardy-Weinberg Equilibrium. <strong>Note:</strong> This data contains case-control, so you don’t need to rename the columns <code>hwe$Phenotype</code> and <code>hwe_zoom$Phenotype</code>. Similarly, avoid using <code>include-nonctrl</code> in PLINK. In the R code, you will also want to avoid the line with <code>facet_wrap</code> separating the group Phenotype for plotting.</p>
<p>4 - Heterozigosity rate. <strong>Note:</strong> You do not have a list of regions with inversions, so do not use <code>--exclude Data/inversion.txt --range</code> in PLINK.</p>
<p><strong>General notes for the exercise:</strong></p>
<ul>
<li>Copy-paste all the commands you need from the tutorial, but remember to carefully check that the file names and options are correct to be used in your data.</li>
<li>Use proper file names for the output of PLINK when filtering and creating new <code>bed/bim/fam</code> files. For example, when you apply the various filterings, create a new output name prefix instead of keeping use <code>mice</code>. Start from <code>mice</code>, then create <code>mice_miss</code>, <code>mice_miss_maf</code>, <code>mice_miss_maf_hwe</code>, <code>mice_miss_maf_hwe_het</code>, so that you can trace back all your work to each analysis step.</li>
<li>Feel free to play with the code for creating plots (e.g.&nbsp;changing colors and cutoff values).</li>
<li>Make use of Generative AI to get help for the code if you think it is not enough to use what is provided in the tutorial, or if you want to explore something beyond the exercise.</li>
</ul>
</div>
</div>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to view answers
</div>
</div>
<div class="callout-body-container callout-body">
<p><i class="bi bi-journal-code"></i> <a href="../develop/exercises/gwas3-mice.html">Download mice-solutions.ipynb</a></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Wrapping up">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Wrapping up
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this tutorial, we have tried several quality control approaches in PLINK and calculated various statistics, which we visualized using <code>R</code>. In the next notebook, you will find even more examples of how you verify the quality of your data.</p>
<p>A table with a small recap of the options used in filtering with PLINK</p>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
</colgroup>
<thead>
<tr class="header">
<th>Step</th>
<th>Command</th>
<th>Function</th>
<th>Thresholds and explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1: Missingness of SNPs</td>
<td><code>--geno</code></td>
<td>Excludes SNPs that are missing in a large proportion of the subjects. In this step, SNPs with low genotype calls are removed</td>
<td>We recommend first filtering SNPs and individuals based on a relaxed threshold (0.2; &gt;20%), as this will filter out SNPs and individuals with very high levels of missingness. Then a filter with a more stringent threshold can be applied (0.02)</td>
</tr>
<tr class="even">
<td>2: Missingness of individuals</td>
<td><code>--mind</code></td>
<td>Excludes individuals who have high rates of genotype missingness. In this step, individuals with low genotype calls are removed</td>
<td>Note, SNP filtering should be performed before individual filtering</td>
</tr>
<tr class="odd">
<td>3: Sex discrepancy</td>
<td><code>--check-sex</code></td>
<td>Checks for discrepancies between the sex of the individuals recorded in the dataset and their sex based on X chromosome heterozygosity/homozygosity rates</td>
<td>Can indicate sample mix‐ups. If many subjects have this discrepancy, the data should be checked carefully. Males should have an X chromosome homozygosity estimate &gt;0.8 and females should have a value &lt;0.2</td>
</tr>
<tr class="even">
<td>4: Minor allele frequency (MAF)</td>
<td><code>--maf</code></td>
<td>Includes only SNPs above the set MAF threshold</td>
<td>SNPs with a low MAF are rare, therefore power is lacking for detecting SNP‐phenotype associations. These SNPs are also more prone to genotyping errors. The MAF threshold should depend on your sample size, larger samples can use lower MAF thresholds. Respectively, for large (N = 100.000) vs.&nbsp;moderate samples (N = 10000), 0.01 and 0.05 are commonly used as MAF thresholds</td>
</tr>
<tr class="odd">
<td>5: Hardy-Weinberg equilibrium (HWE)</td>
<td><code>--hwe</code></td>
<td>Excludes markers which deviate from Hardy–Weinberg equilibrium</td>
<td>Common indicator of genotyping error, may also indicate evolutionary selection. For binary traits, we suggest excluding: HWE p-value &lt;1e−10 in cases and &lt;1e−6 in controls. A less strict case threshold avoids discarding disease‐associated SNPs under selection. For quantitative traits, we recommend HWE p-value &lt;1e‐6.</td>
</tr>
<tr class="even">
<td>6: Heterozygosity</td>
<td><code>R</code> code + <code>--remove</code> option in PLINK</td>
<td>Excludes individuals with high or low heterozygosity rates</td>
<td>Deviations can indicate sample contamination and inbreeding. We suggest removing individuals who deviate ±3 SD from the samples’ heterozygosity rate mean</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="bibliography" class="level1">




</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">Bibliography</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-kanaka_concepts_2023" class="csl-entry" role="listitem">
Kanaka, K. K., Nidhi Sukhija, Rangasai Chandra Goli, Sanjeev Singh, Indrajit Ganguly, S. P. Dixit, Aishwarya Dash, and Anoop Anand Malik. 2023. <span>“On the Concepts and Measures of Diversity in the Genomics Era.”</span> <em>Current Plant Biology</em> 33 (January): 100278. <a href="https://doi.org/10.1016/j.cpb.2023.100278">https://doi.org/10.1016/j.cpb.2023.100278</a>.
</div>
<div id="ref-lachance_hardyweinberg_2016" class="csl-entry" role="listitem">
Lachance, J. 2016. <span>“Hardy–<span>Weinberg</span> <span>Equilibrium</span> and <span>Random</span> <span>Mating</span>.”</span> In <em>Encyclopedia of <span>Evolutionary</span> <span>Biology</span></em>, edited by Richard M. Kliman, 208–11. Oxford: Academic Press. <a href="https://doi.org/10.1016/B978-0-12-800049-6.00022-6">https://doi.org/10.1016/B978-0-12-800049-6.00022-6</a>.
</div>
</div></section><section class="quarto-appendix-contents" id="quarto-copyright"><h2 class="anchored quarto-appendix-heading">Copyright</h2><div class="quarto-appendix-contents"><div>CC-BY-SA 4.0 license</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/hds-sandbox\.github\.io\/GWAS_course\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>