{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1670c4e9",
   "metadata": {
    "Rmd_chunk_options": "setup, include=FALSE",
    "jupyter": {
     "output_hidden": true,
     "source_hidden": true
    },
    "kernel": "R",
    "tags": [
     "scratch"
    ]
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "set.seed(19)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a33b5349",
   "metadata": {},
   "source": [
    "This document is licensed under a [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by-sa/4.0/).\n",
    "\n",
    "The slide set referred to in this document is \"GWAS 10\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6875ffce",
   "metadata": {},
   "source": [
    "### 10.1 Genotype imputation\n",
    "\n",
    "Since long haplotype segments are shared between individuals, the reference panels\n",
    "of haplotypes in different human populations are useful, e.g., to give information\n",
    "for phasing or for choosing tag-SNPs to include on the genotyping arrays.\n",
    "The HapMap project, the 1000 Genomes project, and, more recently, [the Haplotype\n",
    "Reference Consortium](http://www.haplotype-reference-consortium.org/) have\n",
    "collected such data. One of the main use of the\n",
    "haplotype structure in GWAS context is the genotype imputation.\n",
    "\n",
    "**Genotype imputation** (slides 2-9) is the process of predicting or imputing genotypes\n",
    "that are not directly genotyped in a sample of individuals. Most often imputation\n",
    "is done using a reference panel of haplotypes at a dense set of SNPs\n",
    "to impute into a study sample of individuals that have been genotyped\n",
    "at a subset of the SNPs of the reference panel. A review on imputation by\n",
    "[Howie & Marchini (2010)](https://www.nature.com/articles/nrg2796).\n",
    "Imputation software: [IMPUTE4](https://jmarchini.org/impute-4/),\n",
    "[Beagle](https://faculty.washington.edu/browning/beagle/beagle.html)\n",
    "and\n",
    "[Minimac4](https://genome.sph.umich.edu/wiki/Minimac4).\n",
    "\n",
    "As an example, [the UK Biobank](https://www.ukbiobank.ac.uk/)\n",
    "has released genotype data together with\n",
    "thousands of phenotypes on 500,000 British volunteers for research purposes.\n",
    "The samples were genotyped using a genotyping chip with a bit less than 1 million\n",
    "SNPs, but then [imputed](https://www.nature.com/articles/s41586-018-0579-z)\n",
    "to 80 million variants using the existing haplotype reference data.\n",
    "The whole imputed data set is 2 terabytes. The GWAS results on many traits\n",
    "are available for browsing in [Global Biobank Engine](https://biobankengine.stanford.edu/) or [GWAS atlas](http://atlas.ctglab.nl/) and also available for [download](http://www.nealelab.is/uk-biobank/) through Ben Neale's lab.\n",
    "\n",
    "Imputation is important to harmonize cohorts to have the same\n",
    "set of variants before meta-analysis.\n",
    "Imputation is also needed to have as a comprehensive set of variants as possible\n",
    "to be used in fine-mapping.\n",
    "\n",
    "Imputation produces probabilistic genotype predictions, i.e., we don't know with\n",
    "certainty whether each individual has genotype 0, 1 or 2, but instead a probability model\n",
    "gives a probability distribution $(p_{il0}, p_{il1}, p_{il2})$ that\n",
    "describes what are the probabilities that individual $i$ has genotype 0, 1 or 2 at locus $l$.\n",
    "From this distribution, we can compute\n",
    "the expected value of the number of copies of allele 1, so called allele 1 **dosage** as :\n",
    "$\\overline{x}_{il} = 0\\cdot p_{il0} + 1\\cdot p_{il1} + 2\\cdot p_{il2} = p_{il1} + 2\\cdot p_{il2}$.\n",
    "This dosage is then used in the GWAS analysis as if it was an observed genotype.\n",
    "Regression models show their flexibility through\n",
    "their ability to include the imputed dosages in the analysis\n",
    "without any changes to the analysis pipeline of directly observed genotypes.\n",
    "(ANOVA-type group comparisons are not similarly flexible.)\n",
    "\n",
    "**Example 10.1.**\n",
    "Let's consider three SNPs 1,2 and 3, which are not\n",
    "on the genotyping chip that we have used to genotype individual $i$, but which\n",
    "are present in our imputation reference panel, with allele 1 frequencies 0.32, 0.42 an 0.01,\n",
    "respectively.\n",
    "Suppose our imputation results for individual $i$ are\n",
    "\n",
    "|SNP|$p_{0}$|$p_1$|$p_2$| AF1 |\n",
    "|---|-------|-----|-----|---|\n",
    "|1 | 0.8 | 0.19 | 0.01 | 0.32 |\n",
    "|2 | 0.01 | 0.99 | 0.0 | 0.42 |\n",
    "|3 | 0.98 | 0.02| 0.001 | 0.01 |\n",
    "\n",
    "These tell us that\n",
    "we are quite confident that $i$ has genotype 1 at SNP 2,\n",
    "but we have a considerable uncertainty about\n",
    "the genotype at SNP 1. We come back to SNP 3 later.\n",
    "\n",
    "We say that the imputed genotypes are **correctly calibrated**, if, among all\n",
    "individual-locus pairs for which the algorithm estimates genotype $g$\n",
    "with a probability of $p$, a proportion of $p$ indeed have the genotype $g$.\n",
    "For example, if we collect all the individual-locus pairs, such as SNP 1\n",
    "for individual $i$ in Example 10.1 above, where the imputation\n",
    "algorithm has proposed that the individual-locus pair corresponds\n",
    "to genotype 0 with probability in (0.79,0.81),\n",
    "a proportion close to 0.80 do indeed have the genotype 0.\n",
    "Of course we cannot know that the algorithm is correctly calibrated for variants we\n",
    "have not genotyped, but we can at least test the algorithm\n",
    "by masking some of the known genotypes and imputing them and\n",
    "comparing those results to the truth."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2496593",
   "metadata": {},
   "source": [
    "#### 10.1.1 Imputation quality metrics\n",
    "\n",
    "Assuming that the probabilistically imputed genotypes are correctly calibrated,\n",
    "we define a measure of the available information\n",
    "given by imputation compared to the perfect information we would have,\n",
    "had we genotyped the locus of interest.\n",
    "\n",
    "The highest possible information corresponds to the case where no individual\n",
    "has any uncertainty in their imputed genotype. Technically, this happens when\n",
    "the genotype distribution of every\n",
    "individual has one of the probabilities $p_0$, $p_1$ and $p_2$ equal to 1\n",
    "and the other two equal to 0.\n",
    "The **variance** $v_{il}$\n",
    "of individual $i$'s imputed genotype distribution at locus $l$:\n",
    "$$ v_{il} = \\textrm{E}(x_{il}^2|\\pmb{p}_i) - \\textrm{E}(x_{il}|\\pmb{p}_i)^2\n",
    "=4\\cdot p_{il2} + 1\\cdot p_{il1} + 0\\cdot p_{il0} - \\overline{x}_{il}^2\n",
    "=2\\, p_{il2} + \\overline{x}_{il} - \\overline{x}_{il}^2,\n",
    "$$\n",
    "\n",
    "is 0 exactly when there is no uncertainty in the genotype.\n",
    "On the other hand, if LD from imputation reference panel has not provided any\n",
    "new information about the genotype of individual $i$ over what we already had\n",
    "by knowing the allele 1 frequency $f_l$ in the reference panel,\n",
    "then the variance of the imputed genotype for individual $i$ corresponds to the HW variance\n",
    "$w_{l} = 2 f_l (1-f_l).$\n",
    "Thus, if the observed variance $v_{il}$ equals to the HW variance $w_l$, then we have not\n",
    "gained any new information from imputation.\n",
    "Based on these two extreme cases of either perfect information or no information,\n",
    "we define the imputation INFO measure for locus $l$ as\n",
    "$$\n",
    "\\text{INFO}_l = 1- \\frac{1}{n}\\sum_{i=1}^n \\frac{v_{il}}{w_l},\n",
    "$$\n",
    "where the sum is over all imputed individuals.\n",
    "When INFO is 1, there is no uncertainty after imputation, and if INFO is 0, then\n",
    "we have not learned anything new from the imputation over what we could had already guessed based\n",
    "on the population allele frequencies. (Technically, INFO could also go negative,\n",
    "which is interpreted as INFO=0.) This measure is used by versions of the [IMPUTE](http://mathgen.stats.ox.ac.uk/impute/impute_v2.html) software.\n",
    "\n",
    "[Minimac](https://genome.sph.umich.edu/wiki/Minimac)\n",
    "uses a very similar information measure except that the observed\n",
    "variance is not computed as mean over individuals' variances but instead\n",
    "mean of the allele dosages are used to derive the observed variance.\n",
    "For more details about the info measures see [Supplementary 2](https://media.nature.com/original/nature-assets/nrg/journal/v11/n7/extref/nrg2796-s3.pdf)\n",
    "from Marchini & Howie 2010.\n",
    "\n",
    "**Example 10.2.**\n",
    "Consider the imputation probabilities given above in Example 10.1.\n",
    "The variances of the imputed genotype distributions for individual $i$ are"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0eef693c",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "p = matrix(c(0.8, 0.19, 0.01,\n",
    "             0.01, 0.99, 0.0,\n",
    "             0.98, 0.02, 0.001), byrow = T, ncol = 3)\n",
    "f = c(0.32, 0.42, 0.01)\n",
    "cbind(p,f) #check that these are correct\n",
    "x = p %*% 0:2 #allele dosages\n",
    "v = 2*p[,3] + x - x^2 #variance of imputed genotype distribution\n",
    "w = 2*f*(1-f) #HW genotype variance\n",
    "info = 1-v/w #info based on one individual\n",
    "info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "470c1441",
   "metadata": {},
   "source": [
    "We see that even if loci 2 and 3 have similar\n",
    "looking distributions (highest value 0.99 or 0.98),\n",
    "still they have very different info. This is because the distribution\n",
    "at SNP 3 corresponds to\n",
    "the population frequencies under HWE and thus there is no information from\n",
    "imputation, whereas at SNP 2, the imputation has clearly provided\n",
    "new information."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b10963",
   "metadata": {},
   "source": [
    "#### 10.1.2 Summary statistics imputation\n",
    "\n",
    "Suppose that we have access to only the summary association statistics at set $T$ of\n",
    "SNPs genotyped at a GWAS. We would like to impute the summary statistics to a\n",
    "set $U$ of untyped SNPs that we have in our imputation reference panel.\n",
    "The individual level genotype imputation cannot be applied because we don't have\n",
    "access to any individual level genotypes. However, we can impute the summary statistics\n",
    "based on the LD between SNP sets $T$ and $U$ estimated from the reference panel\n",
    "and the observed summary statistics at $T$. We use a standard assumption that the\n",
    "z-scores of the SNPs follow the multivariate Normal distribution.\n",
    "This approach was introduced by Pasaniuc et al. 2014 to their software\n",
    "[Imp-G](http://bogdan.bioinformatics.ucla.edu/software/impg/).\n",
    "\n",
    "Since each untyped SNP will be imputed on its own,\n",
    "we can derive the formulas for imputing a single untyped SNP $u$.\n",
    "We assume that, after reorganising the SNPs so that the typed SNPs $T$\n",
    "come first and $u$ is the last SNP,\n",
    "the (prior) distribution of the z-scores is\n",
    "$$\\pmb{z} = \\left[\n",
    "\\begin{array}{c}\n",
    "\\pmb{z}_T  \\\\ z_u\n",
    "\\end{array}\n",
    "\\right]\n",
    "\\sim \\mathcal{N}_L\\left(\\pmb{0},\n",
    "\\left[\n",
    "\\begin{array}{c|c}\n",
    "\\pmb{R}_{TT}& \\pmb{R}_{Tu} \\\\ \\hline\n",
    "\\pmb{R}_{uT} & \\pmb{R}_{uu}\n",
    "\\end{array}\n",
    "\\right]\n",
    "\\right)\n",
    "$$\n",
    "where we have divided the LD matrix $\\pmb{R}$ into 4 parts according to SNP sets $T$ and $u$.\n",
    "Given the obsereved z-scores $\\pmb{z}_T$ at the typed SNPs, we then\n",
    "use the conditional distribution\n",
    "for the untyped SNPs, that, by the properties of the\n",
    "multivartiate Normal distribution, is\n",
    "$$\n",
    "(z_u\\,|\\, \\pmb{z}_T) = \\pmb{R}_{uT} \\pmb{R}_{TT}^{-1}  \\pmb{z}_T.$$\n",
    "Under the null, this z-score has variance of\n",
    "$v_{u|T} = \\pmb{R}_{uT} \\pmb{R}_{TT}^{-1} \\pmb{R}_{Tu}$ and we use this\n",
    "variance as an information measure of the imputed z-score.\n",
    "If this info value is close to 1 (the true variance of the null z-scores),\n",
    "then the LD structure is very informative about $u$ given $T$, whereas if there\n",
    "is little correlation between $u$ and $T$, then this info value will be close to 0.\n",
    "\n",
    "Note that since some information is lost in imputation compared to the directly\n",
    "observed z-scores, the imputed z-scores will be deflated under the null.\n",
    "This is a consequence of *regression dilution bias*, that causes\n",
    "the regression coefficients to shrink towards zero\n",
    "when predictors in regression model are measured with uncertainty.\n",
    "To correct for this dilution in imputed z-scores,\n",
    "we need to scale them by $1/\\sqrt{v_{u|T}}$.\n",
    "This scaling option is the default z-score in Imp-G\n",
    "and we will also use it in the example below.\n",
    "Thus, our imputed z-score is\n",
    "$$\n",
    "(z_u\\,|\\, \\pmb{z}_T) = \\frac{\\pmb{R}_{uT} \\pmb{R}_{TT}^{-1}  \\pmb{z}_T}{\\sqrt{v_{u|T}}} = \\frac{\\pmb{R}_{uT} \\pmb{R}_{TT}^{-1}  \\pmb{z}_T}{\\sqrt{\\pmb{R}_{uT} \\pmb{R}_{TT}^{-1} \\pmb{R}_{Tu}}}.$$\n",
    "\n",
    "\n",
    "**Example 10.3.**\n",
    "Let's do a GWAS using data from Exercise 6.5 and choose randomly 100 SNPs to be imputed\n",
    "using 20 to 200 other SNPs and\n",
    "compare the imputation results to the true z-scores.\n",
    "Let's color the SNPs according to the estimated imputation info from\n",
    "the summary statistics imputation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae042c82",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "set.seed(118)\n",
    "n = 1000\n",
    "p = 2000\n",
    "X = matrix(scan(\"https://www.mv.helsinki.fi/home/mjxpirin/GWAS_course/2019/material/ex5.2.txt\"), byrow = T, ncol = p, nrow = n)\n",
    "R = cor(X) #LD matrix\n",
    "y = rep(c(1,0),c(n/2,n/2)) #1st 500 cases rest controls\n",
    "gwas = t(apply(X, 2, function(x){ summary( glm( y ~ x, family = \"binomial\") )$coeff[2,] }))\n",
    "Ud = sample(1:p, size = 100) #indexes of the \"untyped\" SNPs to be imputed\n",
    "par(mfrow=c(2,3))\n",
    "par(mar=c(5.5,4.5,4.5,1))\n",
    "for(n.Td in c(10,20,40,60,100,200)){ #\n",
    "  Td = sample(setdiff(1:p, Ud), size = n.Td)\n",
    "  R.TdUd = R[Td,Ud] #LD of typed - untyped pairs\n",
    "  inv.R.Td = solve( R[Td,Td] + 0.001*diag(n.Td) ) #inverse of LD of typed SNPs\n",
    "  W = R[Ud, Td] %*% inv.R.Td #these are the weights that turn typed to imputed\n",
    "  infos = as.numeric(rowSums(W * R[Ud, Td])) #info measures per each untyped\n",
    "  z.imp = (W %*% gwas[Td, 3])/sqrt(infos) #use scaling 1/sqrt(infos) to get var=1 for z-scores\n",
    "  cols = rep(\"darkblue\", length(Td))\n",
    "  cols[infos < 0.75] = \"cyan\"\n",
    "  cols[infos < 0.5] = \"violet\"\n",
    "  cols[infos < 0.25] = \"red\"\n",
    "  plot(gwas[Ud,3],z.imp, col = cols, pch = 19,\n",
    "       main = paste0(\"|Td|=\",n.Td,\"; r=\", signif(cor(gwas[Ud,3],z.imp),2)),\n",
    "       xlab = \"z-score\", ylab = \"Imputed z-score\",\n",
    "       sub = paste0(\"cor(info,acc)=\", signif( cor(-(gwas[Ud,3]-z.imp)^2,infos, method=\"spearman\"),2) ))\n",
    "  abline(0,1, lty=2, col=\"gray\", lwd=2)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afccd270",
   "metadata": {},
   "source": [
    "We can see how the accuracy of the imputed z-scores gets better as we include more SNPs\n",
    "to be used in the imputation. The info measure is also able to tell when the\n",
    "LD-structure is more informative (dark blue) and when it carries little info (red/violet).\n",
    "The correlation value below plots show how the accuracy of the imputed z-scores\n",
    "is correlated with info measure. When info is well calibrated, we expect a positive\n",
    "correlation where increasing info goes with higher imputation accuracy."
   ]
  }
 ],
 "metadata": {
  "Rmd_chunk_options": {
   "author": "Matti Pirinen, University of Helsinki",
   "date": "27-Feb-2019",
   "output": {
    "html_document": "default",
    "pdf_document": "default"
   },
   "title": "GWAS 10: Genotype imputation",
   "urlcolor": "blue"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ],
    [
     "R",
     "ir",
     "",
     ""
    ]
   ]
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
