{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "11ea6652",
   "metadata": {
    "Rmd_chunk_options": "setup, include=FALSE",
    "jupyter": {
     "output_hidden": true,
     "source_hidden": true
    },
    "kernel": "R",
    "tags": [
     "scratch"
    ]
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "set.seed(19)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b3f2021",
   "metadata": {},
   "source": [
    "This document is licensed under a [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by-sa/4.0/).\n",
    "\n",
    "The slide set referred to in this document is \"GWAS 8\".\n",
    "\n",
    "We have seen Manhattan plots from BMI, migraine, schizophrenia etc.\n",
    "where the number of association regions rises to hundreds.\n",
    "How much these regions explain of the population\n",
    "variation in a trait or its liability?\n",
    "How many other genomic regions, or additional variants in the current regions,\n",
    "might also contribute to these phenotypes?\n",
    "Is there something special about the genomic regions that\n",
    "harbor causal variants for a particular phenotype?\n",
    "How well can we predict these phenotypes from genetic data\n",
    "and how larger GWAS will help these predictions?\n",
    "\n",
    "These are questions about the **genetic architecture** of the phenotypes.\n",
    "A primary parameter of the genetic architecture is **heritability**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b4e756a",
   "metadata": {},
   "source": [
    "### 8.1 Heritability\n",
    "\n",
    "Heritability measures, in a particular population,\n",
    "the proportion of variance of the phenotype that is due to genetic differences between individuals.\n",
    "A review by [Visscher et al.](https://www.nature.com/articles/nrg2322)\n",
    "is an excellent overview of the concept and its interpretation.\n",
    "\n",
    "An important point is that heritability is always a property of a particular population,\n",
    "and its value can vary between different environmental conditions as the relative roles of\n",
    "genes and environment change.\n",
    "\n",
    "Heritabilities have been estimated for a long time using phenotypic correlations between relatives\n",
    "of different degrees. (Slides 3-4)\n",
    "For example, the traditional twin estimates compare phenotypic similarity\n",
    "in monozygotic twin pairs (who share whole genome IBD = 2)\n",
    "to similarity in dizygotic twin pairs\n",
    "(who are genetically like any other pairs of full-siblings: IBD0 = 25%,\n",
    "IBD1 = 50%, IBD2 = 25%).\n",
    "Under some (strong!) assumptions, such as that the environmental\n",
    "contribution to the phenotypic similarity would be the same\n",
    "for a monozygotic as for a dizygotic twin pair,\n",
    "and that the genetic effects act additively over loci,\n",
    "it follows that the difference between the correlation\n",
    "estimates of these two types of twin pairs\n",
    "leads to an estimate of heritability.\n",
    "\n",
    "More generally, any pedigree records can been used for estimating how the observed\n",
    "phenotypic correlations can be explained by the estimated genetic sharing,\n",
    "which, under (strong!) assumptions\n",
    "about covariance of the environmental effects between different relative types,\n",
    "again leads to estimates of heritability.\n",
    "\n",
    "Here our interest is in how much heritability do we explain with the already\n",
    "discovered GWAS regions,\n",
    "and whether we could use GWAS data, that typically have only few close relatives,\n",
    "to estimate heritability of a trait.\n",
    "In particular, such estimates should be immune to the confounding\n",
    "factors of the shared environment that occurrs between close relatives."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcda43ee",
   "metadata": {},
   "source": [
    "### 8.2 Heritability of GWAS loci\n",
    "\n",
    "We consider the measure of\n",
    "**heritability in the narrow sense**, $h^2$, which is the heritability due to\n",
    "the **additive genetic effects**. Thus, *dominance effects* within one locus\n",
    "(i.e. the amount by which the heterozygotes' phenotype mean deviates\n",
    "from the average of the two groups of homozygotes), or interaction\n",
    "effects (*epistasis*) between multiple loci, are not included in this estimate.\n",
    "The broad-sense\n",
    "heritability $H^2$, which includes all variance due to genetics, is\n",
    "more difficult to estimate than $h^2$.\n",
    "\n",
    "If SNP $l$ has MAF $f_l$ and allelic causal effect\n",
    "$\\lambda_l$, then the phenotypic variance explained (causally) by\n",
    "the SNP is $2f_l(1-f_l)\\lambda_l^2 = {\\lambda_l^*}^2$.\n",
    "(Here $\\lambda_l^*$ denotes the scaled causal effect introduced in Section 7.)\n",
    "If the trait variance in population is $\\sigma_Y^2$,\n",
    "then the (additive) heritability contributed by the SNP is\n",
    "$$h_l^2 = \\frac{{\\lambda_l^*}^2}{\\sigma_Y^2} = \\frac{2f_l(1-f_l)\\lambda_l^2}{\\sigma_Y^2}.$$\n",
    "In practice, the variance explained is often estimated\n",
    "using an estimate of the marginal effect as\n",
    "$(\\widehat{\\beta}_l^*)^2 = 2 f_l (1-f_l) \\widehat{\\beta}_l^2$.\n",
    "This is the phenotypic variance that is explained by genetic variation\n",
    "tagged by variant $l$, (similar interpretation applies to heritability),\n",
    "but is still often interpeted as \"variance explained by variant $l$\".\n",
    "\n",
    "Let's assume that we have standardized the phenotype ($\\sigma_Y^2=1$) before the GWAS.\n",
    "According to the simplest model, assuming no LD or deviation from\n",
    "the additivity between the causal effects of different variants, the\n",
    "additive heritability over all variants is\n",
    "$h^2 = \\sum_{l=1}^p {\\lambda^*_l}^2.$\n",
    "More generally, allowing LD, $p$ variants in a region\n",
    "contribute to the phenotype of one individual by the quantity\n",
    "${\\pmb{x}^*}^T \\pmb{\\lambda^*},$\n",
    "and the variance contributed by this region is\n",
    "$$\n",
    "\\begin{aligned}\n",
    "h^2_{\\textrm{reg}} &= \\textrm{Var}({\\pmb{x}^*}^T \\pmb{\\lambda^*}) =\n",
    "\\pmb{\\lambda^*}^T \\textrm{Var}(\\pmb{x}^*) \\pmb{\\lambda^*} =\n",
    "\\pmb{\\lambda^*}^T \\pmb{R} \\pmb{\\lambda^*} \\\\\n",
    "&=\n",
    "(\\pmb{R}^{-1}\\pmb{\\beta^*})^T \\pmb{R} (\\pmb{R}^{-1}\\pmb{\\beta^*})=\n",
    "\\pmb{\\beta^*}^T \\pmb{R}^{-1} \\pmb{\\beta^*},\n",
    "\\end{aligned}\n",
    "$$\n",
    "\n",
    "where $\\pmb{R}$ is the LD-matrix (Pearson correlations) of the $p$ variants.\n",
    "The above formula shows how to compute the regional heritability using either\n",
    "the causal effect sizes or the marginal effect sizes (latter of which are directly estimated\n",
    "by the standard GWAS).\n",
    "If there is no LD ($\\pmb{R} = \\pmb{I}$) between the variants,\n",
    "then the heritability is the sum of the squared scaled effects\n",
    "(either causal or marginal effects as they are the same in this case).\n",
    "For example, [FINEMAP](http://www.christianbenner.com/) (see Section 7 of material)\n",
    "gives an estimate of regional heritability for each causal configuration in its .config file\n",
    "according to the formula above.\n",
    "\n",
    "**Example 8.1.**\n",
    "Consider SNPs 1 and 2 whose minor allele correlation is $r_{12}$,\n",
    "MAFs are 0.24 and the marginal estimates are\n",
    "$\\widehat{\\beta}_1 =\\widehat{\\beta}_2 = 0.08$\n",
    "in a QT GWAS with a phenotypic variance of 1.\n",
    "What is the estimate of heritability explained by\n",
    "these SNPs jointly, and how much it differs from the case where they were independent,\n",
    "when $r_{12}$ is 0.99 (almost the same variant twice),\n",
    "0.3 (positively correlated effect alleles),\n",
    "0 (independece) or -0.2 (effect alleles are masking each other)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53e0ea89",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "b.est = c(0.08, 0.08)\n",
    "f = c(0.24, 0.24)\n",
    "b.s = b.est * sqrt(2*f*(1-f)) #scaled marginal effects\n",
    "res = c()\n",
    "for(r.12 in c(0.99, 0.3, 0, -0.2)){\n",
    "  R = matrix(c(1, r.12, r.12, 1),2,2)\n",
    "  #regional heritability b.s^T R^-1 b.s\n",
    "  h2.reg = t(b.s) %*% solve(R) %*% b.s #solve(R) is the inverse of R\n",
    "  h2.ind = sum(b.s^2)\n",
    "  res = rbind(res,c(r.12, h2.reg, h2.ind))\n",
    "}\n",
    "colnames(res) = c(\"r.12\",\"h2.reg\",\"h2.ind\")\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba63f759",
   "metadata": {},
   "source": [
    "When the effect alleles are 0.99 correlated, then they are explaining the same\n",
    "signal and their joint variance explained is, correctly, only half of the sum of the marginals.\n",
    "With a negative correlation, and a similar direction and size in the observed effects,\n",
    "the SNPs must been masking each other's marginal effects and their total contribution\n",
    "is larger than the sum of their marginal contributions.\n",
    "\n",
    "\n",
    "For diseases, the heritability is often measured on the **liability scale**, which requires\n",
    "an estimate of the disease prevalence\n",
    "(see [Box 5](https://www.nature.com/articles/nrg2322) of Visscher et al.).\n",
    "\n",
    "A few examples of the heritability estimates summed over GWAS loci\n",
    "(defined by a variant having P < 5e-8), in decreasing order\n",
    "of heritability:\n",
    "\n",
    "* Paraoxonase-1 level has a heritability of 70% at a single locus\n",
    "(unsurprisingly, harboring the gene *PON1*) [Benner et al. 2018](https://www.biorxiv.org/content/10.1101/318618v1).\n",
    "\n",
    "* LDL-cholesterol study by [Surakka et al. 2014](https://www.nature.com/articles/ng.3300)\n",
    "explained 20% of the variance in LDL-C by ~100 loci.\n",
    "\n",
    "* Height study by [Wood et al. 2014](https://www.nature.com/articles/ng.3097) found that 423\n",
    "loci accounted for 16% of variation in height.\n",
    "\n",
    "* Crohn's disease study by [Jostins et al. 2015](https://www.nature.com/articles/nature11582)\n",
    "explained about 14% of variance in liability from 160 loci.\n",
    "\n",
    "* Schizophrenia study by [Ripke et al. 2014](https://www.nature.com/articles/nature13595) explained\n",
    "about 3.4% of variance in liability using 108 GWS loci.\n",
    "\n",
    "* BMI study by [Locke et al. 2015](https://www.nature.com/articles/nature14177) found that the 97 GWS loci account for 2.7% of BMI variation.\n",
    "\n",
    "In particular, for many complex diseases, such as schizophrenia, the variance\n",
    "explained by GWS loci is still very small. On the other hand, traditional ways to estimate\n",
    "heritabilities suggest a very high heritability for schizophrenia.\n",
    "Where is that heritability, if the first 100 GWAS loci show this little traces of it?\n",
    "Or are the traditional estimates grossly overestimating heritability?\n",
    "This gap is called the [missing heritability problem](https://en.wikipedia.org/wiki/Missing_heritability_problem) (Slides 6-7).\n",
    "\n",
    "Next we will look at two approaches that use GWAS data to estimate\n",
    "the genome-wide **SNP heritability** $h^2_{\\textrm{SNP}}$ that considers the heritability\n",
    "contribution of all SNPs that are included in the study,\n",
    "not just those that happen to reach the genome-wide significance level.\n",
    "The first method, the **linear mixed model**, is based\n",
    "on an efficient way of correlating the variant sharing with the phenotypic similarity\n",
    "in the population sample.\n",
    "The second method, **LD-score regression (LDSC)**, is based on a link between\n",
    "the amount of tagging-by-LD and the amount of GWAS signal seen at a variant,\n",
    "that is induced by a highly polygenic genetic architecture.\n",
    "The linear mixed model requires original phenotype-genotype data whereas\n",
    "LDSC works with the GWAS summary data (marginal effect estimates and their SEs)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fab9507",
   "metadata": {},
   "source": [
    "### 8.3. Linear Mixed Model\n",
    "\n",
    "Let's write down the full linear model for additive effects across the genome for a quantitative\n",
    "phenotype $Y$:\n",
    "$$y_i = \\mu + \\pmb{z}_i^T\\pmb{\\alpha} + {\\pmb{x}_i^*}^T\\pmb{\\lambda}^* + \\varepsilon_i= \\mu + \\pmb{z}_i^T\\pmb{\\alpha} + \\sum_{l=1}^p x_{il}^* \\lambda_l^* + \\varepsilon_i,$$\n",
    "where $\\pmb{z}_i$ is the vector of covariate values for $i$ and $\\varepsilon_i\\sim\\mathcal{N}(0,\\sigma_E^2)$\n",
    "is the (environmental) error term that is assumed to be uncorrelated across individuals.\n",
    "\n",
    "We saw in GWAS 7 that if we try to estimate the parameters $\\lambda_l^*$ using the ordinary\n",
    "least squares estimator, we run into problems because of high correlations between SNPs and,\n",
    "more generally, overfitting.\n",
    "This is because the standard linear regression model\n",
    "is too flexible to adapt to the data when $p$ grows large\n",
    "if the effect sizes are not restricted in any way.\n",
    "To overcome these problems, the linear mixed model treats the effect sizes\n",
    "$\\lambda_l^*$ as **random effects** that share a common (prior) distribution,\n",
    "here chosen to be $\\mathcal{N}(0,\\tau^2)$, where $\\tau^2$ will be estimated from the data.\n",
    "Now the parameters $\\lambda_l^*$ are not allowed freely to choose their\n",
    "values but their magnitude is restricted by a *shared* variance parameter $\\tau^2$.\n",
    "The model is able to learn from the data how the whole set of values of $\\lambda^*$\n",
    "look like when considered together, and then apply that information to keep the\n",
    "magnitude of $\\lambda^*$s appropriate by adjusting a single variance parameter $\\tau^2$.\n",
    "Another way to think about the difference between this random effects model and the\n",
    "least squares estimation is that our focus changes from estimating\n",
    "each of the $p$ values $\\lambda_l^*$ to estimating their *shared* distribution,\n",
    "as determined by the variance parameter $\\tau^2$.\n",
    "Hence we reduce the number of parameters estimated from $p$ to 1,\n",
    "and will avoid overfitting.\n",
    "\n",
    "The name *mixed model* reflects that the model is a *mix* of both **fixed effects** $\\alpha$,\n",
    "whose individual values are estimated as in the standard linear model,\n",
    "and **random effects** $\\lambda^*_l$, whose joint *distribution*\n",
    "is estimated, rather than the individual values of $\\lambda_l^*$s.\n",
    "\n",
    "How can we link the new parameter $\\tau^2$ to the observed values of $y$?\n",
    "The answer is to write down what the random effect\n",
    "assumption means in terms of the observed similarity (mathematically covariance)\n",
    "of the phenotypes of individuals $i$ and $j$.\n",
    "If we follow the random effect formulation, and independently\n",
    "draw each $\\lambda_l^* \\sim \\mathcal{N}(0,\\tau^2)$, what is the\n",
    "consequence on the phenotypic covariance between $i$ and $j$, induced\n",
    "by the terms $\\eta_i = \\sum_{l=1}^p x_{il}^* \\lambda_l^*$\n",
    "and $\\eta_j = \\sum_{l=1}^p x_{jl}^* \\lambda_l^*$?\n",
    "\n",
    "$$\n",
    "\\begin{aligned}\n",
    "\\textrm{Cov}(\\eta_i, \\eta_j) &=\n",
    "\\textrm{Cov}\\left(\\sum_{l=1}^p x_{il}^* \\lambda_l^*, \\sum_{l=1}^p x_{jl}^* \\lambda_l^* \\right) =\n",
    "\\sum_{l=1}^p \\sum_{k=1}^p x_{il}^* x_{jk}^*\\, \\textrm{Cov}(\\lambda_l^*,\\lambda_k^*) =\n",
    "\\sum_{l=1}^p x_{il}^* x_{jl}^*\\, \\textrm{Cov}(\\lambda_l^*,\\lambda_l^*)  \\\\\n",
    "&= \\sum_{l=1}^p x_{il}^* x_{jl}^* \\tau^2\n",
    "= p\\tau^2 \\pmb{G}_{ij},\n",
    "\\end{aligned}\n",
    "$$\n",
    "where $\\pmb{G}$ is the GRM-cor from chapter 5 of the course material,\n",
    "i.e., $\\pmb{G} = \\frac{1}{p}\\pmb{X}^*{\\pmb{X}^*}^T$\n",
    "is the $n\\times n$ empirical correlation matrix for individuals computed across all SNPs.\n",
    "\n",
    "This is saying that the **additive genetic components** $\\eta_i$ of the trait are correlated\n",
    "across the individuals according to the genetic relatedness of the individuals,\n",
    "as measured by GRM-cor, and are scaled so that their\n",
    "variance is $p\\tau^2$, where $\\tau^2$ is the variance of the causal effect sizes.\n",
    "Note that $\\tau^2$ is also the expected phenotypic variance contributed by any one causal effect\n",
    "as $\\textrm{E}({\\lambda_l^*}^2) = \\textrm{Var}(\\lambda_l^*) = \\tau^2$.\n",
    "If we ignore LD between nearby variants, then $\\sigma_G^2  = p\\tau^2$ is the expected phenotypic variance\n",
    "contributed by all $p$ variants together,\n",
    "and we would estimate the heritability as $\\widehat{h}^2 = {\\sigma_G^2}/{(\\sigma_G^2 + \\sigma_E^2)}.$\n",
    "\n",
    "The last step is to write down the joint distribution of the\n",
    "phenotype vector $\\pmb{y}$, as defined by the variance components $\\eta$ and $\\varepsilon$,\n",
    "from the relationship $y_i = \\mu + \\pmb{z}_i^T\\pmb{\\alpha} + \\eta_i + \\varepsilon_i$.\n",
    "The phenotype vector is an $n$-dimensional multivariate normal vector,\n",
    "whose mean $\\pmb{\\mu}$ has components $\\mu_i = \\mu + \\pmb{z}_i^T\\pmb{\\alpha}$\n",
    "and whose covariance matrix\n",
    "$\\pmb{\\Sigma}(\\pmb{\\sigma^2}) =  \\sigma_G^2 \\pmb{G} + \\sigma_E^2 \\pmb{I}_n$\n",
    "is a function of two unknown variance parameters $\\pmb{\\sigma^2} = (\\sigma^2_G,\\sigma_E^2).$\n",
    "(Slide 8.)\n",
    "\n",
    "A naive computation of such $n$-dimensional multivariate Normal likelihoods is\n",
    "expensive -- $\\mathcal{O}(n^3)$ operations -- and in recent years\n",
    "many new ways to speed up the computation have been introduced.\n",
    "Currently,\n",
    "[GCTA](https://cnsgenomics.com/software/gcta/#Overview) with its **fastGWA** module\n",
    "remains a widely-used method and [BOLT-REML](https://data.broadinstitute.org/alkesgroup/BOLT-LMM/) is\n",
    "an efficient implementation of the mixed model applicable to 100,000s of samples.\n",
    "Recently, [REGENIE](https://rgcgithub.github.io/regenie/overview/) has implemented an efficient\n",
    "version of a similar model.\n",
    "\n",
    "In what follows, we will\n",
    "experiment with the mixed model by using a simple trick that can be done easily in R, but\n",
    "which would not generalize to multiple variance components, and therefore differs\n",
    "from more complex methods such as GCTA and BOLT-REML."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a4765dc",
   "metadata": {},
   "source": [
    "#### 8.3.1 A Mixed model estimation method\n",
    "\n",
    "To simplify the setting, let's assume we first regress out the covariates from $y$\n",
    "using linear regression\n",
    "and consider the (quantile normalized) residuals from that regression as our\n",
    "covariate-adjusted phenotype $y'$.\n",
    "Our task is to maximize the multivariate Normal likelihood function of\n",
    "$$\\pmb{y'}\\sim \\mathcal{N}\\left(0,\\pmb{\\Sigma}(\\pmb{\\sigma^2})\\right), \\textrm{ with respect to } \\pmb{\\sigma^2}. $$\n",
    "\n",
    "The log-likelihood of the multivariate Normal is\n",
    "$$L(\\pmb{\\sigma^2}) = -\\frac{1}{2} \\log \\det\\pmb{\\Sigma}(\\pmb{\\sigma^2})\n",
    "-\\frac{1}{2} \\pmb{y'}^T \\pmb{\\Sigma}(\\pmb{\\sigma^2})^{-1} \\pmb{y'}.$$\n",
    "\n",
    "If we make an eigendecomposition of the GRM-cor matrix $\\pmb{G}=\\pmb{U}\\pmb{D}\\pmb{U}^T$,\n",
    "where $\\pmb{U}$ is an orthonormal matrix of eigenvectors and $\\pmb{D}$ is a diagonal matrix\n",
    "of eigenvalues, then the inverse and determinant of the $n$ dimensional matrix\n",
    "can be transformed to those of diagonal matrices by rotating the phenotype with the\n",
    "eigenvectors into a new phenotype $\\widetilde{\\pmb{y}} = \\pmb{U}^T\\pmb{y}':$\n",
    "$$L(\\pmb{\\sigma^2}) = -\\frac{1}{2} \\sum_{i=1}^n \\log(\\sigma_G^2 D_i + \\sigma_E^2) -\\frac{1}{2}\n",
    "\\sum_{i=1}^n \\frac{\\widetilde{y}^2_i}{\\sigma_G^2 D_i + \\sigma_E^2} =\n",
    "-\\frac{1}{2} \\sum_{i=1}^n \\left( \\log(\\sigma_G^2 D_i + \\sigma_E^2) +\n",
    "\\frac{\\widetilde{y}^2_i}{\\sigma_G^2 D_i + \\sigma_E^2} \\right),$$\n",
    "where $D_i$ is the diagonal element $i$ of $\\pmb{D}$.\n",
    "(The derivation of the above transformation\n",
    "in not explained in detail here but is given by\n",
    "[Pirinen et al. 2013](https://projecteuclid.org/euclid.aoas/1365527203).)\n",
    "This version of the log-likelihood is easy to optimize in R by using `optim()`, and\n",
    "we can help `optim()` by giving it also the gradient of the log-likelihood.\n",
    "Here is a function that returns the log-likelihood which we can then maximize by `optim()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fee99c38",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "lmm.loglik <- function(sigma, y, d)\n",
    "{\n",
    "\tsigma.sum = sigma[1] * d + sigma[2]\n",
    "\tres = -0.5 * sum(log(sigma.sum) + y ^ 2 / sigma.sum)\n",
    "\treturn(res) #returns log likelihood\n",
    "}\n",
    "\n",
    "lmm.gradient <- function(sigma, y, d)\n",
    "{\n",
    "\tsigma.sum = sigma[1] * d + sigma[2]\n",
    "\ttmp = y ^ 2 / sigma.sum - 1\n",
    "\tdsigmaG = 0.5 * sum(d / sigma.sum * tmp)\n",
    "\tdsigmaE = 0.5 * sum(1 / sigma.sum * tmp)\n",
    "\treturn(c(dsigmaG, dsigmaE)) #returns gradient\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22e63f83",
   "metadata": {},
   "source": [
    "**Example 8.2.**\n",
    "Let's try the mixed model with $n=2000$ samples and $p=10,000$ independent common variants\n",
    "with MAF 0.5 and simulate trait with $h^2=0.5$.\n",
    "We expect that each variant here would explain only $h^2/p = 0.00005$\n",
    "of the trait variance!\n",
    "We don't expect to have any genome-wide significant findings with $n=2000$ samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1a01f9af",
   "metadata": {
    "Rmd_chunk_options": "fig.width = 4, fig.height=4",
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "p = 10000\n",
    "n = 2000\n",
    "f = 0.5\n",
    "h2 = 0.5\n",
    "\n",
    "X.s = scale(replicate(p, rbinom(n, size = 2, p = f) )) #use scaled genotypes\n",
    "\n",
    "#tau^2 = var(lambda.s) = h2 / p\n",
    "lambda.s = rnorm(p, 0, sqrt( h2 / p)) #scaled effects\n",
    "\n",
    "#generate phenotype: SNP effects + random noise with var=1-h2\n",
    "y = scale( X.s %*% lambda.s + rnorm(n, 0, sqrt(1-h2) )) #scaling makes mean(y)=0 as our LMM ignores intercept!\n",
    "\n",
    "#test individual SNPs and make a QQ-plot\n",
    "pval = as.numeric(apply(X.s, 2, function(x){summary(lm( y ~ x))$coefficients[2,4]}))\n",
    "expect.stats = qchisq(ppoints(p), df = 1, lower = F)\n",
    "obs.stats = qchisq(pval, df = 1, lower = F)\n",
    "lambda = median(obs.stats) / median(expect.stats) #GC lambda = ratio at medians\n",
    "qqplot(expect.stats, obs.stats, xlab = \"chisq expected\", ylab = \"chisq observed\",\n",
    "       sub = paste0(\"lambda=\",signif(lambda,3)), cex = 0.8)\n",
    "abline(0,1)\n",
    "sort(pval)[1:5] #show the 5 smallest P-values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9916c553",
   "metadata": {},
   "source": [
    "We have no genome-wide significant SNPs and the QQ-plot doesn't look very inflated.\n",
    "However, the data were simulated in such a way that we expect that the SNPs explain together\n",
    "half of the phenotypic variance, even if we can't see any individual SNP having a clear effect.\n",
    "\n",
    "What does the mixed model say?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9bf2b170",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "#G = cor(t(X.s)) #make GRM-cor matrix\n",
    "G = (X.s %*% t(X.s))/p #make GRM-cor matrix (see Chapter 5.1.2.)\n",
    "eig = eigen(G) #decompose G = U D t(U)\n",
    "y.Ut = t(eig$vectors) %*% y #transform y to y.Ut by using U^t y\n",
    "d = eig$values #eigenvalues of G\n",
    "\n",
    "start.val = c(1, 1) / 2 #starting values for the optimization of the two variance parameters\n",
    "#1st optim run with robust Nelder-Mead method without gradient\n",
    "res = optim(start.val, fn = lmm.loglik, y = y.Ut, d = d,\n",
    "            method = 'Nelder-Mead', control = list(fnscale = -1)) #fnscale=-1 to maximize, not minimize\n",
    "#2nd optim run refining the estimate by BFGS method using gradient\n",
    "res = optim(res$par, fn = lmm.loglik, gr = lmm.gradient, y = y.Ut, d = d,\n",
    "            method = 'BFGS', hessian = T, control = list(fnscale = -1)) #fnscale=-1 to maximize, not minimize\n",
    "sigma2 = res$par #estimates of sigma.G^2 and sigma.E^2\n",
    "sigma2.SE = sqrt(diag(solve(-res$hessian))) #SE of sigma_G^2 and sigma_E^2\n",
    "res = cbind(sigma2, sigma2.SE) # genetic variance and environmental variance with SEs\n",
    "rownames(res) = c(\"sigma.G^2\",\"sigma.E^2\")\n",
    "res\n",
    "\n",
    "h2.est = sigma2[1]/sum(sigma2)\n",
    "h2.est # heritability is ~sigma.G^2 since var(y)=1. Then we also have SE(h2) = SE(sigma.G^2)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb0c3c8c",
   "metadata": {},
   "source": [
    "Our estimate is close to the true value of 50%.\n",
    "(Note that given the SE of 0.07 we are lucky to get *this* close here.)\n",
    "The mixed model can indeed pick up the\n",
    "joint contribution of all those tiny effects!\n",
    "Note that we haven't derived SE for $\\widehat{h}^2$ but\n",
    "in cases where total variance of phenotype is 1,\n",
    "we can use SE of $\\widehat{\\sigma}_G^2$ as SE of $\\widehat{h}^2$.\n",
    "\n",
    "**Example 8.3.**\n",
    "LMM worked nicely above, but what happens when the true genetic architecture is not 100% polygenic,\n",
    "that is, when only a subset of all variants contribute to the phenotype.\n",
    "The mixed model was derived assuming that all effects are non-zero.\n",
    "What happens if we simulated non-zero effects only for, say 20% of the SNPs?\n",
    "Let's recycle our genotype data and $G$ matrix and its decomposition since\n",
    "those take some time to make. Let's simply simulate new sets of\n",
    "$\\lambda^*$s for a scenario where 20% of SNPs have non-zero\n",
    "effects and together they explain 30%\n",
    "of the trait variance. Let's repeat this 10 times and plot the estimates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ed8fa918",
   "metadata": {
    "Rmd_chunk_options": "eval=T",
    "kernel": "R",
    "tags": [
     "report_cell"
    ]
   },
   "outputs": [],
   "source": [
    "set.seed(12)\n",
    "phi = 0.2 # proportion phi of the SNPs are non-zero\n",
    "h2 = 0.3\n",
    "n.iter = 10 #how many simulations -- all use the same genotype data\n",
    "h2.res = matrix(NA, ncol = 2, nrow = n.iter)\n",
    "\n",
    "for(iter in 1:n.iter){\n",
    "  #choose which SNPs have an effect\n",
    "  c.ind = sort(sample(1:p, size = round(phi*p)))\n",
    "  #var(lambda.s) = h2 / (phi*p)\n",
    "  lambda.s = rnorm(length(c.ind), 0, sqrt( h2 / (phi*p)))\n",
    "\n",
    "  #generate phenotype: SNP effects + random noise with var = 1 - h2\n",
    "  y = scale( X.s[,c.ind] %*% lambda.s + rnorm(n, 0, sqrt(1-h2) ))#scale: mean(y) = 0, sd(y) = 1\n",
    "  y.Ut = t(eig$vectors) %*% y #transform y to y.Ut = U^t y\n",
    "\n",
    "  start.val = c(1, 1) / 2 #starting values for the optimization of the two variance parameters\n",
    "  #1st optim run with robust Nelder-Mead method without gradient\n",
    "  res = optim(start.val, fn = lmm.loglik, y = y.Ut, d = d,\n",
    "              method = 'Nelder-Mead', control = list(fnscale = -1))\n",
    "  #2nd optim run refining the estimate by BFGS method using gradient\n",
    "  res = optim(res$par, fn = lmm.loglik, gr = lmm.gradient, y = y.Ut, d = d,\n",
    "              method = 'BFGS', hessian = T, control = list(fnscale = -1))\n",
    "  sigma2 = res$par #estimates of sigma.G^2 and sigma.E^2\n",
    "  sigma2.SE = sqrt(diag(solve(-res$hessian))) #SE of sigma_G^2 and sigma_E^2\n",
    "  h2.res[iter,] = c(sigma2[1]/sum(sigma2),sigma2.SE[1]) # h2.est with sigma.G^2 SE\n",
    "}\n",
    "plot(1:n.iter, h2.res[,1], pch = 19,\n",
    "     main = paste0(\" n = \",n,\" p = \",p,\" h2 = \",h2,\" phi = \",phi),\n",
    "     xaxt=\"n\", ylab=\"h2 estimate\", ylim = c(0,1), xlab = \"\")\n",
    "arrows(1:n.iter, h2.res[,1]-1.96*h2.res[,2], 1:n.iter, h2.res[,1]+1.96*h2.res[,2],\n",
    "       code = 3, length = 0.05, angle = 90)\n",
    "abline(h = h2, lty = 2, col = \"green\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d05011",
   "metadata": {},
   "source": [
    "It works fine even when only 20% of the SNPs have non-zero effects.\n",
    "It seems that LMM is robust to a considerable proportion of zero effects among the SNPs,\n",
    "which is great news!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d41cba5a",
   "metadata": {},
   "source": [
    "#### 8.3.2 Mixed model heritability estimates\n",
    "\n",
    "LMMs have been used for a long time in animal breeding and pedigree\n",
    "analyses and this approach became popular in GWAS data in humans by the landmark\n",
    "publication of [Yang et al. (2010)](https://www.nature.com/articles/ng.608)\n",
    "where they showed that 45% of the variation in height\n",
    "could be explained by about 300,000 SNPs on a genotyping chip.\n",
    "This was an important piece of\n",
    "information for the discussion surrounding the missing heritability problem, because\n",
    "after that publication it became more widely considered plausible that a lot of the gap between\n",
    "the variance explained by GWS regions (only ~5% for height at that time)\n",
    "and the heritability of height as estimated by twin and sibling studies (~70-80%),\n",
    "may well be just smaller effects not able to become GW-significant\n",
    "with given sample sizes.\n",
    "\n",
    "The authors have explained in their later publication [Visscher et al. 2010](https://www.ncbi.nlm.nih.gov/pubmed/21142928)\n",
    "that \"During the refereeing process\n",
    "(the paper was rejected by two other journals before publication in Nature Genetics)\n",
    "and following the publication of Yang et al. (2010) it became clear to us\n",
    "that the methodology we applied, the interpretation of the results and\n",
    "the consequences of the findings on the genetic architecture of human height\n",
    "and that for other traits such as complex disease are not well understood or\n",
    "appreciated.\"\n",
    "\n",
    "Nowadays, the LMM is routinely used for SNP heritability estimation of quantitative traits.\n",
    "More recent applications suggest that for height and BMI there is not much of a\n",
    "gap anymore between the SNP heritability from very dense marker sets and\n",
    "the estimates of the total heritability by other means than through GWAS data\n",
    "[Yang et al. 2015](https://www.nature.com/articles/ng.3390),\n",
    "[Young 2019](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008222).\n",
    "At the same time, the analysis has got more complex compared to the original\n",
    "version that only included one variance component for all the SNPs,\n",
    "as in our example above.\n",
    "It is now clear that a better model should allow\n",
    "SNPs with different MAFs and different amounts of LD-tagging to\n",
    "have different variance parameters of the effect size distribution.\n",
    "In practice, this means that one should compute separate GRM-cor\n",
    "matrices for each group of SNPs that needs to be modeled separately\n",
    "and then estimate their\n",
    "variance contribution jointly, in a single LMM, that can include\n",
    "tens of different GRM-cor matrices. This is possible with both\n",
    "GCTA and BOLT-REML.\n",
    "\n",
    "Typical heritability analysis using LMM\n",
    "filters out close relatives (at least 2nd degree or closer).\n",
    "This is because closer relatives usually are also positively correlated\n",
    "in some of their environmental factors, which could create correlation\n",
    "in their phenotypes that LMM would falsely pick up as heritability.\n",
    "When the analysis is done in a homogeneous population sample\n",
    "of \"unrelated\" individuals, this concern is greatly reduced.\n",
    "Thus, LMM with GWAS data\n",
    "does something that has been impossible to imagine before:\n",
    "Estimating heritability using \"unrelated\" individuals.\n",
    "A downside of restricting the analysis to unrelateds is that the precision of\n",
    "the variance components is much smaller than if there were a larger\n",
    "range of possible relationships in the data. This means that large\n",
    "samples are needed to get useful estimates of\n",
    "SNP heritability from population data.\n",
    "\n",
    "\n",
    "What about the disease studies? The GCTA approach was also quickly [applied\n",
    "to case-control data](https://www.sciencedirect.com/science/article/pii/S0002929711000206),\n",
    "but the results are much more complicated\n",
    "to interpret than for the quantitative phenotypes.\n",
    "First, there is a transformation from binary phenotype to liability model.\n",
    "Second, there is the case-control ascertainment which complicates statistical\n",
    "modeling (as we saw with the simple covariate adjustments) and makes\n",
    "LMM in general behave unfavorably,\n",
    "as reported by [Golan et al. 2014](https://www.pnas.org/content/111/49/E5272).\n",
    "Third, typically cases and controls have been\n",
    "genotyped/handled/collected differently, and while a careful quality control\n",
    "can make GWS findings reliable and replicable, it remains a concern that the\n",
    "tiny effects picked up by the mixed model are not only polygenic effects but can\n",
    "also contain confounding effects.\n",
    "Hence interepting the variance parameters of\n",
    "a case-control GWAS data as heritabilities of diseases\n",
    "makes quite a many quite strong assumptions.\n",
    "We will come back to this important question about whether the inflation\n",
    "in a QQ-plot is confoundig or polygenicity\n",
    "in section 8.4 below.\n",
    "\n",
    "A perspective of using LMM to estimate heritability by\n",
    "[Yang et al. 2017](https://www.nature.com/articles/ng.3941)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "959328ad",
   "metadata": {},
   "source": [
    "#### 8.3.3 Mixed model in GWAS\n",
    "\n",
    "All the discussion above has been about estimating the variance parameters\n",
    "using the LMM. But mixed models have also become widely-used for\n",
    "running the primary GWAS analysis.\n",
    "\n",
    "Suppose we want to test the effect of SNP $s$ on the phenotype $Y$.\n",
    "The LMM approach does the linear regression by including in the model a random\n",
    "effect from **other variants except $s$ and its LD-friends**:\n",
    "$$y_i = \\mu + \\pmb{z}_i^T\\pmb{\\alpha} + x_{is}\\beta_s + \\eta_i + \\varepsilon_i,$$\n",
    "where $\\pmb{z}_i$ is the vector of covariate values for $i$,\n",
    "$\\eta_i = \\sum_{l=1}^{p_s} x_{il}^*\\lambda_l^*$ is the additive contribution\n",
    "of the rest of the genome *except variant $s$ and its LD-friends*\n",
    "and $\\varepsilon_i \\sim \\mathcal{N}(0,\\sigma_E^2)$\n",
    "is the (environmental) error term that is assumed to be uncorrelated across individuals.\n",
    "The random effect assumption $\\lambda_l^*\\sim \\mathcal{N}(0,\\tau^2)$\n",
    "will then lead to similar computations as above with the heritability estimation\n",
    "except that the GRM-cor matrix $G$ can now be different for different variants tested,\n",
    "and there is an additional fixed effect in the model corresponding to\n",
    "the marginal effect $\\beta_s$ of SNP $s$.\n",
    "\n",
    "In practice, when testing SNPs on, say, chromosome 1,\n",
    "GRM-cor can be computed for all other chromosomes and then used as the covariance\n",
    "structure of the random effect for all SNPs on chr 1 (e.g. BOLT-LMM and GCTA do this).\n",
    "Local updating of the random effect has also been studied\n",
    "[(Listgarten rt al. 2012)](https://www.nature.com/articles/nmeth.2037).\n",
    "\n",
    "There are (at least) two benefits for adding the rest of the genome\n",
    "as a random effect in linear regression:\n",
    "\n",
    "1. The precision of the estimator $\\widehat{\\beta}_s$\n",
    "will increase as the (often substantial) variation in phenotype\n",
    "from the rest of the genome is explained away by the model.\n",
    "This also leads to increased statistical power to detect new associations.\n",
    "\n",
    "2. The rest of the genome captures confounding effects\n",
    "that are due to relatedness and/or\n",
    "population structure and hence $\\widehat{\\beta}_s$ from a LMM\n",
    "has been automatically adjusted for these confounders.\n",
    "\n",
    "Thus, LMM is an alternative for the standard linear regression with leading\n",
    "PCs as covariates, and since LMM also increases power and also accounts for\n",
    "related individuals, it is a very useful in QT GWAS.\n",
    "\n",
    "To be on the safe side, when the data have both close relatives and\n",
    "clear population structure, then these patterns may not be correctly modelled by a single\n",
    "joint random component, and, in general, one should generate different random effects for\n",
    "each major source of phenotypic covariance in the sample.\n",
    "Typically, we still remove one from each pair of closely related individuals\n",
    "because there are not many of these pairs and it is difficult to know whether\n",
    "their environmental correlation has been correctly modelled by LMM.\n",
    "\n",
    "While LMMs have also been applied to GWAS of binary traits, until recently\n",
    "they have only been applicable to the cleanest cases of\n",
    "common variants and balanced case-control ratios.\n",
    "Recent research on efficient mixed model for\n",
    "binary phenotypes have been implemented by\n",
    "[Zhou et al. 2018](https://www.nature.com/articles/s41588-018-0184-y)\n",
    "into [SAIGE](https://github.com/weizhouUMICH/SAIGE) software and\n",
    "by [Mbatchou et al. 2020](https://www.biorxiv.org/content/10.1101/2020.06.19.162354v2)\n",
    "into [REGENIE](https://rgcgithub.github.io/regenie/overview/).\n",
    "\n",
    "\n",
    "Now, let's return to the important question that puzzled us above.\n",
    "We know that both confounding and true polygenic effects can cause\n",
    "inflation in QQ-plots and the genomic-control parameter $\\lambda$.\n",
    "How could we tell whether\n",
    "the inflation is true signal from thousands of small effects, or whether it is\n",
    "a result of some confounding bias?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45cafb24",
   "metadata": {},
   "source": [
    "### 8.4. LD-score regression (LDSC)\n",
    "\n",
    "LDSC developed by [Bulik-Sullivan et al. 2015](https://www.nature.com/articles/ng.3211)\n",
    "can separate confounding from polygenicity by utilizing LD in a clever way.\n",
    "\n",
    "Let's think about consequences of high polygenicity\n",
    "(a lot of non-zero causal effect sizes) on the observed marginal effect sizes\n",
    "when we take into account the differences in LD patterns between the variants.\n",
    "We know that for a region with $p$ variants and LD-matrix $\\pmb{R},$\n",
    "$\\pmb{\\beta}^* = \\pmb{R}\\pmb{\\lambda}^*$. If we assume a highly polygenic\n",
    "model, with each $\\lambda_l^* \\sim \\mathcal{N}(0,\\tau^2)$, we get that\n",
    "$$\\begin{aligned}\n",
    "\\textrm{E}(\\beta_l^*) &= \\textrm{E}\\left(\\sum_{k=1}^p r_{lk} \\lambda_k^*\\right) =\n",
    "\\sum_{k=1}^p r_{lk} \\textrm{E}(\\lambda_k^*) = 0,\\\\\n",
    "\\textrm{E}({\\beta_l^*}^2) &= \\textrm{E}\\left(\\sum_{k=1}^p r_{lk} \\lambda_k^*\\right)^2 =\n",
    "\\sum_{k=1}^pr_{lk}^2\\textrm{E}({\\lambda_k^*}^2) + 2 \\sum_{k<m}r_{lk}r_{lm}\\textrm{E}({\\lambda_k^*}{\\lambda_m^*})\n",
    "=\\sum_{k=1}^pr_{lk}^2\\tau^2+2\\cdot 0 = \\tau^2 r_{l+}^2 ,\n",
    "\\end{aligned}$$\n",
    "where $r_{l+}^2 = \\sum_{k=1}^p r_{lk}^2$ is the **LD-score** of SNP $l$.\n",
    "\n",
    "Thus, while the distribution of scaled causal effects was assumed independent of\n",
    "LD, the distribution of marginal effects has highest variance\n",
    "among the SNPs that tag heavily their neighbors, i.e., among the SNPs whose LD-scores\n",
    "are high. How does this property of true marginal effects\n",
    "translate to the observed estimates that we will get from a finite sample?\n",
    "\n",
    "For scaled effects, SE of the marginal effects is approximately constant\n",
    "across SNPs with small marginal effects:\n",
    "it is $\\sigma_\\varepsilon/\\sqrt{n}$ for quantitative traits\n",
    "(see below about using LD-score regression with binary traits).\n",
    "If we assume that the\n",
    "quantitative trait variance is 1 in the population, then also $\\sigma_\\varepsilon\\approx1$\n",
    "for small effects and SE is $1/\\sqrt{n}$.\n",
    "Assuming no bias due to confounders, our observed marginal effect estimate is\n",
    "$\\widehat{\\beta}_l^* = \\beta_l^* + \\xi_l$, where error is distributed as\n",
    "$\\xi_l \\sim \\mathcal{N}(0,\\textrm{SE}^2)$ and hence\n",
    "\n",
    "$$\\textrm{E}\\left(\\left(\\widehat{\\beta}_l^*\\right)^2\\right) =\n",
    "\\textrm{E}\\left({\\beta_l^*}^2 + \\xi_l^2 + 2\\beta_l^*\\xi_l \\right) =\n",
    "\\textrm{E}\\left({\\beta_l^*}^2\\right) + \\textrm{E}(\\xi_l^2) + 2\\textrm{E}\\left({\\beta_l^*}\\xi_l\\right)=\n",
    "\\tau^2 r_{l+}^2 + \\textrm{SE}^2 + 2\\cdot 0 =\n",
    "\\tau^2 r_{l+}^2 + \\frac{1}{n},\n",
    "$$\n",
    "\n",
    "It follows that the expected\n",
    "chi-square statistic observed at variant $l$ is\n",
    "$$\\textrm{E}(\\chi_l^2) =\n",
    "\\frac{\\textrm{E}\\left(\\left(\\widehat{\\beta}_l^*\\right)^2\\right)}{\\textrm{SE}^2} =\n",
    "\\frac{\\tau^2 r_{l+}^2 + \\frac{1}{n}}{\\frac 1 n}=\n",
    "n\\, \\tau^2\\, r_{l+}^2 + 1,$$\n",
    "and also increases with the LD-score of the variant.\n",
    "(Note that this quantity does not depend on the assumption of the trait variance being 1.)\n",
    "\n",
    "This derivation suggests a simple and testable\n",
    "hypothesis about GWAS summary data.\n",
    "If the trait is (highly)\n",
    "polygenic, then we should see a linear relationship between the observed\n",
    "chi-square statistics and LD-score: Variants that tag more of their neighbors\n",
    "have higher chance of tagging causal variants and hence their marginal effects\n",
    "will have larger magnitude.\n",
    "If, instead, we see overall inflation in chi-square statistic,\n",
    "but the values $(\\chi^2_l-1)$ are not proportional to the LD-scores,\n",
    "then the inflation is likely due to some confounding bias\n",
    "that is affecting the bulk of the test statistics independently of the LD-scores.\n",
    "We denote by such a constant bias factor in the chi-square test statistic by $b$.\n",
    "\n",
    "Since under the polygenic model $h^2=p\\tau^2$, we can replace $\\tau^2$\n",
    "in above formulas with $h^2/p$ and write the LD-score regression equation as\n",
    "$$\\textrm{E}(\\chi_l^2) = 1 + b +\n",
    "\\frac{n h^2}{p} r_{l+}^2.$$\n",
    "If we now take our observed chi-square statistics from a GWAS\n",
    "and regress them on the LD-scores of the variants,\n",
    "we are expecting that\n",
    "\n",
    "* the intercept is elevated from 1 if there is a confounding bias\n",
    "in the results (i.e. when $b>0$),\n",
    "\n",
    "* regardless of the possible confounding bias,\n",
    "the slope of the regression gives an estimate of the SNP heritability\n",
    "when multiplied by $p/n$.\n",
    "\n",
    "**Binary traits.** It is a custom to apply LDSC (and other heritability estimation methods)\n",
    "to binary traits in two steps. First, pretend that the trait is quantitative and compute\n",
    "heritability $h^2_\\text{obs}$ on the **observed scale**, that is, when the binary trait is\n",
    "treated as a quantitative trait with trait values 0 and 1. Second, turn that estimate on the\n",
    "**liability scale** by accounting for the population prevalence of the binary trait,\n",
    "and by also accounting for a possible case-control ascertainment.\n",
    "This is explained by [Lee et al. 2011](https://www.sciencedirect.com/science/article/pii/S0002929711000206?via%3Dihub).\n",
    "\n",
    "\n",
    "LDSC sounds like a simple and useful tool!\n",
    "Let's see some results (slides 9-11)."
   ]
  }
 ],
 "metadata": {
  "Rmd_chunk_options": {
   "author": "Matti Pirinen, University of Helsinki",
   "date": "Latest update: 25-Nov-2020, 1st version: 13-Feb-2019",
   "output": {
    "html_document": "default"
   },
   "title": "GWAS 8: Heritability and mixed models",
   "urlcolor": "blue"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ],
    [
     "R",
     "ir",
     "",
     ""
    ]
   ]
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
